{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare DepMap data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "import janitor\n",
    "from pathlib import Path\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = Path('../data')\n",
    "save_dir = Path('../modeling_data')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'sample_info.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>stripped_cell_line_name</th>\n",
       "      <th>ccle_name</th>\n",
       "      <th>sex</th>\n",
       "      <th>cas9_activity</th>\n",
       "      <th>primary_or_metastasis</th>\n",
       "      <th>primary_disease</th>\n",
       "      <th>subtype</th>\n",
       "      <th>lineage</th>\n",
       "      <th>lineage_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>NIHOVCAR3</td>\n",
       "      <td>NIHOVCAR3_OVARY</td>\n",
       "      <td>Female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>Ovarian Cancer</td>\n",
       "      <td>Adenocarcinoma, high grade serous</td>\n",
       "      <td>ovary</td>\n",
       "      <td>ovary_adenocarcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000002</td>\n",
       "      <td>HL60</td>\n",
       "      <td>HL60_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>Female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>Acute Myelogenous Leukemia (AML), M3 (Promyelo...</td>\n",
       "      <td>blood</td>\n",
       "      <td>AML</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000003</td>\n",
       "      <td>CACO2</td>\n",
       "      <td>CACO2_LARGE_INTESTINE</td>\n",
       "      <td>Male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Colon/Colorectal Cancer</td>\n",
       "      <td>Adenocarcinoma</td>\n",
       "      <td>colorectal</td>\n",
       "      <td>colorectal_adenocarcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000004</td>\n",
       "      <td>HEL</td>\n",
       "      <td>HEL_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>Male</td>\n",
       "      <td>47.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>Acute Myelogenous Leukemia (AML), M6 (Erythrol...</td>\n",
       "      <td>blood</td>\n",
       "      <td>AML</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000005</td>\n",
       "      <td>HEL9217</td>\n",
       "      <td>HEL9217_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>Male</td>\n",
       "      <td>13.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>Acute Myelogenous Leukemia (AML), M6 (Erythrol...</td>\n",
       "      <td>blood</td>\n",
       "      <td>AML</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id stripped_cell_line_name  \\\n",
       "0  ACH-000001               NIHOVCAR3   \n",
       "1  ACH-000002                    HL60   \n",
       "2  ACH-000003                   CACO2   \n",
       "3  ACH-000004                     HEL   \n",
       "4  ACH-000005                 HEL9217   \n",
       "\n",
       "                                    ccle_name     sex cas9_activity  \\\n",
       "0                             NIHOVCAR3_OVARY  Female           NaN   \n",
       "1     HL60_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE  Female           NaN   \n",
       "2                       CACO2_LARGE_INTESTINE    Male           NaN   \n",
       "3      HEL_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE    Male          47.6   \n",
       "4  HEL9217_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE    Male          13.4   \n",
       "\n",
       "  primary_or_metastasis          primary_disease  \\\n",
       "0            Metastasis           Ovarian Cancer   \n",
       "1               Primary                 Leukemia   \n",
       "2                   NaN  Colon/Colorectal Cancer   \n",
       "3                   NaN                 Leukemia   \n",
       "4                   NaN                 Leukemia   \n",
       "\n",
       "                                             subtype     lineage  \\\n",
       "0                  Adenocarcinoma, high grade serous       ovary   \n",
       "1  Acute Myelogenous Leukemia (AML), M3 (Promyelo...       blood   \n",
       "2                                     Adenocarcinoma  colorectal   \n",
       "3  Acute Myelogenous Leukemia (AML), M6 (Erythrol...       blood   \n",
       "4  Acute Myelogenous Leukemia (AML), M6 (Erythrol...       blood   \n",
       "\n",
       "             lineage_subtype  \n",
       "0       ovary_adenocarcinoma  \n",
       "1                        AML  \n",
       "2  colorectal_adenocarcinoma  \n",
       "3                        AML  \n",
       "4                        AML  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_info_columns = [\n",
    "    'depmap_id', 'stripped_cell_line_name', 'ccle_name', 'sex', \n",
    "    'cas9_activity', 'primary_or_metastasis', 'primary_disease',\n",
    "    'subtype', 'lineage', 'lineage_subtype',\n",
    "]\n",
    "\n",
    "sample_info = pd.read_csv(data_dir / 'sample_info.csv') \\\n",
    "    .clean_names() \\\n",
    "    [sample_info_columns]\n",
    "sample_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_info.to_csv(save_dir / 'sample_info.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'Achilles_guide_map.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgrna</th>\n",
       "      <th>genome_alignment</th>\n",
       "      <th>n_alignments</th>\n",
       "      <th>hugo_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAAATCCAGCAATGCAG</td>\n",
       "      <td>chr10_110964620_+</td>\n",
       "      <td>1</td>\n",
       "      <td>SHOC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAAACCCGTAGATAGCCT</td>\n",
       "      <td>chr12_95003615_+</td>\n",
       "      <td>1</td>\n",
       "      <td>NDUFA12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAAAAGAAGAAAAAACCAG</td>\n",
       "      <td>chr4_75970356_-</td>\n",
       "      <td>1</td>\n",
       "      <td>SDAD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAAAAGCTCAAGAAGGAGG</td>\n",
       "      <td>chr2_33588446_-</td>\n",
       "      <td>1</td>\n",
       "      <td>FAM98A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAAAAGGCTGTAAAAGCGT</td>\n",
       "      <td>chr19_19891600_+</td>\n",
       "      <td>1</td>\n",
       "      <td>ZNF253</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sgrna   genome_alignment  n_alignments hugo_symbol\n",
       "0  AAAAAAATCCAGCAATGCAG  chr10_110964620_+             1       SHOC2\n",
       "1  AAAAAACCCGTAGATAGCCT   chr12_95003615_+             1     NDUFA12\n",
       "2  AAAAAAGAAGAAAAAACCAG    chr4_75970356_-             1       SDAD1\n",
       "3  AAAAAAGCTCAAGAAGGAGG    chr2_33588446_-             1      FAM98A\n",
       "4  AAAAAAGGCTGTAAAAGCGT   chr19_19891600_+             1      ZNF253"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "achilles_guide_map = pd.read_csv(data_dir / 'Achilles_guide_map.csv') \\\n",
    "    .clean_names() \\\n",
    "    .assign(hugo_symbol=lambda x: [a.split(' ')[0] for a in x.gene]) \\\n",
    "    .drop(['gene'], axis=1)\n",
    "\n",
    "achilles_guide_map.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'Achilles_dropped_guides.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgrna</th>\n",
       "      <th>genomic_coordinates</th>\n",
       "      <th>gene</th>\n",
       "      <th>n_alignments</th>\n",
       "      <th>fail_reason</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAGCTTCCGCCTGATGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_aligned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAATCCTAAAATAAAATA</td>\n",
       "      <td>chrX_145827835.0_-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>in_dropped_guides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAACAGAATATAGTCAGTG</td>\n",
       "      <td>chrX_145827787.0_-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>guide_dropped_by_ceres</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAACAGGACGATGTGCGGC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_aligned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAACATCGACCGAAAGCGT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_aligned</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sgrna genomic_coordinates gene  n_alignments  \\\n",
       "0  AAAAAGCTTCCGCCTGATGG                 NaN  NaN           0.0   \n",
       "1  AAAAATCCTAAAATAAAATA  chrX_145827835.0_-  NaN           1.0   \n",
       "2  AAAACAGAATATAGTCAGTG  chrX_145827787.0_-  NaN           1.0   \n",
       "3  AAAACAGGACGATGTGCGGC                 NaN  NaN           0.0   \n",
       "4  AAAACATCGACCGAAAGCGT                 NaN  NaN           0.0   \n",
       "\n",
       "              fail_reason  \n",
       "0             not_aligned  \n",
       "1       in_dropped_guides  \n",
       "2  guide_dropped_by_ceres  \n",
       "3             not_aligned  \n",
       "4             not_aligned  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "achilles_dropped_guides = pd.read_csv(data_dir / 'Achilles_dropped_guides.csv') \\\n",
    "    .clean_names() \\\n",
    "    .rename({'unnamed_0': 'sgrna'}, axis=1)\n",
    "\n",
    "achilles_dropped_guides.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2554"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(np.unique(achilles_dropped_guides.sgrna))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "achilles_guide_map = achilles_guide_map[~achilles_guide_map.sgrna.isin(achilles_dropped_guides.sgrna)]\n",
    "achilles_guide_map = achilles_guide_map.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "achilles_guide_map.to_csv(save_dir / 'achilles_guide_map.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'Achilles_replicate_map.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>replicate_id</th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>pdna_batch</th>\n",
       "      <th>passes_qc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pacadd188-311cas9_repb_p6_batch3</td>\n",
       "      <td>ACH-001382</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>kmrc20-311cas9_repa_p6_batch3</td>\n",
       "      <td>ACH-000250</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>253j-311cas9_repa_p5_batch3</td>\n",
       "      <td>ACH-000011</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ocug1-311cas9-repb-p6_batch3</td>\n",
       "      <td>ACH-001619</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>raji-311cas9_repb_p6_batch3</td>\n",
       "      <td>ACH-000654</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       replicate_id   depmap_id  pdna_batch  passes_qc\n",
       "0  pacadd188-311cas9_repb_p6_batch3  ACH-001382           3       True\n",
       "1     kmrc20-311cas9_repa_p6_batch3  ACH-000250           3       True\n",
       "2       253j-311cas9_repa_p5_batch3  ACH-000011           3       True\n",
       "3      ocug1-311cas9-repb-p6_batch3  ACH-001619           3       True\n",
       "4       raji-311cas9_repb_p6_batch3  ACH-000654           3       True"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "achilles_replicate_map = pd.read_csv(data_dir / 'Achilles_replicate_map.csv') \\\n",
    "    .clean_names() \\\n",
    "    .assign(replicate_id=lambda x: x.replicate_id.str.lower())\n",
    "achilles_replicate_map.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.max(achilles_replicate_map.pdna_batch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1638"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(np.unique(achilles_replicate_map.replicate_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.996"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.round(np.mean(achilles_replicate_map.passes_qc), 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "achilles_replicate_map.to_csv(save_dir / 'achilles_replicate_map.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'Achilles_logfold_change.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgrna</th>\n",
       "      <th>replicate_id</th>\n",
       "      <th>lfc</th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>pdna_batch</th>\n",
       "      <th>passes_qc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAAATCCAGCAATGCAG</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>0.289694</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAAACCCGTAGATAGCCT</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>0.170172</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAAAAGAAGAAAAAACCAG</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>-0.695947</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAAAAGCTCAAGAAGGAGG</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>-0.324935</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAAAAGGCTGTAAAAGCGT</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>0.142874</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sgrna                 replicate_id       lfc   depmap_id  \\\n",
       "0  AAAAAAATCCAGCAATGCAG  143b-311cas9_repa_p6_batch3  0.289694  ACH-001001   \n",
       "1  AAAAAACCCGTAGATAGCCT  143b-311cas9_repa_p6_batch3  0.170172  ACH-001001   \n",
       "2  AAAAAAGAAGAAAAAACCAG  143b-311cas9_repa_p6_batch3 -0.695947  ACH-001001   \n",
       "3  AAAAAAGCTCAAGAAGGAGG  143b-311cas9_repa_p6_batch3 -0.324935  ACH-001001   \n",
       "4  AAAAAAGGCTGTAAAAGCGT  143b-311cas9_repa_p6_batch3  0.142874  ACH-001001   \n",
       "\n",
       "   pdna_batch  passes_qc  \n",
       "0           3       True  \n",
       "1           3       True  \n",
       "2           3       True  \n",
       "3           3       True  \n",
       "4           3       True  "
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "achilles_logfold_change = pd.read_csv(data_dir / 'Achilles_logfold_change.csv') \\\n",
    "    .rename({'Construct Barcode': 'sgrna'}, axis=1) \\\n",
    "    .set_index('sgrna') \\\n",
    "    .melt(var_name='replicate_id', value_name='lfc', ignore_index=False) \\\n",
    "    .reset_index() \\\n",
    "    .assign(replicate_id=lambda x: x.replicate_id.str.lower()) \\\n",
    "    .merge(achilles_replicate_map, on='replicate_id', how='left') \\\n",
    "    .pipe(lambda x: x[x.passes_qc])\n",
    "\n",
    "achilles_logfold_change.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "if achilles_logfold_change.depmap_id.isnull().values.any():\n",
    "    raise Exception('Some data points are missing cell line assignments.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not np.all(achilles_logfold_change.passes_qc):\n",
    "    raise Exception('Some data does not pass QC.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(121067627, 6)"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "achilles_logfold_change.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "achilles_logfold_change.to_csv(save_dir / 'achilles_logfold_change.csv', \n",
    "                               index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'CCLE_mutations.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>hugo_symbol</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>tumor_seq_allele1</th>\n",
       "      <th>cdna_change</th>\n",
       "      <th>codon_change</th>\n",
       "      <th>protein_change</th>\n",
       "      <th>isdeleterious</th>\n",
       "      <th>istcgahotspot</th>\n",
       "      <th>iscosmichotspot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000986</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858743</td>\n",
       "      <td>58858743</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>c.1456G&gt;A</td>\n",
       "      <td>c.(1456-1458)Gac&gt;Aac</td>\n",
       "      <td>p.D486N</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000988</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858810</td>\n",
       "      <td>58858810</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>c.1389G&gt;A</td>\n",
       "      <td>c.(1387-1389)caG&gt;caA</td>\n",
       "      <td>p.Q463Q</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-002182</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858867</td>\n",
       "      <td>58858867</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>c.1332G&gt;C</td>\n",
       "      <td>c.(1330-1332)aaG&gt;aaC</td>\n",
       "      <td>p.K444N</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000985</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858872</td>\n",
       "      <td>58858872</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>c.1327G&gt;A</td>\n",
       "      <td>c.(1327-1329)Gtg&gt;Atg</td>\n",
       "      <td>p.V443M</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-001793</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858914</td>\n",
       "      <td>58858914</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>c.1285G&gt;A</td>\n",
       "      <td>c.(1285-1287)Gac&gt;Aac</td>\n",
       "      <td>p.D429N</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id hugo_symbol chromosome  start_position  end_position  \\\n",
       "0  ACH-000986        A1BG         19        58858743      58858743   \n",
       "1  ACH-000988        A1BG         19        58858810      58858810   \n",
       "2  ACH-002182        A1BG         19        58858867      58858867   \n",
       "3  ACH-000985        A1BG         19        58858872      58858872   \n",
       "4  ACH-001793        A1BG         19        58858914      58858914   \n",
       "\n",
       "  variant_classification variant_type reference_allele tumor_seq_allele1  \\\n",
       "0      Missense_Mutation          SNP                C                 T   \n",
       "1                 Silent          SNP                C                 T   \n",
       "2      Missense_Mutation          SNP                C                 G   \n",
       "3      Missense_Mutation          SNP                C                 T   \n",
       "4      Missense_Mutation          SNP                C                 T   \n",
       "\n",
       "  cdna_change          codon_change protein_change isdeleterious  \\\n",
       "0   c.1456G>A  c.(1456-1458)Gac>Aac        p.D486N         False   \n",
       "1   c.1389G>A  c.(1387-1389)caG>caA        p.Q463Q         False   \n",
       "2   c.1332G>C  c.(1330-1332)aaG>aaC        p.K444N         False   \n",
       "3   c.1327G>A  c.(1327-1329)Gtg>Atg        p.V443M         False   \n",
       "4   c.1285G>A  c.(1285-1287)Gac>Aac        p.D429N         False   \n",
       "\n",
       "   istcgahotspot  iscosmichotspot  \n",
       "0          False            False  \n",
       "1          False            False  \n",
       "2          False            False  \n",
       "3          False            False  \n",
       "4          False            False  "
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccle_mutations_columns = [\n",
    "    'depmap_id', \n",
    "    'hugo_symbol', 'chromosome', 'start_position', 'end_position',\n",
    "    'variant_classification', 'variant_type', 'reference_allele', \n",
    "    'tumor_seq_allele1', 'cdna_change', 'codon_change', 'protein_change',\n",
    "    'isdeleterious', 'istcgahotspot', 'iscosmichotspot'\n",
    "]\n",
    "\n",
    "ccle_mutations = pd.read_csv(data_dir / 'CCLE_mutations.csv', \n",
    "                             delimiter='\\t', low_memory=False) \\\n",
    "    .clean_names() \\\n",
    "    [ccle_mutations_columns]\n",
    "\n",
    "ccle_mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "ccle_mutations.to_csv(save_dir / 'ccle_mutations.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### *KRAS* mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>protein_change</th>\n",
       "      <th>isdeleterious</th>\n",
       "      <th>istcgahotspot</th>\n",
       "      <th>iscosmichotspot</th>\n",
       "      <th>codon</th>\n",
       "      <th>is_kras_hotspot</th>\n",
       "      <th>kras_allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000981</td>\n",
       "      <td>25368390</td>\n",
       "      <td>25368390</td>\n",
       "      <td>frame_shift_del</td>\n",
       "      <td>del</td>\n",
       "      <td>p.K185fs</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>185</td>\n",
       "      <td>False</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-001650</td>\n",
       "      <td>25368390</td>\n",
       "      <td>25368390</td>\n",
       "      <td>frame_shift_del</td>\n",
       "      <td>del</td>\n",
       "      <td>p.K185fs</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>185</td>\n",
       "      <td>False</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-002238</td>\n",
       "      <td>25368390</td>\n",
       "      <td>25368390</td>\n",
       "      <td>frame_shift_del</td>\n",
       "      <td>del</td>\n",
       "      <td>p.K185fs</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>185</td>\n",
       "      <td>False</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000996</td>\n",
       "      <td>25368455</td>\n",
       "      <td>25368455</td>\n",
       "      <td>nonsense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.R164*</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>164</td>\n",
       "      <td>False</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000218</td>\n",
       "      <td>25378561</td>\n",
       "      <td>25378561</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.A146V</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>146</td>\n",
       "      <td>True</td>\n",
       "      <td>A146V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id  start_position  end_position variant_classification  \\\n",
       "0  ACH-000981        25368390      25368390        frame_shift_del   \n",
       "1  ACH-001650        25368390      25368390        frame_shift_del   \n",
       "2  ACH-002238        25368390      25368390        frame_shift_del   \n",
       "3  ACH-000996        25368455      25368455      nonsense_mutation   \n",
       "4  ACH-000218        25378561      25378561      missense_mutation   \n",
       "\n",
       "  variant_type protein_change isdeleterious  istcgahotspot  iscosmichotspot  \\\n",
       "0          del       p.K185fs          True          False            False   \n",
       "1          del       p.K185fs          True          False            False   \n",
       "2          del       p.K185fs          True          False            False   \n",
       "3          snp        p.R164*          True          False            False   \n",
       "4          snp        p.A146V         False           True             True   \n",
       "\n",
       "  codon  is_kras_hotspot kras_allele  \n",
       "0   185            False       other  \n",
       "1   185            False       other  \n",
       "2   185            False       other  \n",
       "3   164            False       other  \n",
       "4   146             True       A146V  "
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras_mutations_columns = [\n",
    "    'depmap_id', 'start_position', 'end_position', \n",
    "    'variant_classification', 'variant_type', 'protein_change',\n",
    "    'isdeleterious', 'istcgahotspot', 'iscosmichotspot'\n",
    "]\n",
    "\n",
    "kras_hotspot_codons = ['12', '13', '61', '146']\n",
    "\n",
    "kras_mutations = ccle_mutations[ccle_mutations.hugo_symbol == 'KRAS'] \\\n",
    "    [kras_mutations_columns] \\\n",
    "    .assign(\n",
    "        variant_classification=lambda x: x.variant_classification.str.lower(),\n",
    "        variant_type=lambda x: x.variant_type.str.lower(),\n",
    "        codon=lambda x: [re.sub('\\D', '', a) for a in x.protein_change],\n",
    "        is_kras_hotspot=lambda x: x.codon.isin(kras_hotspot_codons)\n",
    "    ) \\\n",
    "    .pipe(lambda x: x[x.variant_classification != 'silent']) \\\n",
    "    .pipe(lambda x: x[x.is_kras_hotspot | x.iscosmichotspot | x.istcgahotspot | x.isdeleterious]) \\\n",
    "    .assign(kras_allele=lambda x: [a if b else \"other\" for a,b in zip(x.protein_change, x.is_kras_hotspot)]) \\\n",
    "    .assign(kras_allele=lambda x: [re.sub('p.', '', a) for a in x.kras_allele]) \\\n",
    "    .drop_duplicates() \\\n",
    "    .reset_index(drop=True)\n",
    "\n",
    "\n",
    "kras_mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kras_allele</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G12D</th>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12V</th>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12C</th>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>other</th>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12A</th>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G13D</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q61H</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12S</th>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12R</th>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A146T</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G13C</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q61K</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A146V</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q61L</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q61R</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12F</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A146P</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             depmap_id\n",
       "kras_allele           \n",
       "G12D                68\n",
       "G12V                47\n",
       "G12C                27\n",
       "other               18\n",
       "G12A                16\n",
       "G13D                14\n",
       "Q61H                 9\n",
       "G12S                 8\n",
       "G12R                 8\n",
       "A146T                7\n",
       "G13C                 6\n",
       "Q61K                 4\n",
       "A146V                3\n",
       "Q61L                 2\n",
       "Q61R                 2\n",
       "G12F                 1\n",
       "A146P                1"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras_mutations[['kras_allele', 'depmap_id']] \\\n",
    "    .groupby('kras_allele') \\\n",
    "    .count() \\\n",
    "    .sort_values('depmap_id', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "kras_mutations.to_csv(save_dir / 'kras_mutations.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>kras_allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000718</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000249</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000264</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000314</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000344</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACH-001094</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ACH-001378</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ACH-001650</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ACH-001857</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id  kras_allele\n",
       "0  ACH-000718            3\n",
       "1  ACH-000249            2\n",
       "2  ACH-000264            2\n",
       "3  ACH-000314            2\n",
       "4  ACH-000344            2\n",
       "5  ACH-001001            2\n",
       "6  ACH-001094            2\n",
       "7  ACH-001378            2\n",
       "8  ACH-001650            2\n",
       "9  ACH-001857            2"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mult_kras_mutations = kras_mutations[['depmap_id', 'kras_allele']] \\\n",
    "    .groupby('depmap_id') \\\n",
    "    .count() \\\n",
    "    .pipe(lambda x: x[x.kras_allele > 1]) \\\n",
    "    .sort_values('kras_allele', ascending=False) \\\n",
    "    .reset_index(drop=False)\n",
    "\n",
    "mult_kras_mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>protein_change</th>\n",
       "      <th>isdeleterious</th>\n",
       "      <th>istcgahotspot</th>\n",
       "      <th>iscosmichotspot</th>\n",
       "      <th>codon</th>\n",
       "      <th>is_kras_hotspot</th>\n",
       "      <th>kras_allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000249</td>\n",
       "      <td>25380275</td>\n",
       "      <td>25380275</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61H</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000264</td>\n",
       "      <td>25380277</td>\n",
       "      <td>25380277</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACH-000264</td>\n",
       "      <td>25380277</td>\n",
       "      <td>25380278</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>dnp</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000314</td>\n",
       "      <td>25380275</td>\n",
       "      <td>25380275</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61H</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ACH-000314</td>\n",
       "      <td>25398282</td>\n",
       "      <td>25398282</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G13C</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>13</td>\n",
       "      <td>True</td>\n",
       "      <td>G13C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000344</td>\n",
       "      <td>25380277</td>\n",
       "      <td>25380277</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACH-000344</td>\n",
       "      <td>25380277</td>\n",
       "      <td>25380278</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>dnp</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ACH-000718</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ACH-000718</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12C</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ACH-000718</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>dnp</td>\n",
       "      <td>p.G12F</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12S</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ACH-001094</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12S</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ACH-001378</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ACH-001378</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12S</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ACH-001650</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12D</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ACH-001857</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12C</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ACH-001857</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     depmap_id  start_position  end_position variant_classification  \\\n",
       "1   ACH-000249        25380275      25380275      missense_mutation   \n",
       "3   ACH-000264        25380277      25380277      missense_mutation   \n",
       "6   ACH-000264        25380277      25380278      missense_mutation   \n",
       "2   ACH-000314        25380275      25380275      missense_mutation   \n",
       "10  ACH-000314        25398282      25398282      missense_mutation   \n",
       "4   ACH-000344        25380277      25380277      missense_mutation   \n",
       "5   ACH-000344        25380277      25380278      missense_mutation   \n",
       "11  ACH-000718        25398284      25398284      missense_mutation   \n",
       "16  ACH-000718        25398285      25398285      missense_mutation   \n",
       "15  ACH-000718        25398284      25398285      missense_mutation   \n",
       "18  ACH-001001        25398285      25398285      missense_mutation   \n",
       "19  ACH-001094        25398285      25398285      missense_mutation   \n",
       "12  ACH-001378        25398284      25398284      missense_mutation   \n",
       "20  ACH-001378        25398285      25398285      missense_mutation   \n",
       "14  ACH-001650        25398284      25398284      missense_mutation   \n",
       "17  ACH-001857        25398285      25398285      missense_mutation   \n",
       "13  ACH-001857        25398284      25398284      missense_mutation   \n",
       "\n",
       "   variant_type protein_change isdeleterious  istcgahotspot  iscosmichotspot  \\\n",
       "1           snp         p.Q61H         False           True             True   \n",
       "3           snp         p.Q61K         False           True             True   \n",
       "6           dnp         p.Q61K         False           True             True   \n",
       "2           snp         p.Q61H         False           True             True   \n",
       "10          snp         p.G13C         False           True             True   \n",
       "4           snp         p.Q61K         False           True             True   \n",
       "5           dnp         p.Q61K         False           True             True   \n",
       "11          snp         p.G12V         False           True             True   \n",
       "16          snp         p.G12C         False           True             True   \n",
       "15          dnp         p.G12F         False           True             True   \n",
       "18          snp         p.G12S         False           True             True   \n",
       "19          snp         p.G12S         False           True             True   \n",
       "12          snp         p.G12V         False           True             True   \n",
       "20          snp         p.G12S         False           True             True   \n",
       "14          snp         p.G12D         False           True             True   \n",
       "17          snp         p.G12C         False           True             True   \n",
       "13          snp         p.G12V         False           True             True   \n",
       "\n",
       "   codon  is_kras_hotspot kras_allele  \n",
       "1     61             True        Q61H  \n",
       "3     61             True        Q61K  \n",
       "6     61             True        Q61K  \n",
       "2     61             True        Q61H  \n",
       "10    13             True        G13C  \n",
       "4     61             True        Q61K  \n",
       "5     61             True        Q61K  \n",
       "11    12             True        G12V  \n",
       "16    12             True        G12C  \n",
       "15    12             True        G12F  \n",
       "18    12             True        G12S  \n",
       "19    12             True        G12S  \n",
       "12    12             True        G12V  \n",
       "20    12             True        G12S  \n",
       "14    12             True        G12D  \n",
       "17    12             True        G12C  \n",
       "13    12             True        G12V  "
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras_mult_mutations_fix = kras_mutations[kras_mutations.depmap_id.isin(mult_kras_mutations.depmap_id)] \\\n",
    "    .reset_index(drop=True) \\\n",
    "    .pipe(lambda x: x[x.is_kras_hotspot]) \\\n",
    "    .sort_values('depmap_id')\n",
    "\n",
    "kras_mult_mutations_fix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>kras_allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000264</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000314</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000344</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-001378</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-001857</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACH-000249</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ACH-001094</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ACH-001650</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id  kras_allele\n",
       "0  ACH-000264            2\n",
       "1  ACH-000314            2\n",
       "2  ACH-000344            2\n",
       "3  ACH-001378            2\n",
       "4  ACH-001857            2\n",
       "5  ACH-000249            1\n",
       "6  ACH-001001            1\n",
       "7  ACH-001094            1\n",
       "8  ACH-001650            1"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras_mut_blacklist = [\n",
    "    'ACH-000718'\n",
    "]\n",
    "\n",
    "kras_mult_mutations_fix[['depmap_id', 'kras_allele']] \\\n",
    "    .pipe(lambda x: x[~x.depmap_id.isin(kras_mut_blacklist)]) \\\n",
    "    .groupby('depmap_id') \\\n",
    "    .count() \\\n",
    "    .sort_values('kras_allele', ascending=False) \\\n",
    "    .reset_index(drop=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
