{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare DepMap data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "import janitor\n",
    "from pathlib import Path\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = Path(\"../data/depmap_20q3\")\n",
    "save_dir = Path(\"../modeling_data\")\n",
    "\n",
    "if not data_dir.exists():\n",
    "    raise Exception(\n",
    "        \"Data directory does not exist. Make sure to download the DepMap data.\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'sample_info.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>stripped_cell_line_name</th>\n",
       "      <th>ccle_name</th>\n",
       "      <th>sex</th>\n",
       "      <th>cas9_activity</th>\n",
       "      <th>primary_or_metastasis</th>\n",
       "      <th>primary_disease</th>\n",
       "      <th>subtype</th>\n",
       "      <th>lineage</th>\n",
       "      <th>lineage_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>NIHOVCAR3</td>\n",
       "      <td>NIHOVCAR3_OVARY</td>\n",
       "      <td>Female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>Ovarian Cancer</td>\n",
       "      <td>Adenocarcinoma, high grade serous</td>\n",
       "      <td>ovary</td>\n",
       "      <td>ovary_adenocarcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000002</td>\n",
       "      <td>HL60</td>\n",
       "      <td>HL60_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>Female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>Acute Myelogenous Leukemia (AML), M3 (Promyelo...</td>\n",
       "      <td>blood</td>\n",
       "      <td>AML</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000003</td>\n",
       "      <td>CACO2</td>\n",
       "      <td>CACO2_LARGE_INTESTINE</td>\n",
       "      <td>Male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Colon/Colorectal Cancer</td>\n",
       "      <td>Adenocarcinoma</td>\n",
       "      <td>colorectal</td>\n",
       "      <td>colorectal_adenocarcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000004</td>\n",
       "      <td>HEL</td>\n",
       "      <td>HEL_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>Male</td>\n",
       "      <td>52.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>Acute Myelogenous Leukemia (AML), M6 (Erythrol...</td>\n",
       "      <td>blood</td>\n",
       "      <td>AML</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000005</td>\n",
       "      <td>HEL9217</td>\n",
       "      <td>HEL9217_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>Male</td>\n",
       "      <td>86.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>Acute Myelogenous Leukemia (AML), M6 (Erythrol...</td>\n",
       "      <td>blood</td>\n",
       "      <td>AML</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id stripped_cell_line_name  \\\n",
       "0  ACH-000001               NIHOVCAR3   \n",
       "1  ACH-000002                    HL60   \n",
       "2  ACH-000003                   CACO2   \n",
       "3  ACH-000004                     HEL   \n",
       "4  ACH-000005                 HEL9217   \n",
       "\n",
       "                                    ccle_name     sex cas9_activity  \\\n",
       "0                             NIHOVCAR3_OVARY  Female           NaN   \n",
       "1     HL60_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE  Female           NaN   \n",
       "2                       CACO2_LARGE_INTESTINE    Male           NaN   \n",
       "3      HEL_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE    Male          52.4   \n",
       "4  HEL9217_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE    Male          86.6   \n",
       "\n",
       "  primary_or_metastasis          primary_disease  \\\n",
       "0            Metastasis           Ovarian Cancer   \n",
       "1               Primary                 Leukemia   \n",
       "2                   NaN  Colon/Colorectal Cancer   \n",
       "3                   NaN                 Leukemia   \n",
       "4                   NaN                 Leukemia   \n",
       "\n",
       "                                             subtype     lineage  \\\n",
       "0                  Adenocarcinoma, high grade serous       ovary   \n",
       "1  Acute Myelogenous Leukemia (AML), M3 (Promyelo...       blood   \n",
       "2                                     Adenocarcinoma  colorectal   \n",
       "3  Acute Myelogenous Leukemia (AML), M6 (Erythrol...       blood   \n",
       "4  Acute Myelogenous Leukemia (AML), M6 (Erythrol...       blood   \n",
       "\n",
       "             lineage_subtype  \n",
       "0       ovary_adenocarcinoma  \n",
       "1                        AML  \n",
       "2  colorectal_adenocarcinoma  \n",
       "3                        AML  \n",
       "4                        AML  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_info_columns = [\n",
    "    \"depmap_id\",\n",
    "    \"stripped_cell_line_name\",\n",
    "    \"ccle_name\",\n",
    "    \"sex\",\n",
    "    \"cas9_activity\",\n",
    "    \"primary_or_metastasis\",\n",
    "    \"primary_disease\",\n",
    "    \"subtype\",\n",
    "    \"lineage\",\n",
    "    \"lineage_subtype\",\n",
    "]\n",
    "\n",
    "sample_info = pd.read_csv(data_dir / \"sample_info.csv\").clean_names()[\n",
    "    sample_info_columns\n",
    "]\n",
    "sample_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_info.to_csv(save_dir / \"sample_info.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'Achilles_guide_map.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgrna</th>\n",
       "      <th>genome_alignment</th>\n",
       "      <th>n_alignments</th>\n",
       "      <th>hugo_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAAATCCAGCAATGCAG</td>\n",
       "      <td>chr10_110964620_+</td>\n",
       "      <td>1</td>\n",
       "      <td>SHOC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAAACCCGTAGATAGCCT</td>\n",
       "      <td>chr12_95003615_+</td>\n",
       "      <td>1</td>\n",
       "      <td>NDUFA12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAAAAGAAGAAAAAACCAG</td>\n",
       "      <td>chr4_75970356_-</td>\n",
       "      <td>1</td>\n",
       "      <td>SDAD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAAAAGCTCAAGAAGGAGG</td>\n",
       "      <td>chr2_33588446_-</td>\n",
       "      <td>1</td>\n",
       "      <td>FAM98A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAAAAGGCTGTAAAAGCGT</td>\n",
       "      <td>chr19_19891600_+</td>\n",
       "      <td>1</td>\n",
       "      <td>ZNF253</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sgrna   genome_alignment  n_alignments hugo_symbol\n",
       "0  AAAAAAATCCAGCAATGCAG  chr10_110964620_+             1       SHOC2\n",
       "1  AAAAAACCCGTAGATAGCCT   chr12_95003615_+             1     NDUFA12\n",
       "2  AAAAAAGAAGAAAAAACCAG    chr4_75970356_-             1       SDAD1\n",
       "3  AAAAAAGCTCAAGAAGGAGG    chr2_33588446_-             1      FAM98A\n",
       "4  AAAAAAGGCTGTAAAAGCGT   chr19_19891600_+             1      ZNF253"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "achilles_guide_map = (\n",
    "    pd.read_csv(data_dir / \"Achilles_guide_map.csv\")\n",
    "    .clean_names()\n",
    "    .assign(hugo_symbol=lambda x: [a.split(\" \")[0] for a in x.gene])\n",
    "    .drop([\"gene\"], axis=1)\n",
    ")\n",
    "\n",
    "achilles_guide_map.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'Achilles_dropped_guides.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgrna</th>\n",
       "      <th>genomic_coordinates</th>\n",
       "      <th>gene</th>\n",
       "      <th>n_alignments</th>\n",
       "      <th>fail_reason</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAGCTTCCGCCTGATGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_aligned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAATCCTAAAATAAAATA</td>\n",
       "      <td>chrX_145827835.0_-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>in_dropped_guides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAACAGAATATAGTCAGTG</td>\n",
       "      <td>chrX_145827787.0_-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>guide_dropped_by_ceres</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAACAGGACGATGTGCGGC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_aligned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAACATCGACCGAAAGCGT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_aligned</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sgrna genomic_coordinates gene  n_alignments  \\\n",
       "0  AAAAAGCTTCCGCCTGATGG                 NaN  NaN           0.0   \n",
       "1  AAAAATCCTAAAATAAAATA  chrX_145827835.0_-  NaN           1.0   \n",
       "2  AAAACAGAATATAGTCAGTG  chrX_145827787.0_-  NaN           1.0   \n",
       "3  AAAACAGGACGATGTGCGGC                 NaN  NaN           0.0   \n",
       "4  AAAACATCGACCGAAAGCGT                 NaN  NaN           0.0   \n",
       "\n",
       "              fail_reason  \n",
       "0             not_aligned  \n",
       "1       in_dropped_guides  \n",
       "2  guide_dropped_by_ceres  \n",
       "3             not_aligned  \n",
       "4             not_aligned  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "achilles_dropped_guides = (\n",
    "    pd.read_csv(data_dir / \"Achilles_dropped_guides.csv\")\n",
    "    .clean_names()\n",
    "    .rename({\"unnamed_0\": \"sgrna\"}, axis=1)\n",
    ")\n",
    "\n",
    "achilles_dropped_guides.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2554"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(np.unique(achilles_dropped_guides.sgrna))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "achilles_guide_map = achilles_guide_map[\n",
    "    ~achilles_guide_map.sgrna.isin(achilles_dropped_guides.sgrna)\n",
    "]\n",
    "achilles_guide_map = achilles_guide_map.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "achilles_guide_map.to_csv(save_dir / \"achilles_guide_map.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'Achilles_replicate_map.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>replicate_id</th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>pdna_batch</th>\n",
       "      <th>passes_qc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pe_ca-pj-34clone c12-311cas9_repb_p5_batch2</td>\n",
       "      <td>ACH-000606</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nci-h1944-311cas9 rep c p5_batch3</td>\n",
       "      <td>ACH-000414</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mdamb231-311cas9_repb_p5_batch2</td>\n",
       "      <td>ACH-000768</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>skrc20-311cas9_repa_p6_batch3</td>\n",
       "      <td>ACH-000385</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>skpndw-311cas9_repb_p6_batch3</td>\n",
       "      <td>ACH-001193</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  replicate_id   depmap_id  pdna_batch  \\\n",
       "0  pe_ca-pj-34clone c12-311cas9_repb_p5_batch2  ACH-000606           2   \n",
       "1            nci-h1944-311cas9 rep c p5_batch3  ACH-000414           3   \n",
       "2              mdamb231-311cas9_repb_p5_batch2  ACH-000768           2   \n",
       "3                skrc20-311cas9_repa_p6_batch3  ACH-000385           3   \n",
       "4                skpndw-311cas9_repb_p6_batch3  ACH-001193           3   \n",
       "\n",
       "   passes_qc  \n",
       "0       True  \n",
       "1       True  \n",
       "2       True  \n",
       "3       True  \n",
       "4       True  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "achilles_replicate_map = (\n",
    "    pd.read_csv(data_dir / \"Achilles_replicate_map.csv\")\n",
    "    .clean_names()\n",
    "    .assign(replicate_id=lambda x: x.replicate_id.str.lower())\n",
    ")\n",
    "achilles_replicate_map.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.max(achilles_replicate_map.pdna_batch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1759"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(np.unique(achilles_replicate_map.replicate_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.948"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.round(np.mean(achilles_replicate_map.passes_qc), 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "achilles_replicate_map.to_csv(save_dir / \"achilles_replicate_map.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'Achilles_logfold_change.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgrna</th>\n",
       "      <th>replicate_id</th>\n",
       "      <th>lfc</th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>pdna_batch</th>\n",
       "      <th>passes_qc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAAATCCAGCAATGCAG</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>0.289694</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAAACCCGTAGATAGCCT</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>0.170172</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAAAAGAAGAAAAAACCAG</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>-0.695947</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAAAAGCTCAAGAAGGAGG</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>-0.324935</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAAAAGGCTGTAAAAGCGT</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>0.142874</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sgrna                 replicate_id       lfc   depmap_id  \\\n",
       "0  AAAAAAATCCAGCAATGCAG  143b-311cas9_repa_p6_batch3  0.289694  ACH-001001   \n",
       "1  AAAAAACCCGTAGATAGCCT  143b-311cas9_repa_p6_batch3  0.170172  ACH-001001   \n",
       "2  AAAAAAGAAGAAAAAACCAG  143b-311cas9_repa_p6_batch3 -0.695947  ACH-001001   \n",
       "3  AAAAAAGCTCAAGAAGGAGG  143b-311cas9_repa_p6_batch3 -0.324935  ACH-001001   \n",
       "4  AAAAAAGGCTGTAAAAGCGT  143b-311cas9_repa_p6_batch3  0.142874  ACH-001001   \n",
       "\n",
       "   pdna_batch  passes_qc  \n",
       "0           3       True  \n",
       "1           3       True  \n",
       "2           3       True  \n",
       "3           3       True  \n",
       "4           3       True  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "achilles_logfold_change = (\n",
    "    pd.read_csv(data_dir / \"Achilles_logfold_change.csv\")\n",
    "    .rename({\"Construct Barcode\": \"sgrna\"}, axis=1)\n",
    "    .set_index(\"sgrna\")\n",
    "    .melt(var_name=\"replicate_id\", value_name=\"lfc\", ignore_index=False)\n",
    "    .reset_index()\n",
    "    .assign(replicate_id=lambda x: x.replicate_id.str.lower())\n",
    "    .merge(achilles_replicate_map, on=\"replicate_id\", how=\"left\")\n",
    "    .pipe(lambda x: x[x.passes_qc])\n",
    ")\n",
    "\n",
    "achilles_logfold_change.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "if achilles_logfold_change.depmap_id.isnull().values.any():\n",
    "    raise Exception(\"Some data points are missing cell line assignments.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not np.all(achilles_logfold_change.passes_qc):\n",
    "    raise Exception(\"Some data does not pass QC.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(123756359, 6)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "achilles_logfold_change.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "achilles_logfold_change.to_csv(save_dir / \"achilles_logfold_change.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'CCLE_mutations.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>hugo_symbol</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>tumor_seq_allele1</th>\n",
       "      <th>cdna_change</th>\n",
       "      <th>codon_change</th>\n",
       "      <th>protein_change</th>\n",
       "      <th>isdeleterious</th>\n",
       "      <th>istcgahotspot</th>\n",
       "      <th>iscosmichotspot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000986</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858743</td>\n",
       "      <td>58858743</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>c.1456G&gt;A</td>\n",
       "      <td>c.(1456-1458)Gac&gt;Aac</td>\n",
       "      <td>p.D486N</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000988</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858810</td>\n",
       "      <td>58858810</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>c.1389G&gt;A</td>\n",
       "      <td>c.(1387-1389)caG&gt;caA</td>\n",
       "      <td>p.Q463Q</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-002182</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858867</td>\n",
       "      <td>58858867</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>c.1332G&gt;C</td>\n",
       "      <td>c.(1330-1332)aaG&gt;aaC</td>\n",
       "      <td>p.K444N</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000985</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858872</td>\n",
       "      <td>58858872</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>c.1327G&gt;A</td>\n",
       "      <td>c.(1327-1329)Gtg&gt;Atg</td>\n",
       "      <td>p.V443M</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-001793</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858914</td>\n",
       "      <td>58858914</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>c.1285G&gt;A</td>\n",
       "      <td>c.(1285-1287)Gac&gt;Aac</td>\n",
       "      <td>p.D429N</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id hugo_symbol chromosome  start_position  end_position  \\\n",
       "0  ACH-000986        A1BG         19        58858743      58858743   \n",
       "1  ACH-000988        A1BG         19        58858810      58858810   \n",
       "2  ACH-002182        A1BG         19        58858867      58858867   \n",
       "3  ACH-000985        A1BG         19        58858872      58858872   \n",
       "4  ACH-001793        A1BG         19        58858914      58858914   \n",
       "\n",
       "  variant_classification variant_type reference_allele tumor_seq_allele1  \\\n",
       "0      Missense_Mutation          SNP                C                 T   \n",
       "1                 Silent          SNP                C                 T   \n",
       "2      Missense_Mutation          SNP                C                 G   \n",
       "3      Missense_Mutation          SNP                C                 T   \n",
       "4      Missense_Mutation          SNP                C                 T   \n",
       "\n",
       "  cdna_change          codon_change protein_change isdeleterious  \\\n",
       "0   c.1456G>A  c.(1456-1458)Gac>Aac        p.D486N         False   \n",
       "1   c.1389G>A  c.(1387-1389)caG>caA        p.Q463Q         False   \n",
       "2   c.1332G>C  c.(1330-1332)aaG>aaC        p.K444N         False   \n",
       "3   c.1327G>A  c.(1327-1329)Gtg>Atg        p.V443M         False   \n",
       "4   c.1285G>A  c.(1285-1287)Gac>Aac        p.D429N         False   \n",
       "\n",
       "   istcgahotspot  iscosmichotspot  \n",
       "0          False            False  \n",
       "1          False            False  \n",
       "2          False            False  \n",
       "3          False            False  \n",
       "4          False            False  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccle_mutations_columns = [\n",
    "    \"depmap_id\",\n",
    "    \"hugo_symbol\",\n",
    "    \"chromosome\",\n",
    "    \"start_position\",\n",
    "    \"end_position\",\n",
    "    \"variant_classification\",\n",
    "    \"variant_type\",\n",
    "    \"reference_allele\",\n",
    "    \"tumor_seq_allele1\",\n",
    "    \"cdna_change\",\n",
    "    \"codon_change\",\n",
    "    \"protein_change\",\n",
    "    \"isdeleterious\",\n",
    "    \"istcgahotspot\",\n",
    "    \"iscosmichotspot\",\n",
    "]\n",
    "\n",
    "ccle_mutations = pd.read_csv(\n",
    "    data_dir / \"CCLE_mutations.csv\", delimiter=\"\\t\", low_memory=False\n",
    ").clean_names()[ccle_mutations_columns]\n",
    "\n",
    "ccle_mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "ccle_mutations.to_csv(save_dir / \"ccle_mutations.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### *KRAS* mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>protein_change</th>\n",
       "      <th>isdeleterious</th>\n",
       "      <th>istcgahotspot</th>\n",
       "      <th>iscosmichotspot</th>\n",
       "      <th>codon</th>\n",
       "      <th>is_kras_hotspot</th>\n",
       "      <th>kras_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000981</td>\n",
       "      <td>25368390</td>\n",
       "      <td>25368390</td>\n",
       "      <td>frame_shift_del</td>\n",
       "      <td>del</td>\n",
       "      <td>p.K185fs</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>185</td>\n",
       "      <td>False</td>\n",
       "      <td>frame_shift_del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-001650</td>\n",
       "      <td>25368390</td>\n",
       "      <td>25368390</td>\n",
       "      <td>frame_shift_del</td>\n",
       "      <td>del</td>\n",
       "      <td>p.K185fs</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>185</td>\n",
       "      <td>False</td>\n",
       "      <td>frame_shift_del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-002238</td>\n",
       "      <td>25368390</td>\n",
       "      <td>25368390</td>\n",
       "      <td>frame_shift_del</td>\n",
       "      <td>del</td>\n",
       "      <td>p.K185fs</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>185</td>\n",
       "      <td>False</td>\n",
       "      <td>frame_shift_del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000996</td>\n",
       "      <td>25368455</td>\n",
       "      <td>25368455</td>\n",
       "      <td>nonsense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.R164*</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>164</td>\n",
       "      <td>False</td>\n",
       "      <td>nonsense_mutation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000218</td>\n",
       "      <td>25378561</td>\n",
       "      <td>25378561</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.A146V</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>146</td>\n",
       "      <td>True</td>\n",
       "      <td>A146V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id  start_position  end_position variant_classification  \\\n",
       "0  ACH-000981        25368390      25368390        frame_shift_del   \n",
       "1  ACH-001650        25368390      25368390        frame_shift_del   \n",
       "2  ACH-002238        25368390      25368390        frame_shift_del   \n",
       "3  ACH-000996        25368455      25368455      nonsense_mutation   \n",
       "4  ACH-000218        25378561      25378561      missense_mutation   \n",
       "\n",
       "  variant_type protein_change isdeleterious  istcgahotspot  iscosmichotspot  \\\n",
       "0          del       p.K185fs          True          False            False   \n",
       "1          del       p.K185fs          True          False            False   \n",
       "2          del       p.K185fs          True          False            False   \n",
       "3          snp        p.R164*          True          False            False   \n",
       "4          snp        p.A146V         False           True             True   \n",
       "\n",
       "  codon  is_kras_hotspot      kras_mutation  \n",
       "0   185            False    frame_shift_del  \n",
       "1   185            False    frame_shift_del  \n",
       "2   185            False    frame_shift_del  \n",
       "3   164            False  nonsense_mutation  \n",
       "4   146             True              A146V  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras_mutations_columns = [\n",
    "    \"depmap_id\",\n",
    "    \"start_position\",\n",
    "    \"end_position\",\n",
    "    \"variant_classification\",\n",
    "    \"variant_type\",\n",
    "    \"protein_change\",\n",
    "    \"isdeleterious\",\n",
    "    \"istcgahotspot\",\n",
    "    \"iscosmichotspot\",\n",
    "]\n",
    "\n",
    "kras_hotspot_codons = [\"12\", \"13\", \"61\", \"146\"]\n",
    "\n",
    "\n",
    "kras_mutations = (\n",
    "    ccle_mutations[ccle_mutations.hugo_symbol == \"KRAS\"][kras_mutations_columns]\n",
    "    .assign(\n",
    "        variant_classification=lambda x: x.variant_classification.str.lower(),\n",
    "        variant_type=lambda x: x.variant_type.str.lower(),\n",
    "        codon=lambda x: [re.sub(\"\\D\", \"\", a) for a in x.protein_change],\n",
    "        is_kras_hotspot=lambda x: x.codon.isin(kras_hotspot_codons),\n",
    "    )\n",
    "    .pipe(lambda x: x[x.variant_classification != \"silent\"])\n",
    "    .pipe(\n",
    "        lambda x: x[\n",
    "            x.is_kras_hotspot | x.iscosmichotspot | x.istcgahotspot | x.isdeleterious\n",
    "        ]\n",
    "    )\n",
    "    .assign(\n",
    "        kras_mutation=lambda x: [\n",
    "            pc if vt == \"missense_mutation\" else vt\n",
    "            for vt, pc in zip(x.variant_classification, x.protein_change)\n",
    "        ]\n",
    "    )\n",
    "    .assign(kras_mutation=lambda x: [re.sub(\"p\\\\.\", \"\", a) for a in x.kras_mutation])\n",
    "    .drop_duplicates()\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "\n",
    "kras_mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kras_mutation</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G12D</th>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12V</th>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12C</th>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12A</th>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G13D</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q61H</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12R</th>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12S</th>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A146T</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G13C</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q61K</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>frame_shift_del</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A59T</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K117N</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A146V</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V14I</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q61L</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q61R</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>in_frame_ins</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nonsense_mutation</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V14L</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A146P</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L19F</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12F</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D33E</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A59G</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>splice_site</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   depmap_id\n",
       "kras_mutation               \n",
       "G12D                      68\n",
       "G12V                      47\n",
       "G12C                      27\n",
       "G12A                      16\n",
       "G13D                      14\n",
       "Q61H                       9\n",
       "G12R                       8\n",
       "G12S                       8\n",
       "A146T                      7\n",
       "G13C                       6\n",
       "Q61K                       4\n",
       "frame_shift_del            3\n",
       "A59T                       3\n",
       "K117N                      3\n",
       "A146V                      3\n",
       "V14I                       2\n",
       "Q61L                       2\n",
       "Q61R                       2\n",
       "in_frame_ins               1\n",
       "nonsense_mutation          1\n",
       "V14L                       1\n",
       "A146P                      1\n",
       "L19F                       1\n",
       "G12F                       1\n",
       "D33E                       1\n",
       "A59G                       1\n",
       "splice_site                1"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras_mutation_counts = (\n",
    "    kras_mutations[[\"kras_mutation\", \"depmap_id\"]]\n",
    "    .groupby(\"kras_mutation\")\n",
    "    .count()\n",
    "    .sort_values(\"depmap_id\", ascending=False)\n",
    ")\n",
    "\n",
    "# List of KRAS mutations to group as \"other\".\n",
    "kras_others = (\n",
    "    kras_mutation_counts.reset_index().pipe(lambda x: x[x.depmap_id < 2]).kras_mutation\n",
    ")\n",
    "\n",
    "kras_mutation_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assign rare KRAS mutations to \"other\" group.\n",
    "kras_mutations = kras_mutations.assign(\n",
    "    kras_mutation=lambda x: [\n",
    "        \"other\" if kras_others.isin([a]).any() else a for a in x.kras_mutation\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>kras_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000718</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000249</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000264</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000314</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000344</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACH-001094</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ACH-001378</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ACH-001650</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ACH-001857</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id  kras_mutation\n",
       "0  ACH-000718              3\n",
       "1  ACH-000249              2\n",
       "2  ACH-000264              2\n",
       "3  ACH-000314              2\n",
       "4  ACH-000344              2\n",
       "5  ACH-001001              2\n",
       "6  ACH-001094              2\n",
       "7  ACH-001378              2\n",
       "8  ACH-001650              2\n",
       "9  ACH-001857              2"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mult_kras_mutations = (\n",
    "    kras_mutations[[\"depmap_id\", \"kras_mutation\"]]\n",
    "    .groupby(\"depmap_id\")\n",
    "    .count()\n",
    "    .pipe(lambda x: x[x.kras_mutation > 1])\n",
    "    .sort_values(\"kras_mutation\", ascending=False)\n",
    "    .reset_index(drop=False)\n",
    ")\n",
    "\n",
    "mult_kras_mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>protein_change</th>\n",
       "      <th>isdeleterious</th>\n",
       "      <th>istcgahotspot</th>\n",
       "      <th>iscosmichotspot</th>\n",
       "      <th>codon</th>\n",
       "      <th>is_kras_hotspot</th>\n",
       "      <th>kras_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000249</td>\n",
       "      <td>25380275</td>\n",
       "      <td>25380275</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61H</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000264</td>\n",
       "      <td>25380277</td>\n",
       "      <td>25380277</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000264</td>\n",
       "      <td>25380277</td>\n",
       "      <td>25380278</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>dnp</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000314</td>\n",
       "      <td>25380275</td>\n",
       "      <td>25380275</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61H</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000314</td>\n",
       "      <td>25398282</td>\n",
       "      <td>25398282</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G13C</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>13</td>\n",
       "      <td>True</td>\n",
       "      <td>G13C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACH-000344</td>\n",
       "      <td>25380277</td>\n",
       "      <td>25380277</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACH-000344</td>\n",
       "      <td>25380277</td>\n",
       "      <td>25380278</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>dnp</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ACH-000718</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ACH-000718</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12C</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ACH-000718</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>dnp</td>\n",
       "      <td>p.G12F</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12S</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ACH-001094</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12S</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ACH-001378</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ACH-001378</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12S</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ACH-001650</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12D</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ACH-001857</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12C</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ACH-001857</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     depmap_id  start_position  end_position variant_classification  \\\n",
       "0   ACH-000249        25380275      25380275      missense_mutation   \n",
       "1   ACH-000264        25380277      25380277      missense_mutation   \n",
       "2   ACH-000264        25380277      25380278      missense_mutation   \n",
       "3   ACH-000314        25380275      25380275      missense_mutation   \n",
       "4   ACH-000314        25398282      25398282      missense_mutation   \n",
       "5   ACH-000344        25380277      25380277      missense_mutation   \n",
       "6   ACH-000344        25380277      25380278      missense_mutation   \n",
       "7   ACH-000718        25398284      25398284      missense_mutation   \n",
       "8   ACH-000718        25398285      25398285      missense_mutation   \n",
       "9   ACH-000718        25398284      25398285      missense_mutation   \n",
       "10  ACH-001001        25398285      25398285      missense_mutation   \n",
       "11  ACH-001094        25398285      25398285      missense_mutation   \n",
       "12  ACH-001378        25398284      25398284      missense_mutation   \n",
       "13  ACH-001378        25398285      25398285      missense_mutation   \n",
       "14  ACH-001650        25398284      25398284      missense_mutation   \n",
       "15  ACH-001857        25398285      25398285      missense_mutation   \n",
       "16  ACH-001857        25398284      25398284      missense_mutation   \n",
       "\n",
       "   variant_type protein_change isdeleterious  istcgahotspot  iscosmichotspot  \\\n",
       "0           snp         p.Q61H         False           True             True   \n",
       "1           snp         p.Q61K         False           True             True   \n",
       "2           dnp         p.Q61K         False           True             True   \n",
       "3           snp         p.Q61H         False           True             True   \n",
       "4           snp         p.G13C         False           True             True   \n",
       "5           snp         p.Q61K         False           True             True   \n",
       "6           dnp         p.Q61K         False           True             True   \n",
       "7           snp         p.G12V         False           True             True   \n",
       "8           snp         p.G12C         False           True             True   \n",
       "9           dnp         p.G12F         False           True             True   \n",
       "10          snp         p.G12S         False           True             True   \n",
       "11          snp         p.G12S         False           True             True   \n",
       "12          snp         p.G12V         False           True             True   \n",
       "13          snp         p.G12S         False           True             True   \n",
       "14          snp         p.G12D         False           True             True   \n",
       "15          snp         p.G12C         False           True             True   \n",
       "16          snp         p.G12V         False           True             True   \n",
       "\n",
       "   codon  is_kras_hotspot kras_mutation  \n",
       "0     61             True          Q61H  \n",
       "1     61             True          Q61K  \n",
       "2     61             True          Q61K  \n",
       "3     61             True          Q61H  \n",
       "4     13             True          G13C  \n",
       "5     61             True          Q61K  \n",
       "6     61             True          Q61K  \n",
       "7     12             True          G12V  \n",
       "8     12             True          G12C  \n",
       "9     12             True         other  \n",
       "10    12             True          G12S  \n",
       "11    12             True          G12S  \n",
       "12    12             True          G12V  \n",
       "13    12             True          G12S  \n",
       "14    12             True          G12D  \n",
       "15    12             True          G12C  \n",
       "16    12             True          G12V  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras_mult_mutations_fix = (\n",
    "    kras_mutations[kras_mutations.depmap_id.isin(mult_kras_mutations.depmap_id)]\n",
    "    .reset_index(drop=True)\n",
    "    .pipe(lambda x: x[x.is_kras_hotspot])\n",
    "    .sort_values(\"depmap_id\")\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "kras_mult_mutations_fix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>kras_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000718</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000264</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000314</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000344</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-001378</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACH-001857</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACH-000249</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ACH-001094</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ACH-001650</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id  kras_mutation\n",
       "0  ACH-000718              3\n",
       "1  ACH-000264              2\n",
       "2  ACH-000314              2\n",
       "3  ACH-000344              2\n",
       "4  ACH-001378              2\n",
       "5  ACH-001857              2\n",
       "6  ACH-000249              1\n",
       "7  ACH-001001              1\n",
       "8  ACH-001094              1\n",
       "9  ACH-001650              1"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Cell lines to ignore because they have multiple KRAS mutations that are\n",
    "# not easily resolved.\n",
    "true_kras_double_muts = [\"ACH-000718\", \"ACH-000314\", \"ACH-001378\", \"ACH-001857\"]\n",
    "\n",
    "(\n",
    "    kras_mult_mutations_fix[[\"depmap_id\", \"kras_mutation\"]]\n",
    "    .groupby(\"depmap_id\")\n",
    "    .count()\n",
    "    .sort_values(\"kras_mutation\", ascending=False)\n",
    "    .reset_index(drop=False)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>protein_change</th>\n",
       "      <th>isdeleterious</th>\n",
       "      <th>istcgahotspot</th>\n",
       "      <th>iscosmichotspot</th>\n",
       "      <th>codon</th>\n",
       "      <th>is_kras_hotspot</th>\n",
       "      <th>kras_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000249</td>\n",
       "      <td>25380275</td>\n",
       "      <td>25380275</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61H</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000264</td>\n",
       "      <td>25380277</td>\n",
       "      <td>25380277</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000314</td>\n",
       "      <td>25380275</td>\n",
       "      <td>25380275</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61H</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>double_mut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000344</td>\n",
       "      <td>25380277</td>\n",
       "      <td>25380277</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000718</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>double_mut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12S</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACH-001094</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12S</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ACH-001378</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>double_mut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ACH-001650</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12D</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ACH-001857</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12C</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>double_mut</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id  start_position  end_position variant_classification  \\\n",
       "0  ACH-000249        25380275      25380275      missense_mutation   \n",
       "1  ACH-000264        25380277      25380277      missense_mutation   \n",
       "2  ACH-000314        25380275      25380275      missense_mutation   \n",
       "3  ACH-000344        25380277      25380277      missense_mutation   \n",
       "4  ACH-000718        25398284      25398284      missense_mutation   \n",
       "5  ACH-001001        25398285      25398285      missense_mutation   \n",
       "6  ACH-001094        25398285      25398285      missense_mutation   \n",
       "7  ACH-001378        25398284      25398284      missense_mutation   \n",
       "8  ACH-001650        25398284      25398284      missense_mutation   \n",
       "9  ACH-001857        25398285      25398285      missense_mutation   \n",
       "\n",
       "  variant_type protein_change  isdeleterious  istcgahotspot  iscosmichotspot  \\\n",
       "0          snp         p.Q61H          False           True             True   \n",
       "1          snp         p.Q61K          False           True             True   \n",
       "2          snp         p.Q61H          False           True             True   \n",
       "3          snp         p.Q61K          False           True             True   \n",
       "4          snp         p.G12V          False           True             True   \n",
       "5          snp         p.G12S          False           True             True   \n",
       "6          snp         p.G12S          False           True             True   \n",
       "7          snp         p.G12V          False           True             True   \n",
       "8          snp         p.G12D          False           True             True   \n",
       "9          snp         p.G12C          False           True             True   \n",
       "\n",
       "  codon  is_kras_hotspot kras_mutation  \n",
       "0    61             True          Q61H  \n",
       "1    61             True          Q61K  \n",
       "2    61             True    double_mut  \n",
       "3    61             True          Q61K  \n",
       "4    12             True    double_mut  \n",
       "5    12             True          G12S  \n",
       "6    12             True          G12S  \n",
       "7    12             True    double_mut  \n",
       "8    12             True          G12D  \n",
       "9    12             True    double_mut  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras_mult_mutations_fix = (\n",
    "    kras_mult_mutations_fix.assign(\n",
    "        kras_mutation=lambda x: [\n",
    "            \"double_mut\" if a in true_kras_double_muts else b\n",
    "            for a, b in zip(x.depmap_id, x.kras_mutation)\n",
    "        ]\n",
    "    )\n",
    "    .groupby(\"depmap_id\")\n",
    "    .first()\n",
    "    .reset_index(drop=False)\n",
    ")\n",
    "kras_mult_mutations_fix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "kras_mutations = kras_mutations.pipe(\n",
    "    lambda x: x[~x.depmap_id.isin(kras_mult_mutations_fix.depmap_id)]\n",
    ")\n",
    "\n",
    "kras_mutations = pd.concat([kras_mutations, kras_mult_mutations_fix])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of cell lines with multiple KRAS mutations: 0\n"
     ]
    }
   ],
   "source": [
    "mult_kras_mutations = (\n",
    "    kras_mutations[[\"depmap_id\", \"kras_mutation\"]]\n",
    "    .groupby(\"depmap_id\")\n",
    "    .count()\n",
    "    .pipe(lambda x: x[x.kras_mutation > 1])\n",
    "    .sort_values(\"kras_mutation\", ascending=False)\n",
    "    .reset_index(drop=False)\n",
    ")\n",
    "\n",
    "print(f\"Number of cell lines with multiple KRAS mutations: {len(mult_kras_mutations)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "kras_mutations.to_csv(save_dir / \"kras_mutations.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'CCLE_segment_cn.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>num_probes</th>\n",
       "      <th>segment_mean</th>\n",
       "      <th>source</th>\n",
       "      <th>copy_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1969745</td>\n",
       "      <td>286</td>\n",
       "      <td>2.546065</td>\n",
       "      <td>Sanger WES</td>\n",
       "      <td>5.840389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>1</td>\n",
       "      <td>1969746</td>\n",
       "      <td>6354345</td>\n",
       "      <td>365</td>\n",
       "      <td>2.175759</td>\n",
       "      <td>Sanger WES</td>\n",
       "      <td>4.518235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>1</td>\n",
       "      <td>6354346</td>\n",
       "      <td>6958256</td>\n",
       "      <td>100</td>\n",
       "      <td>3.109430</td>\n",
       "      <td>Sanger WES</td>\n",
       "      <td>8.630414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>1</td>\n",
       "      <td>6958257</td>\n",
       "      <td>15977206</td>\n",
       "      <td>884</td>\n",
       "      <td>2.134831</td>\n",
       "      <td>Sanger WES</td>\n",
       "      <td>4.391857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>1</td>\n",
       "      <td>15977207</td>\n",
       "      <td>16174774</td>\n",
       "      <td>57</td>\n",
       "      <td>2.952592</td>\n",
       "      <td>Sanger WES</td>\n",
       "      <td>7.741386</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id chromosome     start       end  num_probes  segment_mean  \\\n",
       "0  ACH-000001          1         1   1969745         286      2.546065   \n",
       "1  ACH-000001          1   1969746   6354345         365      2.175759   \n",
       "2  ACH-000001          1   6354346   6958256         100      3.109430   \n",
       "3  ACH-000001          1   6958257  15977206         884      2.134831   \n",
       "4  ACH-000001          1  15977207  16174774          57      2.952592   \n",
       "\n",
       "       source  copy_number  \n",
       "0  Sanger WES     5.840389  \n",
       "1  Sanger WES     4.518235  \n",
       "2  Sanger WES     8.630414  \n",
       "3  Sanger WES     4.391857  \n",
       "4  Sanger WES     7.741386  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccle_segment_cn = (\n",
    "    pd.read_csv(data_dir / \"CCLE_segment_cn.csv\")\n",
    "    .clean_names()\n",
    "    .assign(copy_number=lambda x: [2 ** a for a in x.segment_mean])\n",
    ")\n",
    "ccle_segment_cn.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "ccle_segment_cn.to_csv(save_dir / \"ccle_semgent_cn.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'CCLE_gene_cn.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>hugo_symbol</th>\n",
       "      <th>log2_cn_p1</th>\n",
       "      <th>copy_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1.179621</td>\n",
       "      <td>2.253141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000002</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1.009801</td>\n",
       "      <td>1.745055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000003</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1.022828</td>\n",
       "      <td>1.781048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000004</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1.232225</td>\n",
       "      <td>2.428849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000005</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1.151880</td>\n",
       "      <td>2.164136</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id hugo_symbol  log2_cn_p1  copy_number\n",
       "0  ACH-000001        A1BG    1.179621     2.253141\n",
       "1  ACH-000002        A1BG    1.009801     1.745055\n",
       "2  ACH-000003        A1BG    1.022828     1.781048\n",
       "3  ACH-000004        A1BG    1.232225     2.428849\n",
       "4  ACH-000005        A1BG    1.151880     2.164136"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccle_gene_cn = (\n",
    "    pd.read_csv(data_dir / \"CCLE_gene_cn.csv\")\n",
    "    .rename({\"Unnamed: 0\": \"depmap_id\"}, axis=1)\n",
    "    .melt(id_vars=\"depmap_id\", var_name=\"hugo_symbol\", value_name=\"log2_cn_p1\")\n",
    "    .assign(\n",
    "        hugo_symbol=lambda x: [a.split(\" \")[0] for a in x.hugo_symbol],\n",
    "        copy_number=lambda x: np.exp(x.log2_cn_p1) - 1,\n",
    "    )\n",
    ")\n",
    "\n",
    "ccle_gene_cn.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAtQAAAHSCAYAAADMnFxwAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAbgElEQVR4nO3dcZDnd13f8dfbHKACFjBLSEmuR2wERdqg57UmFTCoRMsIOILJWJoq9XQKDjRiCzhTbDvMOFainbbCHCQlthiIhChaqqaEhjooIcEICQENGOTM3eYIbcHq4CR594/7nSxh93Yvn/39vr+9ezxmdva3n+/vd/tmfsnmyfc++/1WdwcAAHhovmLqAQAAYCcT1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMGDX1AOMOP3003vPnj1TjwEAwEnulltu+Ux3r6x3bEcH9Z49e3LzzTdPPQYAACe5qvrURsds+QAAgAGCGgAABghqAAAYIKgBAGCAoAYAgAGCGgAABghqAAAYIKgBAGCAoAYAgAGCGgAABghqAAAYIKgBAGCAoAYAgAGCGgAABghqAAAYIKgBAGCAoAYAgAGCGgAABghqAAAYsGvqAYDtd/4zL8zh1SMbHn/CGSt5/403LHAiADh5CWo4CR1ePZJ9lx3Y8PhNl+9f4DQAcHKb25aPqjq7qt5bVXdU1e1V9fLZ+uOq6vqq+uPZ58euec2rq+rOqvp4VT1nXrMBAMB2mece6vuS/GR3f0OSv5/kpVX1jUleleQ93X1ukvfMvs7s2MVJnprkoiS/VFWnzXE+AAAYNreg7u5D3f2h2ePPJ7kjyROTPC/JVbOnXZXk+bPHz0vytu7+Qnf/SZI7k+yb13wAALAdFnKVj6rak+TpST6Q5IzuPpQcje4kj5897YlJPr3mZQdnawAAsLTmHtRV9agk1yZ5RXd/7nhPXWet1/nz9lfVzVV185EjG1/FAAAAFmGuQV1VD8vRmH5rd79ztrxaVWfOjp+Z5J7Z+sEkZ695+VlJ7n7wn9ndB7p7b3fvXVlZmd/wAACwBfO8ykcluSLJHd19+ZpD70py6ezxpUl+fc36xVX1iKp6UpJzk9w0r/kAAGA7zPM61BckeXGSj1TVrbO11yT52STXVNVLkvxpkhcmSXffXlXXJPlojl4h5KXdff8c5wMAgGFzC+ru/t2svy86SZ69wWtel+R185oJAAC220Ku8gEAACcrQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAPmFtRVdWVV3VNVt61Ze3tV3Tr7uKuqbp2t76mqv1xz7I3zmgsAALbTrjn+2W9J8h+T/PKxhe7+wWOPq+r1Sf7vmud/orvPm+M8AACw7eYW1N39vqras96xqqokL0py4by+PwAALMJUe6i/Pclqd//xmrUnVdUfVNWNVfXtE80FAAAnZJ5bPo7nkiRXr/n6UJLd3X1vVX1Lkl+rqqd29+ce/MKq2p9kf5Ls3r17IcMCAMBGFn6Guqp2Jfn+JG8/ttbdX+jue2ePb0nyiSRfv97ru/tAd+/t7r0rKyuLGBkAADY0xZaP70zyse4+eGyhqlaq6rTZ43OSnJvkkxPMBgAAJ2Sel827OsnvJXlyVR2sqpfMDl2cL93ukSTPSPLhqvrDJO9I8uPd/dl5zQYAANtlnlf5uGSD9X+yztq1Sa6d1ywAADAv7pQIAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAyYW1BX1ZVVdU9V3bZm7Weq6s+q6tbZx/euOfbqqrqzqj5eVc+Z11wAALCd5nmG+i1JLlpn/Re6+7zZx7uTpKq+McnFSZ46e80vVdVpc5wNAAC2xdyCurvfl+SzW3z685K8rbu/0N1/kuTOJPvmNRsAAGyXKfZQv6yqPjzbEvLY2doTk3x6zXMOztYAAGCpLTqo35Dk65Kcl+RQktfP1mud5/Z6f0BV7a+qm6vq5iNHjsxlSAAA2KqFBnV3r3b3/d39QJI35YvbOg4mOXvNU89KcvcGf8aB7t7b3XtXVlbmOzAAAGxioUFdVWeu+fIFSY5dAeRdSS6uqkdU1ZOSnJvkpkXOBgAAD8Wuef3BVXV1kmclOb2qDiZ5bZJnVdV5Obqd464kP5Yk3X17VV2T5KNJ7kvy0u6+f16zAQDAdplbUHf3JessX3Gc578uyevmNQ8AAMyDOyUCAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAPmFtRVdWVV3VNVt61Z+3dV9bGq+nBVXVdVj5mt76mqv6yqW2cfb5zXXAAAsJ3meYb6LUkuetDa9Um+qbv/TpI/SvLqNcc+0d3nzT5+fI5zAQDAtplbUHf3+5J89kFrv9Pd982+/P0kZ83r+wMAwCJMuYf6R5L89zVfP6mq/qCqbqyqb59qKAAAOBG7pvimVfXTSe5L8tbZ0qEku7v73qr6liS/VlVP7e7PrfPa/Un2J8nu3bsXNTIAAKxr4Weoq+rSJM9N8kPd3UnS3V/o7ntnj29J8okkX7/e67v7QHfv7e69KysrixobAADWtdCgrqqLkvzLJN/X3X+xZn2lqk6bPT4nyblJPrnI2QAA4KGY25aPqro6ybOSnF5VB5O8Nkev6vGIJNdXVZL8/uyKHs9I8m+q6r4k9yf58e7+7Lp/MAAALJG5BXV3X7LO8hUbPPfaJNfOaxYAAJgXd0oEAIABk1zlA5jWoUN355ynPG3D4084YyXvv/GGBU4EADuXoIZT0AOd7LvswIbHb7p8/wKnAYCdzZYPAAAYIKgBAGCAoAYAgAGCGgAABghqAAAYIKgBAGCAoAYAgAGCGgAABghqAAAYIKgBAGCAoAYAgAGCGgAABghqAAAYIKgBAGCAoAYAgAGCGgAABmwpqKvqgq2sAQDAqWarZ6j/wxbXAADglLLreAer6tuSnJ9kpaouW3Poa5KcNs/BAABgJzhuUCd5eJJHzZ736DXrn0vyA/MaCgAAdorjBnV335jkxqp6S3d/akEzAQDAjrHZGepjHlFVB5LsWfua7r5wHkMBAMBOsdWg/tUkb0zy5iT3z28cAADYWbYa1Pd19xvmOgkAAOxAW71s3m9U1T+rqjOr6nHHPuY6GQAA7ABbPUN96ezzT61Z6yTnbO84AACws2wpqLv7SfMeBAAAdqItBXVV/eP11rv7l7d3HAAA2Fm2uuXjW9c8/sokz07yoSSCGgCAU9pWt3z8xNqvq+pvJPkvc5kIAAB2kK1e5ePB/iLJuds5CAAA7ERb3UP9Gzl6VY8kOS3JNyS5Zl5DAQDATrHVPdQ/v+bxfUk+1d0H5zAPAADsKFva8tHdNyb5WJJHJ3lskr+a51AAALBTbCmoq+pFSW5K8sIkL0rygar6gXkOBgAAO8FWt3z8dJJv7e57kqSqVpL8jyTvmNdgAACwE2z1Kh9fcSymZ+49gdcCAMBJa6tnqH+rqn47ydWzr38wybvnMxIAAOwcxw3qqvrbSc7o7p+qqu9P8g+SVJLfS/LWBcwHAABLbbNtG7+Y5PNJ0t3v7O7Luvuf5+jZ6V883gur6sqquqeqbluz9riqur6q/nj2+bFrjr26qu6sqo9X1XMe6v8gAABYpM2Cek93f/jBi919c5I9m7z2LUkuetDaq5K8p7vPTfKe2depqm9McnGSp85e80tVddpmwwMAwNQ2C+qvPM6xrzreC7v7fUk++6Dl5yW5avb4qiTPX7P+tu7+Qnf/SZI7k+zbZDYAAJjcZkH9war60QcvVtVLktzyEL7fGd19KElmnx8/W39ikk+ved7B2RoAACy1za7y8Yok11XVD+WLAb03ycOTvGAb56h11nrdJ1btT7I/SXbv3r2NIwAAwIk7blB392qS86vqO5J802z5v3X3DQ/x+61W1Zndfaiqzkxy7NrWB5OcveZ5ZyW5e4OZDiQ5kCR79+5dN7oBAGBRtnQd6u5+b5L3bsP3e1eSS5P87Ozzr69Z/5WqujzJ30xybo7e6hwAAJbaVm/scsKq6uokz0pyelUdTPLaHA3pa2Z7sP80yQuTpLtvr6prknw0yX1JXtrd989rNgAA2C5zC+ruvmSDQ8/e4PmvS/K6ec0DAADzsNlVPgAAgOMQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMGDXor9hVT05ydvXLJ2T5F8leUySH01yZLb+mu5+92KnAwCAE7PwoO7ujyc5L0mq6rQkf5bkuiQ/nOQXuvvnFz0TAAA8VFNv+Xh2kk9096cmngMAAB6SqYP64iRXr/n6ZVX14aq6sqoeO9VQAACwVZMFdVU9PMn3JfnV2dIbknxdjm4HOZTk9Ru8bn9V3VxVNx85cmS9pwAAwMJMeYb6e5J8qLtXk6S7V7v7/u5+IMmbkuxb70XdfaC793b33pWVlQWOCwAAX27KoL4ka7Z7VNWZa469IMltC58IAABO0MKv8pEkVfXVSb4ryY+tWf65qjovSSe560HHAABgKU0S1N39F0m+9kFrL55iFgAAGDH1VT4AAGBHE9QAADBAUAMAwABBDQAAAwQ1AAAMENQAADBAUAMAwABBDQAAAwQ1AAAMENQAADBAUAMAwIBdUw8AnLjzn3lhDq8e2fD44dXVBU4DAKc2QQ070OHVI9l32YENj1/3yucucBoAOLXZ8gEAAAMENQAADBDUAAAwQFADAMAAv5QIfJlDh+7OOU952obHn3DGSt5/4w0LnAgAlpegBr7MA53jXkXkpsv3L3AaAFhutnwAAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADBDUAAAwQFADAMAAQQ0AAAMENQAADNg1xTetqruSfD7J/Unu6+69VfW4JG9PsifJXUle1N3/e4r5AABgq6Y8Q/0d3X1ed++dff2qJO/p7nOTvGf2NQAALLVl2vLxvCRXzR5fleT5040CAABbM1VQd5Lfqapbqmr/bO2M7j6UJLPPj59oNgAA2LJJ9lAnuaC7766qxye5vqo+ttUXzgJ8f5Ls3r17XvMBAMCWTHKGurvvnn2+J8l1SfYlWa2qM5Nk9vmeDV57oLv3dvfelZWVRY0MAADrWnhQV9Ujq+rRxx4n+e4ktyV5V5JLZ0+7NMmvL3o2AAA4UVNs+TgjyXVVdez7/0p3/1ZVfTDJNVX1kiR/muSFE8wGAAAnZOFB3d2fTPJ311m/N8mzFz0PTOH8Z16Yw6tHNjz+hDNW8v4bb1jgRADAQzXVLyXCKe3w6pHsu+zAhsdvunz/hscAgOWyTNehBgCAHUdQAwDAAFs+YAkdOnR3znnK0zY8fnh1dYHTAADHI6hhCT3QOe4e6+te+dwFTgMAHI8tHwAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAgF1TDwCces5/5oU5vHpkw+NPOGMl77/xhgVOBAAPnaAGFu7w6pHsu+zAhsdvunz/AqcBgDG2fAAAwABBDQAAA2z5AE7YoUN355ynPG3D4/ZAA3AqEdTACXugYw80AMzY8gEAAAMENQAADLDlA9h2m+2xPry6usBpAGC+BDWw7TbbY33dK5+7wGkAYL5s+QAAgAGCGgAABghqAAAYIKgBAGCAoAYAgAGCGgAABghqAAAY4DrUMAfnP/PCHF49suFxNzYBgJOHoIY5OLx6xI1NAOAUYcsHAAAMENQAADBAUAMAwABBDQAAAxYe1FV1dlW9t6ruqKrbq+rls/Wfqao/q6pbZx/fu+jZAADgRE1xlY/7kvxkd3+oqh6d5Jaqun527Be6++cnmAkAAB6ShQd1dx9Kcmj2+PNVdUeSJy56DgAA2A6T7qGuqj1Jnp7kA7Oll1XVh6vqyqp67HSTAQDA1kx2Y5eqelSSa5O8ors/V1VvSPJvk/Ts8+uT/Mg6r9ufZH+S7N69e3EDwxruhAgAHDNJUFfVw3I0pt/a3e9Mku5eXXP8TUl+c73XdveBJAeSZO/evT3/aeHLuRMiAHDMFFf5qCRXJLmjuy9fs37mmqe9IMlti54NAABO1BRnqC9I8uIkH6mqW2drr0lySVWdl6NbPu5K8mMTzAYAACdkiqt8/G6SWufQuxc9C2zEHmkAYKsm+6VEWGb2SAMAWyWogaVz6NDdOecpT9vw+BPOWMn7b7xhgRMBwMYENbB0Hugc928Ibrp8/wKnAYDjm/TGLgAAsNMJagAAGGDLByelza7Sce9njuRrT1/Z8LireAAAWyWoOSlt5SodruIBAGwHWz4AAGCAoAYAgAGCGgAABghqAAAY4JcSgR1nszspbnYVF3daBGA7CWpgx9nsToqbXcXFnRYB2E62fAAAwABBDQAAA2z5AE45m+3BHt1jvdmdOu3hBji5CGrglLPZHuzRPdab3anTHm6Ak4ugZik5w8cy2+yfz8OrqwucBoCpCWqWkjN8LLPN/vm87pXPXeA0AExNUDMJZ/gAgJOFoGYSzvABACcLQc2OtNlVGpzhBgAWRVCzI23lTnkAAIvgxi4AADBAUAMAwABbPgCWjOuwA+wsghpgybgOO8DOYssHAAAMENQAADDAlg+ABRu9jvpmr7fHGmCxBDXAgo1eR32z19tjDbBYtnwAAMAAZ6gBHsSt7QE4EYIa4EHc2h6AEyGomYvNbkzhDB8AcLIQ1MzFZjemcIYPADhZ+KVEAAAYIKgBAGCALR+sa7M90Pd+5ki+9vSVDY/bIw2nrtGfH25MA6euzX5+LOvPB0HNurayB9oeaWA9oz8/3JgGTl2b/fxY1p8PghrgJDPvW5O7ig/AlxLUACeZed+a3FV8AL7U0gV1VV2U5N8nOS3Jm7v7ZyceaUfaqXuQgPmb9xlsxvj5DTvPUgV1VZ2W5D8l+a4kB5N8sKre1d0fnXaynWen7kEC5m/eZ7AZ4+c37DxLFdRJ9iW5s7s/mSRV9bYkz0uydEG903+LfbMzVPZAAlPZ7OfTZj9f5338VN+DPnoGfeoz8Kf69x+10+efl2UL6icm+fSarw8m+XsTzXJcO/232Dc7Q2UPJDCVrfx8mvL4qb4HffQM+tRn4E/17z9qp88/L9XdU8/w16rqhUme093/dPb1i5Ps6+6fWPOc/UmOvVtPTvLxhQ+anJ7kMxN8X47P+7J8vCfLyfuyfLwny8n7snymfE/+Vnev+9dXy3aG+mCSs9d8fVaSu9c+obsPJNn4/xotQFXd3N17p5yBL+d9WT7ek+XkfVk+3pPl5H1ZPsv6nizbrcc/mOTcqnpSVT08ycVJ3jXxTAAAsKGlOkPd3fdV1cuS/HaOXjbvyu6+feKxAABgQ0sV1EnS3e9O8u6p59jEpFtO2JD3Zfl4T5aT92X5eE+Wk/dl+Szle7JUv5QIAAA7zbLtoQYAgB1FUJ+gqrqoqj5eVXdW1aumnoekqq6sqnuq6rapZ+Goqjq7qt5bVXdU1e1V9fKpZzrVVdVXVtVNVfWHs/fkX089E19UVadV1R9U1W9OPQtHVdVdVfWRqrq1qm6eeh6SqnpMVb2jqj42++/Lt0090zG2fJyA2a3R/yhrbo2e5BK3Rp9WVT0jyZ8n+eXu/qap5yGpqjOTnNndH6qqRye5Jcnz/bsynaqqJI/s7j+vqocl+d0kL+/u3594NJJU1WVJ9ib5mu5e7ju7nCKq6q4ke7vbdaiXRFVdleR/dfebZ1eD++ru/j8Tj5XEGeoT9de3Ru/uv0py7NboTKi735fks1PPwRd196Hu/tDs8eeT3JGjd0JlIn3Un8++fNjswxmVJVBVZyX5h0nePPUssKyq6muSPCPJFUnS3X+1LDGdCOoTtd6t0UUCHEdV7Uny9CQfmHiUU95sW8GtSe5Jcn13e0+Wwy8m+RdJHph4Dr5UJ/mdqrpldpdmpnVOkiNJ/vNse9Sbq+qRUw91jKA+MbXOmjM8sIGqelSSa5O8ors/N/U8p7ruvr+7z8vRu9DuqypbpCZWVc9Nck933zL1LHyZC7r7m5N8T5KXzrYXMp1dSb45yRu6++lJ/l+SpfldNkF9Yja9NTpw1Gyf7rVJ3trd75x6Hr5o9tek/zPJRdNOQpILknzfbL/u25JcWFX/ddqRSJLuvnv2+Z4k1+Xotk+mczDJwTV/s/aOHA3spSCoT4xbo8MWzH4B7ookd3T35VPPQ1JVK1X1mNnjr0rynUk+NulQpLtf3d1ndfeeHP1vyg3d/Y8mHuuUV1WPnP1CdWbbCr47iStJTai7Dyf5dFU9ebb07CRL84vuS3enxGXm1ujLqaquTvKsJKdX1cEkr+3uK6ad6pR3QZIXJ/nIbM9ukrxmdidUpnFmkqtmVyv6iiTXdLdLtMH6zkhy3dFzA9mV5Fe6+7emHYkkP5HkrbOTmp9M8sMTz/PXXDYPAAAG2PIBAAADBDUAAAwQ1AAAMEBQAwDAAEENAAADBDUAAAwQ1AAAMEBQAwDAgP8P/mb4GjS/9fAAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 864x576 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig = plt.figure(figsize=(12, 8))\n",
    "x = [np.min([a, 6]) for a in ccle_gene_cn.copy_number.sample(n=1000)]\n",
    "sns.histplot(x, kde=False)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "ccle_gene_cn.to_csv(save_dir / \"ccle_gene_cn.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'CCLE_fusions.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>fusion_name</th>\n",
       "      <th>junction_read_count</th>\n",
       "      <th>spanning_frag_count</th>\n",
       "      <th>splice_type</th>\n",
       "      <th>left_gene</th>\n",
       "      <th>left_break_point</th>\n",
       "      <th>right_gene</th>\n",
       "      <th>right_break_point</th>\n",
       "      <th>large_anchor_support</th>\n",
       "      <th>ffpm</th>\n",
       "      <th>left_break_dinuc</th>\n",
       "      <th>left_break_entropy</th>\n",
       "      <th>right_break_dinuc</th>\n",
       "      <th>right_break_entropy</th>\n",
       "      <th>annots</th>\n",
       "      <th>ccle_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-001113</td>\n",
       "      <td>CACNA2D1--SEMA3E</td>\n",
       "      <td>48</td>\n",
       "      <td>53</td>\n",
       "      <td>ONLY_REF_SPLICE</td>\n",
       "      <td>CACNA2D1 (ENSG00000153956)</td>\n",
       "      <td>chr7:82335135:-</td>\n",
       "      <td>SEMA3E (ENSG00000170381)</td>\n",
       "      <td>chr7:83469302:-</td>\n",
       "      <td>YES_LDAS</td>\n",
       "      <td>1.5703</td>\n",
       "      <td>GT</td>\n",
       "      <td>1.9899</td>\n",
       "      <td>AG</td>\n",
       "      <td>1.9899</td>\n",
       "      <td>[\"CCLE_StarF2019\",\"INTRACHROMOSOMAL[chr7:0.92M...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-001113</td>\n",
       "      <td>ADNP2--PLIN3</td>\n",
       "      <td>50</td>\n",
       "      <td>40</td>\n",
       "      <td>INCL_NON_REF_SPLICE</td>\n",
       "      <td>ADNP2 (ENSG00000101544)</td>\n",
       "      <td>chr18:80117650:+</td>\n",
       "      <td>PLIN3 (ENSG00000105355)</td>\n",
       "      <td>chr19:4867657:-</td>\n",
       "      <td>YES_LDAS</td>\n",
       "      <td>1.3993</td>\n",
       "      <td>GT</td>\n",
       "      <td>1.7819</td>\n",
       "      <td>AG</td>\n",
       "      <td>1.8323</td>\n",
       "      <td>[\"CCLE_StarF2019\",\"INTERCHROMOSOMAL[chr18--chr...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-001113</td>\n",
       "      <td>XRCC4--SUGCT</td>\n",
       "      <td>23</td>\n",
       "      <td>104</td>\n",
       "      <td>ONLY_REF_SPLICE</td>\n",
       "      <td>XRCC4 (ENSG00000152422)</td>\n",
       "      <td>chr5:83111203:+</td>\n",
       "      <td>SUGCT (ENSG00000175600)</td>\n",
       "      <td>chr7:40749434:+</td>\n",
       "      <td>YES_LDAS</td>\n",
       "      <td>1.9745</td>\n",
       "      <td>GT</td>\n",
       "      <td>1.9656</td>\n",
       "      <td>AG</td>\n",
       "      <td>1.9656</td>\n",
       "      <td>[\"CCLE_StarF2019\",\"INTERCHROMOSOMAL[chr5--chr7]\"]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-001113</td>\n",
       "      <td>PCYT1A--RBFOX1</td>\n",
       "      <td>26</td>\n",
       "      <td>48</td>\n",
       "      <td>ONLY_REF_SPLICE</td>\n",
       "      <td>PCYT1A (ENSG00000161217)</td>\n",
       "      <td>chr3:196247367:-</td>\n",
       "      <td>RBFOX1 (ENSG00000078328)</td>\n",
       "      <td>chr16:6654603:+</td>\n",
       "      <td>YES_LDAS</td>\n",
       "      <td>1.1505</td>\n",
       "      <td>GT</td>\n",
       "      <td>1.7819</td>\n",
       "      <td>AG</td>\n",
       "      <td>1.8295</td>\n",
       "      <td>[\"CCLE_StarF2019\",\"INTERCHROMOSOMAL[chr3--chr1...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-001113</td>\n",
       "      <td>AF117829.1--PIP4P2</td>\n",
       "      <td>25</td>\n",
       "      <td>31</td>\n",
       "      <td>ONLY_REF_SPLICE</td>\n",
       "      <td>AF117829.1 (ENSG00000251136)</td>\n",
       "      <td>chr8:89757045:-</td>\n",
       "      <td>PIP4P2 (ENSG00000155099)</td>\n",
       "      <td>chr8:90996744:-</td>\n",
       "      <td>YES_LDAS</td>\n",
       "      <td>0.8707</td>\n",
       "      <td>GT</td>\n",
       "      <td>1.8062</td>\n",
       "      <td>AG</td>\n",
       "      <td>1.9329</td>\n",
       "      <td>[\"INTRACHROMOSOMAL[chr8:1.24Mb]\"]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id         fusion_name  junction_read_count  spanning_frag_count  \\\n",
       "0  ACH-001113    CACNA2D1--SEMA3E                   48                   53   \n",
       "1  ACH-001113        ADNP2--PLIN3                   50                   40   \n",
       "2  ACH-001113        XRCC4--SUGCT                   23                  104   \n",
       "3  ACH-001113      PCYT1A--RBFOX1                   26                   48   \n",
       "4  ACH-001113  AF117829.1--PIP4P2                   25                   31   \n",
       "\n",
       "           splice_type                     left_gene  left_break_point  \\\n",
       "0      ONLY_REF_SPLICE    CACNA2D1 (ENSG00000153956)   chr7:82335135:-   \n",
       "1  INCL_NON_REF_SPLICE       ADNP2 (ENSG00000101544)  chr18:80117650:+   \n",
       "2      ONLY_REF_SPLICE       XRCC4 (ENSG00000152422)   chr5:83111203:+   \n",
       "3      ONLY_REF_SPLICE      PCYT1A (ENSG00000161217)  chr3:196247367:-   \n",
       "4      ONLY_REF_SPLICE  AF117829.1 (ENSG00000251136)   chr8:89757045:-   \n",
       "\n",
       "                 right_gene right_break_point large_anchor_support    ffpm  \\\n",
       "0  SEMA3E (ENSG00000170381)   chr7:83469302:-             YES_LDAS  1.5703   \n",
       "1   PLIN3 (ENSG00000105355)   chr19:4867657:-             YES_LDAS  1.3993   \n",
       "2   SUGCT (ENSG00000175600)   chr7:40749434:+             YES_LDAS  1.9745   \n",
       "3  RBFOX1 (ENSG00000078328)   chr16:6654603:+             YES_LDAS  1.1505   \n",
       "4  PIP4P2 (ENSG00000155099)   chr8:90996744:-             YES_LDAS  0.8707   \n",
       "\n",
       "  left_break_dinuc  left_break_entropy right_break_dinuc  right_break_entropy  \\\n",
       "0               GT              1.9899                AG               1.9899   \n",
       "1               GT              1.7819                AG               1.8323   \n",
       "2               GT              1.9656                AG               1.9656   \n",
       "3               GT              1.7819                AG               1.8295   \n",
       "4               GT              1.8062                AG               1.9329   \n",
       "\n",
       "                                              annots  ccle_count  \n",
       "0  [\"CCLE_StarF2019\",\"INTRACHROMOSOMAL[chr7:0.92M...           2  \n",
       "1  [\"CCLE_StarF2019\",\"INTERCHROMOSOMAL[chr18--chr...           2  \n",
       "2  [\"CCLE_StarF2019\",\"INTERCHROMOSOMAL[chr5--chr7]\"]           2  \n",
       "3  [\"CCLE_StarF2019\",\"INTERCHROMOSOMAL[chr3--chr1...           4  \n",
       "4                  [\"INTRACHROMOSOMAL[chr8:1.24Mb]\"]           2  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccle_fusions = (\n",
    "    pd.read_csv(data_dir / \"CCLE_fusions.csv\", delimiter=\"\\t\")\n",
    "    .clean_names()\n",
    "    .rename(\n",
    "        {\n",
    "            \"#fusionname\": \"fusion_name\",\n",
    "            \"junctionreadcount\": \"junction_read_count\",\n",
    "            \"spanningfragcount\": \"spanning_frag_count\",\n",
    "            \"splicetype\": \"splice_type\",\n",
    "            \"leftgene\": \"left_gene\",\n",
    "            \"leftbreakpoint\": \"left_break_point\",\n",
    "            \"rightgene\": \"right_gene\",\n",
    "            \"rightbreakpoint\": \"right_break_point\",\n",
    "            \"largeanchorsupport\": \"large_anchor_support\",\n",
    "            \"leftbreakdinuc\": \"left_break_dinuc\",\n",
    "            \"leftbreakentropy\": \"left_break_entropy\",\n",
    "            \"rightbreakdinuc\": \"right_break_dinuc\",\n",
    "            \"rightbreakentropy\": \"right_break_entropy\",\n",
    "        },\n",
    "        axis=1,\n",
    "    )\n",
    ")\n",
    "ccle_fusions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "ccle_fusions.to_csv(save_dir / \"ccle_fusions.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'CCLE_expression_full_v2.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>hugo_symbol</th>\n",
       "      <th>rna_expr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-001097</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-001636</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-001804</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>4.934988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000534</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>0.839960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-001498</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>1.744161</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id hugo_symbol  rna_expr\n",
       "0  ACH-001097      TSPAN6  0.000000\n",
       "1  ACH-001636      TSPAN6  0.000000\n",
       "2  ACH-001804      TSPAN6  4.934988\n",
       "3  ACH-000534      TSPAN6  0.839960\n",
       "4  ACH-001498      TSPAN6  1.744161"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccle_expression = (\n",
    "    pd.read_csv(data_dir / \"CCLE_expression_full.csv\")\n",
    "    .rename({\"Unnamed: 0\": \"depmap_id\"}, axis=1)\n",
    "    .melt(id_vars=\"depmap_id\", var_name=\"hugo_symbol\", value_name=\"rna_expr\")\n",
    "    .assign(hugo_symbol=lambda x: [a.split(\" \")[0] for a in x.hugo_symbol])\n",
    ")\n",
    "\n",
    "ccle_expression.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "ccle_expression.to_csv(save_dir / \"ccle_expression.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'Achilles_gene_effect.csv' & 'Achilles_gene_effect_unscaled.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>hugo_symbol</th>\n",
       "      <th>gene_effect</th>\n",
       "      <th>gene_effect_unscaled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000004</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.181037</td>\n",
       "      <td>0.542787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000005</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>-0.090250</td>\n",
       "      <td>0.440018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000007</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.071568</td>\n",
       "      <td>0.104889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000009</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.110311</td>\n",
       "      <td>0.478258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000011</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.275160</td>\n",
       "      <td>0.589944</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id hugo_symbol  gene_effect  gene_effect_unscaled\n",
       "0  ACH-000004        A1BG     0.181037              0.542787\n",
       "1  ACH-000005        A1BG    -0.090250              0.440018\n",
       "2  ACH-000007        A1BG     0.071568              0.104889\n",
       "3  ACH-000009        A1BG     0.110311              0.478258\n",
       "4  ACH-000011        A1BG     0.275160              0.589944"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def prep_gene_effect_data(fpath, col_name):\n",
    "    d = (\n",
    "        pd.read_csv(fpath)\n",
    "        .rename({\"DepMap_ID\": \"depmap_id\"}, axis=1)\n",
    "        .melt(id_vars=\"depmap_id\", var_name=\"hugo_symbol\", value_name=col_name)\n",
    "        .assign(hugo_symbol=lambda x: [a.split(\" \")[0] for a in x.hugo_symbol])\n",
    "    )\n",
    "    return d\n",
    "\n",
    "\n",
    "gene_effect_scaled = prep_gene_effect_data(\n",
    "    data_dir / \"Achilles_gene_effect.csv\", \"gene_effect\"\n",
    ")\n",
    "gene_effect_unscaled = prep_gene_effect_data(\n",
    "    data_dir / \"Achilles_gene_effect_unscaled.csv\", \"gene_effect_unscaled\"\n",
    ")\n",
    "\n",
    "# Check that the two data frames are the same shape.\n",
    "if not gene_effect_scaled.shape == gene_effect_unscaled.shape:\n",
    "    raise Exception(\"Gene effect data frames are not the same shape as expected.\")\n",
    "\n",
    "gene_effect = pd.merge(\n",
    "    gene_effect_scaled, gene_effect_unscaled, on=[\"depmap_id\", \"hugo_symbol\"]\n",
    ")\n",
    "\n",
    "gene_effect.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_effect.to_csv(save_dir / \"achilles_gene_effect.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
