{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare DepMap data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "import janitor\n",
    "from pathlib import Path\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = Path('../data')\n",
    "save_dir = Path('../modeling_data')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'sample_info.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>stripped_cell_line_name</th>\n",
       "      <th>ccle_name</th>\n",
       "      <th>sex</th>\n",
       "      <th>cas9_activity</th>\n",
       "      <th>primary_or_metastasis</th>\n",
       "      <th>primary_disease</th>\n",
       "      <th>subtype</th>\n",
       "      <th>lineage</th>\n",
       "      <th>lineage_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>NIHOVCAR3</td>\n",
       "      <td>NIHOVCAR3_OVARY</td>\n",
       "      <td>Female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>Ovarian Cancer</td>\n",
       "      <td>Adenocarcinoma, high grade serous</td>\n",
       "      <td>ovary</td>\n",
       "      <td>ovary_adenocarcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000002</td>\n",
       "      <td>HL60</td>\n",
       "      <td>HL60_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>Female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>Acute Myelogenous Leukemia (AML), M3 (Promyelo...</td>\n",
       "      <td>blood</td>\n",
       "      <td>AML</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000003</td>\n",
       "      <td>CACO2</td>\n",
       "      <td>CACO2_LARGE_INTESTINE</td>\n",
       "      <td>Male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Colon/Colorectal Cancer</td>\n",
       "      <td>Adenocarcinoma</td>\n",
       "      <td>colorectal</td>\n",
       "      <td>colorectal_adenocarcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000004</td>\n",
       "      <td>HEL</td>\n",
       "      <td>HEL_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>Male</td>\n",
       "      <td>47.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>Acute Myelogenous Leukemia (AML), M6 (Erythrol...</td>\n",
       "      <td>blood</td>\n",
       "      <td>AML</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000005</td>\n",
       "      <td>HEL9217</td>\n",
       "      <td>HEL9217_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>Male</td>\n",
       "      <td>13.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>Acute Myelogenous Leukemia (AML), M6 (Erythrol...</td>\n",
       "      <td>blood</td>\n",
       "      <td>AML</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id stripped_cell_line_name  \\\n",
       "0  ACH-000001               NIHOVCAR3   \n",
       "1  ACH-000002                    HL60   \n",
       "2  ACH-000003                   CACO2   \n",
       "3  ACH-000004                     HEL   \n",
       "4  ACH-000005                 HEL9217   \n",
       "\n",
       "                                    ccle_name     sex cas9_activity  \\\n",
       "0                             NIHOVCAR3_OVARY  Female           NaN   \n",
       "1     HL60_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE  Female           NaN   \n",
       "2                       CACO2_LARGE_INTESTINE    Male           NaN   \n",
       "3      HEL_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE    Male          47.6   \n",
       "4  HEL9217_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE    Male          13.4   \n",
       "\n",
       "  primary_or_metastasis          primary_disease  \\\n",
       "0            Metastasis           Ovarian Cancer   \n",
       "1               Primary                 Leukemia   \n",
       "2                   NaN  Colon/Colorectal Cancer   \n",
       "3                   NaN                 Leukemia   \n",
       "4                   NaN                 Leukemia   \n",
       "\n",
       "                                             subtype     lineage  \\\n",
       "0                  Adenocarcinoma, high grade serous       ovary   \n",
       "1  Acute Myelogenous Leukemia (AML), M3 (Promyelo...       blood   \n",
       "2                                     Adenocarcinoma  colorectal   \n",
       "3  Acute Myelogenous Leukemia (AML), M6 (Erythrol...       blood   \n",
       "4  Acute Myelogenous Leukemia (AML), M6 (Erythrol...       blood   \n",
       "\n",
       "             lineage_subtype  \n",
       "0       ovary_adenocarcinoma  \n",
       "1                        AML  \n",
       "2  colorectal_adenocarcinoma  \n",
       "3                        AML  \n",
       "4                        AML  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_info_columns = [\n",
    "    'depmap_id', 'stripped_cell_line_name', 'ccle_name', 'sex', \n",
    "    'cas9_activity', 'primary_or_metastasis', 'primary_disease',\n",
    "    'subtype', 'lineage', 'lineage_subtype',\n",
    "]\n",
    "\n",
    "sample_info = pd.read_csv(data_dir / 'sample_info.csv') \\\n",
    "    .clean_names() \\\n",
    "    [sample_info_columns]\n",
    "sample_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_info.to_csv(save_dir / 'sample_info.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'Achilles_guide_map.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgrna</th>\n",
       "      <th>genome_alignment</th>\n",
       "      <th>n_alignments</th>\n",
       "      <th>hugo_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAAATCCAGCAATGCAG</td>\n",
       "      <td>chr10_110964620_+</td>\n",
       "      <td>1</td>\n",
       "      <td>SHOC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAAACCCGTAGATAGCCT</td>\n",
       "      <td>chr12_95003615_+</td>\n",
       "      <td>1</td>\n",
       "      <td>NDUFA12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAAAAGAAGAAAAAACCAG</td>\n",
       "      <td>chr4_75970356_-</td>\n",
       "      <td>1</td>\n",
       "      <td>SDAD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAAAAGCTCAAGAAGGAGG</td>\n",
       "      <td>chr2_33588446_-</td>\n",
       "      <td>1</td>\n",
       "      <td>FAM98A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAAAAGGCTGTAAAAGCGT</td>\n",
       "      <td>chr19_19891600_+</td>\n",
       "      <td>1</td>\n",
       "      <td>ZNF253</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sgrna   genome_alignment  n_alignments hugo_symbol\n",
       "0  AAAAAAATCCAGCAATGCAG  chr10_110964620_+             1       SHOC2\n",
       "1  AAAAAACCCGTAGATAGCCT   chr12_95003615_+             1     NDUFA12\n",
       "2  AAAAAAGAAGAAAAAACCAG    chr4_75970356_-             1       SDAD1\n",
       "3  AAAAAAGCTCAAGAAGGAGG    chr2_33588446_-             1      FAM98A\n",
       "4  AAAAAAGGCTGTAAAAGCGT   chr19_19891600_+             1      ZNF253"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "achilles_guide_map = pd.read_csv(data_dir / 'Achilles_guide_map.csv') \\\n",
    "    .clean_names() \\\n",
    "    .assign(hugo_symbol=lambda x: [a.split(' ')[0] for a in x.gene]) \\\n",
    "    .drop(['gene'], axis=1)\n",
    "\n",
    "achilles_guide_map.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'Achilles_dropped_guides.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgrna</th>\n",
       "      <th>genomic_coordinates</th>\n",
       "      <th>gene</th>\n",
       "      <th>n_alignments</th>\n",
       "      <th>fail_reason</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAGCTTCCGCCTGATGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_aligned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAATCCTAAAATAAAATA</td>\n",
       "      <td>chrX_145827835.0_-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>in_dropped_guides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAACAGAATATAGTCAGTG</td>\n",
       "      <td>chrX_145827787.0_-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>guide_dropped_by_ceres</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAACAGGACGATGTGCGGC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_aligned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAACATCGACCGAAAGCGT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_aligned</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sgrna genomic_coordinates gene  n_alignments  \\\n",
       "0  AAAAAGCTTCCGCCTGATGG                 NaN  NaN           0.0   \n",
       "1  AAAAATCCTAAAATAAAATA  chrX_145827835.0_-  NaN           1.0   \n",
       "2  AAAACAGAATATAGTCAGTG  chrX_145827787.0_-  NaN           1.0   \n",
       "3  AAAACAGGACGATGTGCGGC                 NaN  NaN           0.0   \n",
       "4  AAAACATCGACCGAAAGCGT                 NaN  NaN           0.0   \n",
       "\n",
       "              fail_reason  \n",
       "0             not_aligned  \n",
       "1       in_dropped_guides  \n",
       "2  guide_dropped_by_ceres  \n",
       "3             not_aligned  \n",
       "4             not_aligned  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "achilles_dropped_guides = pd.read_csv(data_dir / 'Achilles_dropped_guides.csv') \\\n",
    "    .clean_names() \\\n",
    "    .rename({'unnamed_0': 'sgrna'}, axis=1)\n",
    "\n",
    "achilles_dropped_guides.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2554"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(np.unique(achilles_dropped_guides.sgrna))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "achilles_guide_map = achilles_guide_map[~achilles_guide_map.sgrna.isin(achilles_dropped_guides.sgrna)]\n",
    "achilles_guide_map = achilles_guide_map.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "achilles_guide_map.to_csv(save_dir / 'achilles_guide_map.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'Achilles_replicate_map.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>replicate_id</th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>pdna_batch</th>\n",
       "      <th>passes_qc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pacadd188-311cas9_repb_p6_batch3</td>\n",
       "      <td>ACH-001382</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>kmrc20-311cas9_repa_p6_batch3</td>\n",
       "      <td>ACH-000250</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>253j-311cas9_repa_p5_batch3</td>\n",
       "      <td>ACH-000011</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ocug1-311cas9-repb-p6_batch3</td>\n",
       "      <td>ACH-001619</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>raji-311cas9_repb_p6_batch3</td>\n",
       "      <td>ACH-000654</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       replicate_id   depmap_id  pdna_batch  passes_qc\n",
       "0  pacadd188-311cas9_repb_p6_batch3  ACH-001382           3       True\n",
       "1     kmrc20-311cas9_repa_p6_batch3  ACH-000250           3       True\n",
       "2       253j-311cas9_repa_p5_batch3  ACH-000011           3       True\n",
       "3      ocug1-311cas9-repb-p6_batch3  ACH-001619           3       True\n",
       "4       raji-311cas9_repb_p6_batch3  ACH-000654           3       True"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "achilles_replicate_map = pd.read_csv(data_dir / 'Achilles_replicate_map.csv') \\\n",
    "    .clean_names() \\\n",
    "    .assign(replicate_id=lambda x: x.replicate_id.str.lower())\n",
    "achilles_replicate_map.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.max(achilles_replicate_map.pdna_batch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1638"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(np.unique(achilles_replicate_map.replicate_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.996"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.round(np.mean(achilles_replicate_map.passes_qc), 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "achilles_replicate_map.to_csv(save_dir / 'achilles_replicate_map.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'Achilles_logfold_change.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgrna</th>\n",
       "      <th>replicate_id</th>\n",
       "      <th>lfc</th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>pdna_batch</th>\n",
       "      <th>passes_qc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAAATCCAGCAATGCAG</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>0.289694</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAAACCCGTAGATAGCCT</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>0.170172</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAAAAGAAGAAAAAACCAG</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>-0.695947</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAAAAGCTCAAGAAGGAGG</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>-0.324935</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAAAAGGCTGTAAAAGCGT</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>0.142874</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sgrna                 replicate_id       lfc   depmap_id  \\\n",
       "0  AAAAAAATCCAGCAATGCAG  143b-311cas9_repa_p6_batch3  0.289694  ACH-001001   \n",
       "1  AAAAAACCCGTAGATAGCCT  143b-311cas9_repa_p6_batch3  0.170172  ACH-001001   \n",
       "2  AAAAAAGAAGAAAAAACCAG  143b-311cas9_repa_p6_batch3 -0.695947  ACH-001001   \n",
       "3  AAAAAAGCTCAAGAAGGAGG  143b-311cas9_repa_p6_batch3 -0.324935  ACH-001001   \n",
       "4  AAAAAAGGCTGTAAAAGCGT  143b-311cas9_repa_p6_batch3  0.142874  ACH-001001   \n",
       "\n",
       "   pdna_batch  passes_qc  \n",
       "0           3       True  \n",
       "1           3       True  \n",
       "2           3       True  \n",
       "3           3       True  \n",
       "4           3       True  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "achilles_logfold_change = pd.read_csv(data_dir / 'Achilles_logfold_change.csv') \\\n",
    "    .rename({'Construct Barcode': 'sgrna'}, axis=1) \\\n",
    "    .set_index('sgrna') \\\n",
    "    .melt(var_name='replicate_id', value_name='lfc', ignore_index=False) \\\n",
    "    .reset_index() \\\n",
    "    .assign(replicate_id=lambda x: x.replicate_id.str.lower()) \\\n",
    "    .merge(achilles_replicate_map, on='replicate_id', how='left') \\\n",
    "    .pipe(lambda x: x[x.passes_qc])\n",
    "\n",
    "achilles_logfold_change.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "if achilles_logfold_change.depmap_id.isnull().values.any():\n",
    "    raise Exception('Some data points are missing cell line assignments.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not np.all(achilles_logfold_change.passes_qc):\n",
    "    raise Exception('Some data does not pass QC.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(121067627, 6)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "achilles_logfold_change.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "achilles_logfold_change.to_csv(save_dir / 'achilles_logfold_change.csv', \n",
    "                               index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'CCLE_mutations.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>hugo_symbol</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>tumor_seq_allele1</th>\n",
       "      <th>cdna_change</th>\n",
       "      <th>codon_change</th>\n",
       "      <th>protein_change</th>\n",
       "      <th>isdeleterious</th>\n",
       "      <th>istcgahotspot</th>\n",
       "      <th>iscosmichotspot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000986</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858743</td>\n",
       "      <td>58858743</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>c.1456G&gt;A</td>\n",
       "      <td>c.(1456-1458)Gac&gt;Aac</td>\n",
       "      <td>p.D486N</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000988</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858810</td>\n",
       "      <td>58858810</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>c.1389G&gt;A</td>\n",
       "      <td>c.(1387-1389)caG&gt;caA</td>\n",
       "      <td>p.Q463Q</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-002182</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858867</td>\n",
       "      <td>58858867</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>c.1332G&gt;C</td>\n",
       "      <td>c.(1330-1332)aaG&gt;aaC</td>\n",
       "      <td>p.K444N</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000985</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858872</td>\n",
       "      <td>58858872</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>c.1327G&gt;A</td>\n",
       "      <td>c.(1327-1329)Gtg&gt;Atg</td>\n",
       "      <td>p.V443M</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-001793</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858914</td>\n",
       "      <td>58858914</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>c.1285G&gt;A</td>\n",
       "      <td>c.(1285-1287)Gac&gt;Aac</td>\n",
       "      <td>p.D429N</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id hugo_symbol chromosome  start_position  end_position  \\\n",
       "0  ACH-000986        A1BG         19        58858743      58858743   \n",
       "1  ACH-000988        A1BG         19        58858810      58858810   \n",
       "2  ACH-002182        A1BG         19        58858867      58858867   \n",
       "3  ACH-000985        A1BG         19        58858872      58858872   \n",
       "4  ACH-001793        A1BG         19        58858914      58858914   \n",
       "\n",
       "  variant_classification variant_type reference_allele tumor_seq_allele1  \\\n",
       "0      Missense_Mutation          SNP                C                 T   \n",
       "1                 Silent          SNP                C                 T   \n",
       "2      Missense_Mutation          SNP                C                 G   \n",
       "3      Missense_Mutation          SNP                C                 T   \n",
       "4      Missense_Mutation          SNP                C                 T   \n",
       "\n",
       "  cdna_change          codon_change protein_change isdeleterious  \\\n",
       "0   c.1456G>A  c.(1456-1458)Gac>Aac        p.D486N         False   \n",
       "1   c.1389G>A  c.(1387-1389)caG>caA        p.Q463Q         False   \n",
       "2   c.1332G>C  c.(1330-1332)aaG>aaC        p.K444N         False   \n",
       "3   c.1327G>A  c.(1327-1329)Gtg>Atg        p.V443M         False   \n",
       "4   c.1285G>A  c.(1285-1287)Gac>Aac        p.D429N         False   \n",
       "\n",
       "   istcgahotspot  iscosmichotspot  \n",
       "0          False            False  \n",
       "1          False            False  \n",
       "2          False            False  \n",
       "3          False            False  \n",
       "4          False            False  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccle_mutations_columns = [\n",
    "    'depmap_id', \n",
    "    'hugo_symbol', 'chromosome', 'start_position', 'end_position',\n",
    "    'variant_classification', 'variant_type', 'reference_allele', \n",
    "    'tumor_seq_allele1', 'cdna_change', 'codon_change', 'protein_change',\n",
    "    'isdeleterious', 'istcgahotspot', 'iscosmichotspot'\n",
    "]\n",
    "\n",
    "ccle_mutations = pd.read_csv(data_dir / 'CCLE_mutations.csv', \n",
    "                             delimiter='\\t', low_memory=False) \\\n",
    "    .clean_names() \\\n",
    "    [ccle_mutations_columns]\n",
    "\n",
    "ccle_mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "ccle_mutations.to_csv(save_dir / 'ccle_mutations.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### *KRAS* mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>protein_change</th>\n",
       "      <th>isdeleterious</th>\n",
       "      <th>istcgahotspot</th>\n",
       "      <th>iscosmichotspot</th>\n",
       "      <th>codon</th>\n",
       "      <th>is_kras_hotspot</th>\n",
       "      <th>kras_allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000981</td>\n",
       "      <td>25368390</td>\n",
       "      <td>25368390</td>\n",
       "      <td>frame_shift_del</td>\n",
       "      <td>del</td>\n",
       "      <td>p.K185fs</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>185</td>\n",
       "      <td>False</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-001650</td>\n",
       "      <td>25368390</td>\n",
       "      <td>25368390</td>\n",
       "      <td>frame_shift_del</td>\n",
       "      <td>del</td>\n",
       "      <td>p.K185fs</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>185</td>\n",
       "      <td>False</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-002238</td>\n",
       "      <td>25368390</td>\n",
       "      <td>25368390</td>\n",
       "      <td>frame_shift_del</td>\n",
       "      <td>del</td>\n",
       "      <td>p.K185fs</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>185</td>\n",
       "      <td>False</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000996</td>\n",
       "      <td>25368455</td>\n",
       "      <td>25368455</td>\n",
       "      <td>nonsense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.R164*</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>164</td>\n",
       "      <td>False</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000218</td>\n",
       "      <td>25378561</td>\n",
       "      <td>25378561</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.A146V</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>146</td>\n",
       "      <td>True</td>\n",
       "      <td>A146V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id  start_position  end_position variant_classification  \\\n",
       "0  ACH-000981        25368390      25368390        frame_shift_del   \n",
       "1  ACH-001650        25368390      25368390        frame_shift_del   \n",
       "2  ACH-002238        25368390      25368390        frame_shift_del   \n",
       "3  ACH-000996        25368455      25368455      nonsense_mutation   \n",
       "4  ACH-000218        25378561      25378561      missense_mutation   \n",
       "\n",
       "  variant_type protein_change isdeleterious  istcgahotspot  iscosmichotspot  \\\n",
       "0          del       p.K185fs          True          False            False   \n",
       "1          del       p.K185fs          True          False            False   \n",
       "2          del       p.K185fs          True          False            False   \n",
       "3          snp        p.R164*          True          False            False   \n",
       "4          snp        p.A146V         False           True             True   \n",
       "\n",
       "  codon  is_kras_hotspot kras_allele  \n",
       "0   185            False       other  \n",
       "1   185            False       other  \n",
       "2   185            False       other  \n",
       "3   164            False       other  \n",
       "4   146             True       A146V  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras_mutations_columns = [\n",
    "    'depmap_id', 'start_position', 'end_position', \n",
    "    'variant_classification', 'variant_type', 'protein_change',\n",
    "    'isdeleterious', 'istcgahotspot', 'iscosmichotspot'\n",
    "]\n",
    "\n",
    "kras_hotspot_codons = ['12', '13', '61', '146']\n",
    "\n",
    "kras_mutations = ccle_mutations[ccle_mutations.hugo_symbol == 'KRAS'] \\\n",
    "    [kras_mutations_columns] \\\n",
    "    .assign(\n",
    "        variant_classification=lambda x: x.variant_classification.str.lower(),\n",
    "        variant_type=lambda x: x.variant_type.str.lower(),\n",
    "        codon=lambda x: [re.sub('\\D', '', a) for a in x.protein_change],\n",
    "        is_kras_hotspot=lambda x: x.codon.isin(kras_hotspot_codons)\n",
    "    ) \\\n",
    "    .pipe(lambda x: x[x.variant_classification != 'silent']) \\\n",
    "    .pipe(lambda x: x[x.is_kras_hotspot | x.iscosmichotspot | x.istcgahotspot | x.isdeleterious]) \\\n",
    "    .assign(kras_allele=lambda x: [a if b else \"other\" for a,b in zip(x.protein_change, x.is_kras_hotspot)]) \\\n",
    "    .assign(kras_allele=lambda x: [re.sub('p.', '', a) for a in x.kras_allele]) \\\n",
    "    .drop_duplicates() \\\n",
    "    .reset_index(drop=True)\n",
    "\n",
    "\n",
    "kras_mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kras_allele</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G12D</th>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12V</th>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12C</th>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>other</th>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12A</th>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G13D</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q61H</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12S</th>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12R</th>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A146T</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G13C</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q61K</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A146V</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q61L</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q61R</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12F</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A146P</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             depmap_id\n",
       "kras_allele           \n",
       "G12D                68\n",
       "G12V                47\n",
       "G12C                27\n",
       "other               18\n",
       "G12A                16\n",
       "G13D                14\n",
       "Q61H                 9\n",
       "G12S                 8\n",
       "G12R                 8\n",
       "A146T                7\n",
       "G13C                 6\n",
       "Q61K                 4\n",
       "A146V                3\n",
       "Q61L                 2\n",
       "Q61R                 2\n",
       "G12F                 1\n",
       "A146P                1"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras_mutations[['kras_allele', 'depmap_id']] \\\n",
    "    .groupby('kras_allele') \\\n",
    "    .count() \\\n",
    "    .sort_values('depmap_id', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "kras_mutations.to_csv(save_dir / 'kras_mutations.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>kras_allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000718</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000249</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000264</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000314</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000344</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACH-001094</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ACH-001378</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ACH-001650</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ACH-001857</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id  kras_allele\n",
       "0  ACH-000718            3\n",
       "1  ACH-000249            2\n",
       "2  ACH-000264            2\n",
       "3  ACH-000314            2\n",
       "4  ACH-000344            2\n",
       "5  ACH-001001            2\n",
       "6  ACH-001094            2\n",
       "7  ACH-001378            2\n",
       "8  ACH-001650            2\n",
       "9  ACH-001857            2"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mult_kras_mutations = kras_mutations[['depmap_id', 'kras_allele']] \\\n",
    "    .groupby('depmap_id') \\\n",
    "    .count() \\\n",
    "    .pipe(lambda x: x[x.kras_allele > 1]) \\\n",
    "    .sort_values('kras_allele', ascending=False) \\\n",
    "    .reset_index(drop=False)\n",
    "\n",
    "mult_kras_mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>protein_change</th>\n",
       "      <th>isdeleterious</th>\n",
       "      <th>istcgahotspot</th>\n",
       "      <th>iscosmichotspot</th>\n",
       "      <th>codon</th>\n",
       "      <th>is_kras_hotspot</th>\n",
       "      <th>kras_allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000249</td>\n",
       "      <td>25380275</td>\n",
       "      <td>25380275</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61H</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000264</td>\n",
       "      <td>25380277</td>\n",
       "      <td>25380277</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACH-000264</td>\n",
       "      <td>25380277</td>\n",
       "      <td>25380278</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>dnp</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000314</td>\n",
       "      <td>25380275</td>\n",
       "      <td>25380275</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61H</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ACH-000314</td>\n",
       "      <td>25398282</td>\n",
       "      <td>25398282</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G13C</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>13</td>\n",
       "      <td>True</td>\n",
       "      <td>G13C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000344</td>\n",
       "      <td>25380277</td>\n",
       "      <td>25380277</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACH-000344</td>\n",
       "      <td>25380277</td>\n",
       "      <td>25380278</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>dnp</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ACH-000718</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ACH-000718</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12C</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ACH-000718</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>dnp</td>\n",
       "      <td>p.G12F</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12S</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ACH-001094</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12S</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ACH-001378</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ACH-001378</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12S</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ACH-001650</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12D</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ACH-001857</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12C</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ACH-001857</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     depmap_id  start_position  end_position variant_classification  \\\n",
       "1   ACH-000249        25380275      25380275      missense_mutation   \n",
       "3   ACH-000264        25380277      25380277      missense_mutation   \n",
       "6   ACH-000264        25380277      25380278      missense_mutation   \n",
       "2   ACH-000314        25380275      25380275      missense_mutation   \n",
       "10  ACH-000314        25398282      25398282      missense_mutation   \n",
       "4   ACH-000344        25380277      25380277      missense_mutation   \n",
       "5   ACH-000344        25380277      25380278      missense_mutation   \n",
       "11  ACH-000718        25398284      25398284      missense_mutation   \n",
       "16  ACH-000718        25398285      25398285      missense_mutation   \n",
       "15  ACH-000718        25398284      25398285      missense_mutation   \n",
       "18  ACH-001001        25398285      25398285      missense_mutation   \n",
       "19  ACH-001094        25398285      25398285      missense_mutation   \n",
       "12  ACH-001378        25398284      25398284      missense_mutation   \n",
       "20  ACH-001378        25398285      25398285      missense_mutation   \n",
       "14  ACH-001650        25398284      25398284      missense_mutation   \n",
       "17  ACH-001857        25398285      25398285      missense_mutation   \n",
       "13  ACH-001857        25398284      25398284      missense_mutation   \n",
       "\n",
       "   variant_type protein_change isdeleterious  istcgahotspot  iscosmichotspot  \\\n",
       "1           snp         p.Q61H         False           True             True   \n",
       "3           snp         p.Q61K         False           True             True   \n",
       "6           dnp         p.Q61K         False           True             True   \n",
       "2           snp         p.Q61H         False           True             True   \n",
       "10          snp         p.G13C         False           True             True   \n",
       "4           snp         p.Q61K         False           True             True   \n",
       "5           dnp         p.Q61K         False           True             True   \n",
       "11          snp         p.G12V         False           True             True   \n",
       "16          snp         p.G12C         False           True             True   \n",
       "15          dnp         p.G12F         False           True             True   \n",
       "18          snp         p.G12S         False           True             True   \n",
       "19          snp         p.G12S         False           True             True   \n",
       "12          snp         p.G12V         False           True             True   \n",
       "20          snp         p.G12S         False           True             True   \n",
       "14          snp         p.G12D         False           True             True   \n",
       "17          snp         p.G12C         False           True             True   \n",
       "13          snp         p.G12V         False           True             True   \n",
       "\n",
       "   codon  is_kras_hotspot kras_allele  \n",
       "1     61             True        Q61H  \n",
       "3     61             True        Q61K  \n",
       "6     61             True        Q61K  \n",
       "2     61             True        Q61H  \n",
       "10    13             True        G13C  \n",
       "4     61             True        Q61K  \n",
       "5     61             True        Q61K  \n",
       "11    12             True        G12V  \n",
       "16    12             True        G12C  \n",
       "15    12             True        G12F  \n",
       "18    12             True        G12S  \n",
       "19    12             True        G12S  \n",
       "12    12             True        G12V  \n",
       "20    12             True        G12S  \n",
       "14    12             True        G12D  \n",
       "17    12             True        G12C  \n",
       "13    12             True        G12V  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras_mult_mutations_fix = kras_mutations[kras_mutations.depmap_id.isin(mult_kras_mutations.depmap_id)] \\\n",
    "    .reset_index(drop=True) \\\n",
    "    .pipe(lambda x: x[x.is_kras_hotspot]) \\\n",
    "    .sort_values('depmap_id')\n",
    "\n",
    "kras_mult_mutations_fix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>kras_allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000264</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000314</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000344</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-001378</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-001857</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACH-000249</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ACH-001094</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ACH-001650</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id  kras_allele\n",
       "0  ACH-000264            2\n",
       "1  ACH-000314            2\n",
       "2  ACH-000344            2\n",
       "3  ACH-001378            2\n",
       "4  ACH-001857            2\n",
       "5  ACH-000249            1\n",
       "6  ACH-001001            1\n",
       "7  ACH-001094            1\n",
       "8  ACH-001650            1"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras_mut_blacklist = [\n",
    "    'ACH-000718'\n",
    "]\n",
    "\n",
    "kras_mult_mutations_fix[['depmap_id', 'kras_allele']] \\\n",
    "    .pipe(lambda x: x[~x.depmap_id.isin(kras_mut_blacklist)]) \\\n",
    "    .groupby('depmap_id') \\\n",
    "    .count() \\\n",
    "    .sort_values('kras_allele', ascending=False) \\\n",
    "    .reset_index(drop=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'CCLE_segment_cn.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>num_probes</th>\n",
       "      <th>segment_mean</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1969745</td>\n",
       "      <td>286</td>\n",
       "      <td>2.546065</td>\n",
       "      <td>Sanger WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>1</td>\n",
       "      <td>1969746</td>\n",
       "      <td>6354345</td>\n",
       "      <td>365</td>\n",
       "      <td>2.175759</td>\n",
       "      <td>Sanger WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>1</td>\n",
       "      <td>6354346</td>\n",
       "      <td>6958256</td>\n",
       "      <td>100</td>\n",
       "      <td>3.109430</td>\n",
       "      <td>Sanger WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>1</td>\n",
       "      <td>6958257</td>\n",
       "      <td>15977206</td>\n",
       "      <td>884</td>\n",
       "      <td>2.134831</td>\n",
       "      <td>Sanger WES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>1</td>\n",
       "      <td>15977207</td>\n",
       "      <td>16174774</td>\n",
       "      <td>57</td>\n",
       "      <td>2.952592</td>\n",
       "      <td>Sanger WES</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id chromosome     start       end  num_probes  segment_mean  \\\n",
       "0  ACH-000001          1         1   1969745         286      2.546065   \n",
       "1  ACH-000001          1   1969746   6354345         365      2.175759   \n",
       "2  ACH-000001          1   6354346   6958256         100      3.109430   \n",
       "3  ACH-000001          1   6958257  15977206         884      2.134831   \n",
       "4  ACH-000001          1  15977207  16174774          57      2.952592   \n",
       "\n",
       "       source  \n",
       "0  Sanger WES  \n",
       "1  Sanger WES  \n",
       "2  Sanger WES  \n",
       "3  Sanger WES  \n",
       "4  Sanger WES  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccle_segment_cn = pd.read_csv(data_dir / 'CCLE_segment_cn.csv') \\\n",
    "    .clean_names()\n",
    "ccle_segment_cn.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "ccle_segment_cn.to_csv(save_dir / 'ccle_semgent_cn.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'CCLE_gene_cn.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>hugo_symbol</th>\n",
       "      <th>log2_cn_p1</th>\n",
       "      <th>cn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1.179621</td>\n",
       "      <td>2.253141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000002</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1.009801</td>\n",
       "      <td>1.745055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000003</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1.022828</td>\n",
       "      <td>1.781048</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000004</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1.232225</td>\n",
       "      <td>2.428849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000005</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1.151880</td>\n",
       "      <td>2.164136</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id hugo_symbol  log2_cn_p1        cn\n",
       "0  ACH-000001        A1BG    1.179621  2.253141\n",
       "1  ACH-000002        A1BG    1.009801  1.745055\n",
       "2  ACH-000003        A1BG    1.022828  1.781048\n",
       "3  ACH-000004        A1BG    1.232225  2.428849\n",
       "4  ACH-000005        A1BG    1.151880  2.164136"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccle_gene_cn = pd.read_csv(data_dir / 'CCLE_gene_cn.csv') \\\n",
    "    .rename({'Unnamed: 0': 'depmap_id'}, axis=1) \\\n",
    "    .melt(id_vars='depmap_id', var_name='hugo_symbol', value_name='log2_cn_p1') \\\n",
    "    .assign(hugo_symbol=lambda x: [a.split(' ')[0] for a in x.hugo_symbol],\n",
    "            cn=lambda x: np.exp(x.log2_cn_p1) - 1)\n",
    "\n",
    "ccle_gene_cn.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsYAAAHSCAYAAADvxw2lAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/d3fzzAAAACXBIWXMAAAsTAAALEwEAmpwYAAAUUklEQVR4nO3dX6ykd3nY8e8TltIWiALyglz/qWnkVoVKNdHKbmUpoqUNNI1qckFlnCJUITkXpgI1UgXckF5YykVDelOQHEzjqhjXCiCsCqWhlIpyUcOaugFjaCxw7Y1dvAmtgF4Q2fx6sePkFM6y/87snLU/H2l1Zt55Z85jvSz71Tu/mXfWWgEAwPPdT+x6AAAAOAyEMQAAJIwBAKASxgAAUAljAACohDEAAFR1ZNcDVF122WXrmmuu2fUYAAA8xz3wwAN/uNY6ut9jhyKMr7nmmo4fP77rMQAAeI6bmf95uscspQAAgIQxAABUwhgAACphDAAAlTAGAIBKGAMAQCWMAQCgEsYAAFAJYwAAqIQxAABUwhgAACphDAAAlTAGAIBKGAMAQCWMAQCgEsYAAFAJYwAAqIQxAABUwhgAAKo6susB4Lng7vsfO6f9b7nh6i1NAgCcL2eMAQAgYQwAAJUwBgCAShgDAEAljAEAoBLGAABQCWMAAKiEMQAAVMIYAAAqYQwAAJUwBgCAShgDAEAljAEAoBLGAABQCWMAAKiEMQAAVMIYAAAqYQwAAJUwBgCAShgDAEAljAEAoBLGAABQCWMAAKiEMQAAVMIYAAAqYQwAANVZhPHMXDUzn52Zh2fmoZl552b7r87MH8zMg5s/P7/nOe+ZmUdm5usz84Zt/gcAAMBBOHIW+zxd/cpa60sz89LqgZn59Oax31hr/Yu9O8/Mq6ubq9dUf6H6jzPzl9dazxzk4AAAcJDOeMZ4rfXkWutLm9vfrR6urvgxT7mpumet9f211jerR6rrD2JYAADYlnNaYzwz11Svre7fbHrHzPzezHx4Zl622XZF9fiep51on5CemVtn5vjMHD958uS5Tw4AAAforMN4Zl5Sfax611rrO9UHq5+urquerH792V33efr6kQ1r3bHWOrbWOnb06NFznRsAAA7UWYXxzLywU1H8kbXWx6vWWt9aaz2z1vpB9Zv96XKJE9VVe55+ZfXEwY0MAAAH72y+lWKqO6uH11rv37P98j27/WL1lc3t+6qbZ+ZFM/Oq6trqCwc3MgAAHLyz+VaKG6u3Vl+emQc3295bvWVmruvUMolHq1+uWms9NDP3Vl/t1Dda3OYbKQAAOOzOGMZrrc+3/7rhT/2Y59xe3X4BcwEAwEXlyncAAJAwBgCAShgDAEAljAEAoBLGAABQCWMAAKiEMQAAVMIYAAAqYQwAAJUwBgCAShgDAEAljAEAoBLGAABQCWMAAKiEMQAAVMIYAAAqYQwAAJUwBgCAShgDAEAljAEAoBLGAABQCWMAAKiEMQAAVMIYAAAqYQwAAJUwBgCAShgDAEAljAEAoBLGAABQCWMAAKiEMQAAVMIYAAAqYQwAAJUwBgCAShgDAEAljAEAoBLGAABQCWMAAKiEMQAAVMIYAAAqYQwAAJUwBgCAShgDAEAljAEAoBLGAABQCWMAAKiEMQAAVMIYAAAqYQwAAJUwBgCAShgDAEAljAEAoBLGAABQCWMAAKiEMQAAVMIYAAAqYQwAAJUwBgCAShgDAEAljAEAoBLGAABQCWMAAKiEMQAAVMIYAAAqYQwAAJUwBgCAShgDAEB1FmE8M1fNzGdn5uGZeWhm3rnZ/vKZ+fTM/P7m58v2POc9M/PIzHx9Zt6wzf8AAAA4CGdzxvjp6lfWWn+1+hvVbTPz6urd1WfWWtdWn9ncb/PYzdVrqjdWH5iZF2xjeAAAOChnDOO11pNrrS9tbn+3eri6orqpumuz213Vmza3b6ruWWt9f631zeqR6voDnhsAAA7UOa0xnplrqtdW91evXGs9WafiuXrFZrcrqsf3PO3EZhsAABxaZx3GM/OS6mPVu9Za3/lxu+6zbe3zerfOzPGZOX7y5MmzHQMAALbirMJ4Zl7YqSj+yFrr45vN35qZyzePX149tdl+orpqz9OvrJ744ddca92x1jq21jp29OjR850fAAAOxNl8K8VUd1YPr7Xev+eh+6q3bW6/rfrknu03z8yLZuZV1bXVFw5uZAAAOHhHzmKfG6u3Vl+emQc3295b/Vp178y8vXqsenPVWuuhmbm3+mqnvtHitrXWMwc9OAAAHKQzhvFa6/Ptv2646vWnec7t1e0XMBcAAFxUrnwHAAAJYwAAqIQxAABUwhgAACphDAAAlTAGAIBKGAMAQCWMAQCgEsYAAFAJYwAAqIQxAABUwhgAACphDAAAlTAGAIBKGAMAQCWMAQCgEsYAAFAJYwAAqIQxAABUwhgAACphDAAAlTAGAIBKGAMAQCWMAQCgEsYAAFAJYwAAqIQxAABUwhgAACphDAAAlTAGAIBKGAMAQCWMAQCgEsYAAFAJYwAAqIQxAABUwhgAACphDAAAlTAGAIBKGAMAQCWMAQCgEsYAAFAJYwAAqIQxAABUwhgAACphDAAAlTAGAIBKGAMAQCWMAQCgEsYAAFAJYwAAqIQxAABUwhgAACphDAAAlTAGAIBKGAMAQCWMAQCgEsYAAFAJYwAAqIQxAABUwhgAACphDAAAlTAGAIBKGAMAQCWMAQCgEsYAAFAJYwAAqIQxAABUwhgAAKqzCOOZ+fDMPDUzX9mz7Vdn5g9m5sHNn5/f89h7ZuaRmfn6zLxhW4MDAMBBOpszxr9VvXGf7b+x1rpu8+dTVTPz6urm6jWb53xgZl5wUMMCAMC2nDGM11qfq759lq93U3XPWuv7a61vVo9U11/AfAAAcFFcyBrjd8zM722WWrxss+2K6vE9+5zYbAMAgEPtfMP4g9VPV9dVT1a/vtk+++y79nuBmbl1Zo7PzPGTJ0+e5xgAAHAwziuM11rfWms9s9b6QfWb/elyiRPVVXt2vbJ64jSvccda69ha69jRo0fPZwwAADgw5xXGM3P5nru/WD37jRX3VTfPzItm5lXVtdUXLmxEAADYviNn2mFmPlq9rrpsZk5U76teNzPXdWqZxKPVL1ettR6amXurr1ZPV7ettZ7ZyuQAAHCAzhjGa6237LP5zh+z/+3V7RcyFAAAXGyufAcAAAljAACohDEAAFTCGAAAKmEMAACVMAYAgEoYAwBAJYwBAKASxgAAUAljAACohDEAAFTCGAAAKmEMAACVMAYAgEoYAwBAJYwBAKASxgAAUAljAACohDEAAFTCGAAAKmEMAACVMAYAgEoYAwBAJYwBAKASxgAAUAljAACohDEAAFTCGAAAKmEMAACVMAYAgEoYAwBAJYwBAKASxgAAUAljAACohDEAAFTCGAAAKmEMAACVMAYAgEoYAwBAJYwBAKASxgAAUAljAACohDEAAFTCGAAAKmEMAACVMAYAgEoYAwBAJYwBAKASxgAAUAljAACohDEAAFTCGAAAKmEMAACVMAYAgEoYAwBAJYwBAKASxgAAUAljAACohDEAAFTCGAAAKmEMAABVHdn1APB8dPf9j53T/rfccPWWJgEAnuWMMQAAJIwBAKASxgAAUAljAACohDEAAFTCGAAAqrMI45n58Mw8NTNf2bPt5TPz6Zn5/c3Pl+157D0z88jMfH1m3rCtwQEA4CCdzRnj36re+EPb3l19Zq11bfWZzf1m5tXVzdVrNs/5wMy84MCmBQCALTljGK+1Pld9+4c231Tdtbl9V/WmPdvvWWt9f631zeqR6vqDGRUAALbnfNcYv3Kt9WTV5ucrNtuvqB7fs9+JzbYfMTO3zszxmTl+8uTJ8xwDAAAOxkF/+G722bb223Gtdcda69ha69jRo0cPeAwAADg35xvG35qZy6s2P5/abD9RXbVnvyurJ85/PAAAuDjON4zvq962uf226pN7tt88My+amVdV11ZfuLARAQBg+46caYeZ+Wj1uuqymTlRva/6teremXl79Vj15qq11kMzc2/11erp6ra11jNbmh0AAA7MGcN4rfWW0zz0+tPsf3t1+4UMBQAAF5sr3wEAQMIYAAAqYQwAAJUwBgCAShgDAEAljAEAoBLGAABQCWMAAKiEMQAAVMIYAAAqYQwAAJUwBgCAShgDAEAljAEAoBLGAABQCWMAAKiEMQAAVMIYAAAqYQwAAJUwBgCAqo7segA4jO6+/7FdjwAAXGTOGAMAQMIYAAAqYQwAAJUwBgCAShgDAEAljAEAoBLGAABQCWMAAKiEMQAAVMIYAAAqYQwAAJUwBgCAShgDAEAljAEAoBLGAABQCWMAAKjqyK4HAM7s7vsfO6f9b7nh6i1NAgDPXc4YAwBAwhgAACphDAAAlTAGAIBKGAMAQCWMAQCgEsYAAFAJYwAAqIQxAABUwhgAACphDAAAlTAGAIBKGAMAQCWMAQCgEsYAAFAJYwAAqIQxAABUwhgAACphDAAAlTAGAIBKGAMAQFVHdj0AXAx33//YrkcAAA45Z4wBACBhDAAAlTAGAIBKGAMAQCWMAQCgEsYAAFD5ujZ4Tjqfr6e75YartzAJAFw6LiiMZ+bR6rvVM9XTa61jM/Py6t9V11SPVv9wrfW/L2xMAADYroNYSvG31lrXrbWObe6/u/rMWuva6jOb+wAAcKhtY43xTdVdm9t3VW/awu8AAIADdaFhvKrfnZkHZubWzbZXrrWerNr8fMUF/g4AANi6C/3w3Y1rrSdm5hXVp2fma2f7xE1I31p19dU+9AMAwG5d0BnjtdYTm59PVZ+orq++NTOXV21+PnWa596x1jq21jp29OjRCxkDAAAu2HmH8cy8eGZe+uzt6ueqr1T3VW/b7Pa26pMXOiQAAGzbhSyleGX1iZl59nXuXmv9zsx8sbp3Zt5ePVa9+cLHBACA7TrvMF5rfaP66/ts/6Pq9RcyFAAAXGwuCQ0AAAljAACohDEAAFTCGAAAKmEMAACVMAYAgOrCLwkNF93d9z+26xEAgOcgZ4wBACBhDAAAlTAGAIBKGAMAQCWMAQCgEsYAAFAJYwAAqIQxAABUwhgAACphDAAAlUtCcwi4xDMAcBg4YwwAAAljAACohDEAAFTWGLMF1gwDAJciZ4wBACBnjIGL5FzfSbjlhqu3NAkA7E8YA+fFkhkAnmsspQAAgIQxAABUwhgAACphDAAAlQ/fARs+TAfA850zxgAAkDAGAIBKGAMAQCWMAQCgEsYAAFAJYwAAqIQxAABUwhgAACphDAAAlTAGAIDKJaGBQ+pcL1F9yw1Xb2kSAJ4vnDEGAICEMQAAVMIYAAAqa4yfl6zdhHP/e1D+LgA81zljDAAACWMAAKgspeAsnM9bznCx+d8pABfKGWMAAEgYAwBAZSnFJc/bxwAAB8MZYwAASBgDAEAljAEAoBLGAABQCWMAAKiEMQAAVMIYAAAqYQwAAJUwBgCAShgDAEAljAEAoBLGAABQ1ZFdDwDA+bn7/sfOaf9bbrh6q69/Pr8D4DARxgBnadshCsBuWUoBAAA5Y7x1zjDB89f5LEXYpsM2D8Bh87wPY+EKcHC2Hd/+PxjYpq0tpZiZN87M12fmkZl597Z+DwAAHIStnDGemRdU/6r6u9WJ6oszc99a66vb+H3PJd7qBAAuNc+Vd+C3tZTi+uqRtdY3qmbmnuqm6pIPY+EKsDvPlX98L2UX42v8HOfde772zraWUlxRPb7n/onNNgAAOJS2dcZ49tm2/r8dZm6tbt3c/d7M/FH1h1uah4NzWY7TpcKxujQ4Tlv0Swf7co7VeTrg43Cm13ecLgG/tNvj9BdP98C2wvhEddWe+1dWT+zdYa11R3XHs/dn5vha69iW5uGAOE6XDsfq0uA4XTocq0uD43RpOKzHaVtLKb5YXTszr5qZP1PdXN23pd8FAAAXbCtnjNdaT8/MO6r/UL2g+vBa66Ft/C4AADgIW7vAx1rrU9WnzuEpd5x5Fw4Bx+nS4VhdGhynS4djdWlwnC4Nh/I4zVrrzHsBAMBz3NaufAcAAJeSnYexS0dfGmbmwzPz1Mx8ZdezcHozc9XMfHZmHp6Zh2bmnbueif3NzJ+dmS/MzH/fHKt/vuuZOL2ZecHM/LeZ+fe7noXTm5lHZ+bLM/PgzBzf9Tzsb2Z+amZ+e2a+tvn36m/ueqZn7XQpxebS0f+jPZeOrt7i0tGHz8z8bPW96t+stf7arudhfzNzeXX5WutLM/PS6oHqTf5OHT4zM9WL11rfm5kXVp+v3rnW+q87Ho19zMw/rY5VP7nW+oVdz8P+ZubR6thay/cYH2Izc1f1X9ZaH9p8e9mfX2v9nx2PVe3+jPGfXDp6rfXH1bOXjuaQWWt9rvr2rufgx1trPbnW+tLm9nerh3PVyUNpnfK9zd0Xbv740MchNDNXVn+/+tCuZ4FL3cz8ZPWz1Z1Va60/PixRXLsPY5eOhi2ZmWuq11b373gUTmPz9vyD1VPVp9dajtXh9C+rf1b9YMdzcGar+t2ZeWBzhV0On79Unaz+9WZ50odm5sW7HupZuw7jM146Gjh3M/OS6mPVu9Za39n1POxvrfXMWuu6Tl0d9PqZsUzpkJmZX6ieWms9sOtZOCs3rrV+pvp71W2bZYAcLkeqn6k+uNZ6bfV/q0PzGbNdh/EZLx0NnJvNetWPVR9Za3181/NwZpu3Ef9z9cbdTsI+bqz+wWbt6j3V356Zf7vbkTidtdYTm59PVZ/o1JJNDpcT1Yk975D9dqdC+VDYdRi7dDQcoM0Huu6sHl5rvX/X83B6M3N0Zn5qc/vPVX+n+tpOh+JHrLXes9a6cq11Taf+jfpPa61/tOOx2MfMvHjzoeM2b83/XOWblA6Ztdb/qh6fmb+y2fT66tB8QHxrV747Gy4dfemYmY9Wr6sum5kT1fvWWnfudir2cWP11urLm7WrVe/dXImSw+Xy6q7Nt/P8RHXvWstXgcH5e2X1iVPnBzpS3b3W+p3djsRp/JPqI5uTot+o/vGO5/kTrnwHAADtfikFAAAcCsIYAAASxgAAUAljAACohDEAAFTCGAAAKmEMAACVMAYAgKr+H2zz0xjpcdTUAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 864x576 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig = plt.figure(figsize=(12, 8))\n",
    "x = [np.min([a, 6]) for a in ccle_gene_cn.cn.sample(n=1000)]\n",
    "sns.distplot(x, kde=False)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'CCLE_fusions.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>fusion_name</th>\n",
       "      <th>junction_read_count</th>\n",
       "      <th>spanning_frag_count</th>\n",
       "      <th>splice_type</th>\n",
       "      <th>left_gene</th>\n",
       "      <th>left_break_point</th>\n",
       "      <th>right_gene</th>\n",
       "      <th>right_break_point</th>\n",
       "      <th>large_anchor_support</th>\n",
       "      <th>ffpm</th>\n",
       "      <th>left_break_dinuc</th>\n",
       "      <th>left_break_entropy</th>\n",
       "      <th>right_break_dinuc</th>\n",
       "      <th>right_break_entropy</th>\n",
       "      <th>annots</th>\n",
       "      <th>ccle_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-001113</td>\n",
       "      <td>CACNA2D1--SEMA3E</td>\n",
       "      <td>48</td>\n",
       "      <td>53</td>\n",
       "      <td>ONLY_REF_SPLICE</td>\n",
       "      <td>CACNA2D1 (ENSG00000153956)</td>\n",
       "      <td>chr7:82335135:-</td>\n",
       "      <td>SEMA3E (ENSG00000170381)</td>\n",
       "      <td>chr7:83469302:-</td>\n",
       "      <td>YES_LDAS</td>\n",
       "      <td>1.5703</td>\n",
       "      <td>GT</td>\n",
       "      <td>1.9899</td>\n",
       "      <td>AG</td>\n",
       "      <td>1.9899</td>\n",
       "      <td>[\"CCLE_StarF2019\",\"INTRACHROMOSOMAL[chr7:0.92M...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-001113</td>\n",
       "      <td>ADNP2--PLIN3</td>\n",
       "      <td>50</td>\n",
       "      <td>40</td>\n",
       "      <td>INCL_NON_REF_SPLICE</td>\n",
       "      <td>ADNP2 (ENSG00000101544)</td>\n",
       "      <td>chr18:80117650:+</td>\n",
       "      <td>PLIN3 (ENSG00000105355)</td>\n",
       "      <td>chr19:4867657:-</td>\n",
       "      <td>YES_LDAS</td>\n",
       "      <td>1.3993</td>\n",
       "      <td>GT</td>\n",
       "      <td>1.7819</td>\n",
       "      <td>AG</td>\n",
       "      <td>1.8323</td>\n",
       "      <td>[\"CCLE_StarF2019\",\"INTERCHROMOSOMAL[chr18--chr...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-001113</td>\n",
       "      <td>XRCC4--SUGCT</td>\n",
       "      <td>23</td>\n",
       "      <td>104</td>\n",
       "      <td>ONLY_REF_SPLICE</td>\n",
       "      <td>XRCC4 (ENSG00000152422)</td>\n",
       "      <td>chr5:83111203:+</td>\n",
       "      <td>SUGCT (ENSG00000175600)</td>\n",
       "      <td>chr7:40749434:+</td>\n",
       "      <td>YES_LDAS</td>\n",
       "      <td>1.9745</td>\n",
       "      <td>GT</td>\n",
       "      <td>1.9656</td>\n",
       "      <td>AG</td>\n",
       "      <td>1.9656</td>\n",
       "      <td>[\"CCLE_StarF2019\",\"INTERCHROMOSOMAL[chr5--chr7]\"]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-001113</td>\n",
       "      <td>PCYT1A--RBFOX1</td>\n",
       "      <td>26</td>\n",
       "      <td>48</td>\n",
       "      <td>ONLY_REF_SPLICE</td>\n",
       "      <td>PCYT1A (ENSG00000161217)</td>\n",
       "      <td>chr3:196247367:-</td>\n",
       "      <td>RBFOX1 (ENSG00000078328)</td>\n",
       "      <td>chr16:6654603:+</td>\n",
       "      <td>YES_LDAS</td>\n",
       "      <td>1.1505</td>\n",
       "      <td>GT</td>\n",
       "      <td>1.7819</td>\n",
       "      <td>AG</td>\n",
       "      <td>1.8295</td>\n",
       "      <td>[\"CCLE_StarF2019\",\"INTERCHROMOSOMAL[chr3--chr1...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-001113</td>\n",
       "      <td>AF117829.1--PIP4P2</td>\n",
       "      <td>25</td>\n",
       "      <td>31</td>\n",
       "      <td>ONLY_REF_SPLICE</td>\n",
       "      <td>AF117829.1 (ENSG00000251136)</td>\n",
       "      <td>chr8:89757045:-</td>\n",
       "      <td>PIP4P2 (ENSG00000155099)</td>\n",
       "      <td>chr8:90996744:-</td>\n",
       "      <td>YES_LDAS</td>\n",
       "      <td>0.8707</td>\n",
       "      <td>GT</td>\n",
       "      <td>1.8062</td>\n",
       "      <td>AG</td>\n",
       "      <td>1.9329</td>\n",
       "      <td>[\"INTRACHROMOSOMAL[chr8:1.24Mb]\"]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id         fusion_name  junction_read_count  spanning_frag_count  \\\n",
       "0  ACH-001113    CACNA2D1--SEMA3E                   48                   53   \n",
       "1  ACH-001113        ADNP2--PLIN3                   50                   40   \n",
       "2  ACH-001113        XRCC4--SUGCT                   23                  104   \n",
       "3  ACH-001113      PCYT1A--RBFOX1                   26                   48   \n",
       "4  ACH-001113  AF117829.1--PIP4P2                   25                   31   \n",
       "\n",
       "           splice_type                     left_gene  left_break_point  \\\n",
       "0      ONLY_REF_SPLICE    CACNA2D1 (ENSG00000153956)   chr7:82335135:-   \n",
       "1  INCL_NON_REF_SPLICE       ADNP2 (ENSG00000101544)  chr18:80117650:+   \n",
       "2      ONLY_REF_SPLICE       XRCC4 (ENSG00000152422)   chr5:83111203:+   \n",
       "3      ONLY_REF_SPLICE      PCYT1A (ENSG00000161217)  chr3:196247367:-   \n",
       "4      ONLY_REF_SPLICE  AF117829.1 (ENSG00000251136)   chr8:89757045:-   \n",
       "\n",
       "                 right_gene right_break_point large_anchor_support    ffpm  \\\n",
       "0  SEMA3E (ENSG00000170381)   chr7:83469302:-             YES_LDAS  1.5703   \n",
       "1   PLIN3 (ENSG00000105355)   chr19:4867657:-             YES_LDAS  1.3993   \n",
       "2   SUGCT (ENSG00000175600)   chr7:40749434:+             YES_LDAS  1.9745   \n",
       "3  RBFOX1 (ENSG00000078328)   chr16:6654603:+             YES_LDAS  1.1505   \n",
       "4  PIP4P2 (ENSG00000155099)   chr8:90996744:-             YES_LDAS  0.8707   \n",
       "\n",
       "  left_break_dinuc  left_break_entropy right_break_dinuc  right_break_entropy  \\\n",
       "0               GT              1.9899                AG               1.9899   \n",
       "1               GT              1.7819                AG               1.8323   \n",
       "2               GT              1.9656                AG               1.9656   \n",
       "3               GT              1.7819                AG               1.8295   \n",
       "4               GT              1.8062                AG               1.9329   \n",
       "\n",
       "                                              annots  ccle_count  \n",
       "0  [\"CCLE_StarF2019\",\"INTRACHROMOSOMAL[chr7:0.92M...           2  \n",
       "1  [\"CCLE_StarF2019\",\"INTERCHROMOSOMAL[chr18--chr...           2  \n",
       "2  [\"CCLE_StarF2019\",\"INTERCHROMOSOMAL[chr5--chr7]\"]           2  \n",
       "3  [\"CCLE_StarF2019\",\"INTERCHROMOSOMAL[chr3--chr1...           4  \n",
       "4                  [\"INTRACHROMOSOMAL[chr8:1.24Mb]\"]           2  "
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccle_fusions = pd.read_csv(data_dir / 'CCLE_fusions.csv', delimiter='\\t') \\\n",
    "    .clean_names() \\\n",
    "    .rename({'#fusionname': 'fusion_name',\n",
    "             'junctionreadcount': 'junction_read_count',\n",
    "             'spanningfragcount': 'spanning_frag_count',\n",
    "             'splicetype': 'splice_type',\n",
    "             'leftgene': 'left_gene',\n",
    "            'leftbreakpoint': 'left_break_point',\n",
    "            'rightgene': 'right_gene',\n",
    "            'rightbreakpoint': 'right_break_point',\n",
    "            'largeanchorsupport': 'large_anchor_support',\n",
    "            'leftbreakdinuc': 'left_break_dinuc',\n",
    "            'leftbreakentropy': 'left_break_entropy',\n",
    "            'rightbreakdinuc': 'right_break_dinuc',\n",
    "            'rightbreakentropy': 'right_break_entropy'},\n",
    "            axis=1)\n",
    "ccle_fusions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "ccle_fusions.to_csv(save_dir / 'ccle_fusions.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'CCLE_expression_full_v2.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>TSPAN6 (ENSG00000000003)</th>\n",
       "      <th>TNMD (ENSG00000000005)</th>\n",
       "      <th>DPM1 (ENSG00000000419)</th>\n",
       "      <th>SCYL3 (ENSG00000000457)</th>\n",
       "      <th>C1orf112 (ENSG00000000460)</th>\n",
       "      <th>FGR (ENSG00000000938)</th>\n",
       "      <th>CFH (ENSG00000000971)</th>\n",
       "      <th>FUCA2 (ENSG00000001036)</th>\n",
       "      <th>GCLC (ENSG00000001084)</th>\n",
       "      <th>...</th>\n",
       "      <th>AC136297.1 (ENSG00000285985)</th>\n",
       "      <th>BX248415.1 (ENSG00000285986)</th>\n",
       "      <th>AL157886.1 (ENSG00000285987)</th>\n",
       "      <th>AL392086.3 (ENSG00000285988)</th>\n",
       "      <th>AL357123.1 (ENSG00000285989)</th>\n",
       "      <th>AL589743.7 (ENSG00000285990)</th>\n",
       "      <th>AL355312.5 (ENSG00000285991)</th>\n",
       "      <th>AC120036.5 (ENSG00000285992)</th>\n",
       "      <th>AC018931.1 (ENSG00000285993)</th>\n",
       "      <th>AL731559.1 (ENSG00000285994)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-001097</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.667324</td>\n",
       "      <td>1.761285</td>\n",
       "      <td>3.554589</td>\n",
       "      <td>4.358959</td>\n",
       "      <td>0.641546</td>\n",
       "      <td>2.201634</td>\n",
       "      <td>1.941106</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.056584</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-001636</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.198887</td>\n",
       "      <td>2.032101</td>\n",
       "      <td>3.755956</td>\n",
       "      <td>2.601697</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.415488</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.111031</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-001804</td>\n",
       "      <td>4.934988</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.635464</td>\n",
       "      <td>1.929791</td>\n",
       "      <td>2.503349</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.883621</td>\n",
       "      <td>6.772546</td>\n",
       "      <td>3.889474</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.298658</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000534</td>\n",
       "      <td>0.839960</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.376082</td>\n",
       "      <td>2.687061</td>\n",
       "      <td>4.440288</td>\n",
       "      <td>4.938286</td>\n",
       "      <td>0.536053</td>\n",
       "      <td>3.584963</td>\n",
       "      <td>2.935460</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.084064</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-001498</td>\n",
       "      <td>1.744161</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.846744</td>\n",
       "      <td>2.147307</td>\n",
       "      <td>3.666757</td>\n",
       "      <td>5.252098</td>\n",
       "      <td>0.014355</td>\n",
       "      <td>1.641546</td>\n",
       "      <td>4.100978</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.111031</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACH-000742</td>\n",
       "      <td>3.722466</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.029674</td>\n",
       "      <td>2.192194</td>\n",
       "      <td>2.533563</td>\n",
       "      <td>0.150560</td>\n",
       "      <td>5.933809</td>\n",
       "      <td>6.973382</td>\n",
       "      <td>5.192194</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.604071</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACH-002018</td>\n",
       "      <td>5.121015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.989139</td>\n",
       "      <td>1.863938</td>\n",
       "      <td>4.035624</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>0.056584</td>\n",
       "      <td>5.948601</td>\n",
       "      <td>4.627607</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.097611</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ACH-000836</td>\n",
       "      <td>4.632268</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.704180</td>\n",
       "      <td>2.792855</td>\n",
       "      <td>4.079805</td>\n",
       "      <td>0.097611</td>\n",
       "      <td>1.150560</td>\n",
       "      <td>5.729553</td>\n",
       "      <td>4.802193</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.871844</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ACH-000545</td>\n",
       "      <td>4.251719</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.651052</td>\n",
       "      <td>3.440952</td>\n",
       "      <td>3.275007</td>\n",
       "      <td>1.541019</td>\n",
       "      <td>6.659496</td>\n",
       "      <td>6.478648</td>\n",
       "      <td>5.566206</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ACH-001617</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.916357</td>\n",
       "      <td>2.722466</td>\n",
       "      <td>4.120186</td>\n",
       "      <td>0.432959</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.056584</td>\n",
       "      <td>4.426936</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.056584</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 58677 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  TSPAN6 (ENSG00000000003)  TNMD (ENSG00000000005)  \\\n",
       "0  ACH-001097                  0.000000                     0.0   \n",
       "1  ACH-001636                  0.000000                     0.0   \n",
       "2  ACH-001804                  4.934988                     0.0   \n",
       "3  ACH-000534                  0.839960                     0.0   \n",
       "4  ACH-001498                  1.744161                     0.0   \n",
       "5  ACH-000742                  3.722466                     0.0   \n",
       "6  ACH-002018                  5.121015                     0.0   \n",
       "7  ACH-000836                  4.632268                     0.0   \n",
       "8  ACH-000545                  4.251719                     0.0   \n",
       "9  ACH-001617                  0.000000                     0.0   \n",
       "\n",
       "   DPM1 (ENSG00000000419)  SCYL3 (ENSG00000000457)  \\\n",
       "0                4.667324                 1.761285   \n",
       "1                6.198887                 2.032101   \n",
       "2                6.635464                 1.929791   \n",
       "3                5.376082                 2.687061   \n",
       "4                6.846744                 2.147307   \n",
       "5                6.029674                 2.192194   \n",
       "6                5.989139                 1.863938   \n",
       "7                6.704180                 2.792855   \n",
       "8                5.651052                 3.440952   \n",
       "9                6.916357                 2.722466   \n",
       "\n",
       "   C1orf112 (ENSG00000000460)  FGR (ENSG00000000938)  CFH (ENSG00000000971)  \\\n",
       "0                    3.554589               4.358959               0.641546   \n",
       "1                    3.755956               2.601697               0.000000   \n",
       "2                    2.503349               0.000000               4.883621   \n",
       "3                    4.440288               4.938286               0.536053   \n",
       "4                    3.666757               5.252098               0.014355   \n",
       "5                    2.533563               0.150560               5.933809   \n",
       "6                    4.035624               0.028569               0.056584   \n",
       "7                    4.079805               0.097611               1.150560   \n",
       "8                    3.275007               1.541019               6.659496   \n",
       "9                    4.120186               0.432959               0.000000   \n",
       "\n",
       "   FUCA2 (ENSG00000001036)  GCLC (ENSG00000001084)  ...  \\\n",
       "0                 2.201634                1.941106  ...   \n",
       "1                 0.000000                3.415488  ...   \n",
       "2                 6.772546                3.889474  ...   \n",
       "3                 3.584963                2.935460  ...   \n",
       "4                 1.641546                4.100978  ...   \n",
       "5                 6.973382                5.192194  ...   \n",
       "6                 5.948601                4.627607  ...   \n",
       "7                 5.729553                4.802193  ...   \n",
       "8                 6.478648                5.566206  ...   \n",
       "9                 0.056584                4.426936  ...   \n",
       "\n",
       "   AC136297.1 (ENSG00000285985)  BX248415.1 (ENSG00000285986)  \\\n",
       "0                           0.0                           0.0   \n",
       "1                           0.0                           0.0   \n",
       "2                           0.0                           0.0   \n",
       "3                           0.0                           0.0   \n",
       "4                           0.0                           0.0   \n",
       "5                           0.0                           0.0   \n",
       "6                           0.0                           0.0   \n",
       "7                           0.0                           0.0   \n",
       "8                           0.0                           0.0   \n",
       "9                           0.0                           0.0   \n",
       "\n",
       "   AL157886.1 (ENSG00000285987)  AL392086.3 (ENSG00000285988)  \\\n",
       "0                      0.056584                      0.000000   \n",
       "1                      0.000000                      0.000000   \n",
       "2                      0.000000                      0.000000   \n",
       "3                      0.000000                      0.028569   \n",
       "4                      0.000000                      0.000000   \n",
       "5                      0.000000                      0.000000   \n",
       "6                      0.000000                      0.000000   \n",
       "7                      0.000000                      0.000000   \n",
       "8                      0.000000                      0.000000   \n",
       "9                      0.000000                      0.000000   \n",
       "\n",
       "   AL357123.1 (ENSG00000285989)  AL589743.7 (ENSG00000285990)  \\\n",
       "0                           0.0                      0.000000   \n",
       "1                           0.0                      0.000000   \n",
       "2                           0.0                      0.000000   \n",
       "3                           0.0                      1.084064   \n",
       "4                           0.0                      0.111031   \n",
       "5                           0.0                      0.000000   \n",
       "6                           0.0                      0.000000   \n",
       "7                           0.0                      0.871844   \n",
       "8                           0.0                      0.000000   \n",
       "9                           0.0                      0.000000   \n",
       "\n",
       "   AL355312.5 (ENSG00000285991)  AC120036.5 (ENSG00000285992)  \\\n",
       "0                      0.000000                           0.0   \n",
       "1                      0.111031                           0.0   \n",
       "2                      0.298658                           0.0   \n",
       "3                      0.000000                           0.0   \n",
       "4                      0.000000                           0.0   \n",
       "5                      0.604071                           0.0   \n",
       "6                      0.097611                           0.0   \n",
       "7                      0.000000                           0.0   \n",
       "8                      0.000000                           0.0   \n",
       "9                      0.056584                           0.0   \n",
       "\n",
       "   AC018931.1 (ENSG00000285993)  AL731559.1 (ENSG00000285994)  \n",
       "0                           0.0                           0.0  \n",
       "1                           0.0                           0.0  \n",
       "2                           0.0                           0.0  \n",
       "3                           0.0                           0.0  \n",
       "4                           0.0                           0.0  \n",
       "5                           0.0                           0.0  \n",
       "6                           0.0                           0.0  \n",
       "7                           0.0                           0.0  \n",
       "8                           0.0                           0.0  \n",
       "9                           0.0                           0.0  \n",
       "\n",
       "[10 rows x 58677 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(data_dir / 'CCLE_expression_full_v2.csv', nrows=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
