{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare DepMap data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:30:41.523431Z",
     "iopub.status.busy": "2021-05-31T14:30:41.522483Z",
     "iopub.status.idle": "2021-05-31T14:30:45.009706Z",
     "shell.execute_reply": "2021-05-31T14:30:45.010361Z"
    }
   },
   "outputs": [],
   "source": [
    "import re\n",
    "from pathlib import Path\n",
    "\n",
    "import janitor\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from matplotlib import pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:30:45.013442Z",
     "iopub.status.busy": "2021-05-31T14:30:45.012547Z",
     "iopub.status.idle": "2021-05-31T14:30:45.016779Z",
     "shell.execute_reply": "2021-05-31T14:30:45.017393Z"
    }
   },
   "outputs": [],
   "source": [
    "data_dir = Path(\"../data/depmap_20q3\")\n",
    "save_dir = Path(\"../modeling_data\")\n",
    "\n",
    "if not data_dir.exists():\n",
    "    raise Exception(\n",
    "        \"Data directory does not exist. Make sure to download the DepMap data.\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'sample_info.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:30:45.020367Z",
     "iopub.status.busy": "2021-05-31T14:30:45.019463Z",
     "iopub.status.idle": "2021-05-31T14:30:45.063364Z",
     "shell.execute_reply": "2021-05-31T14:30:45.064006Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>stripped_cell_line_name</th>\n",
       "      <th>ccle_name</th>\n",
       "      <th>sex</th>\n",
       "      <th>cas9_activity</th>\n",
       "      <th>primary_or_metastasis</th>\n",
       "      <th>primary_disease</th>\n",
       "      <th>subtype</th>\n",
       "      <th>lineage</th>\n",
       "      <th>lineage_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>NIHOVCAR3</td>\n",
       "      <td>NIHOVCAR3_OVARY</td>\n",
       "      <td>Female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>Ovarian Cancer</td>\n",
       "      <td>Adenocarcinoma, high grade serous</td>\n",
       "      <td>ovary</td>\n",
       "      <td>ovary_adenocarcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000002</td>\n",
       "      <td>HL60</td>\n",
       "      <td>HL60_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>Female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>Acute Myelogenous Leukemia (AML), M3 (Promyelo...</td>\n",
       "      <td>blood</td>\n",
       "      <td>AML</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000003</td>\n",
       "      <td>CACO2</td>\n",
       "      <td>CACO2_LARGE_INTESTINE</td>\n",
       "      <td>Male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Colon/Colorectal Cancer</td>\n",
       "      <td>Adenocarcinoma</td>\n",
       "      <td>colorectal</td>\n",
       "      <td>colorectal_adenocarcinoma</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000004</td>\n",
       "      <td>HEL</td>\n",
       "      <td>HEL_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>Male</td>\n",
       "      <td>52.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>Acute Myelogenous Leukemia (AML), M6 (Erythrol...</td>\n",
       "      <td>blood</td>\n",
       "      <td>AML</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000005</td>\n",
       "      <td>HEL9217</td>\n",
       "      <td>HEL9217_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>Male</td>\n",
       "      <td>86.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>Acute Myelogenous Leukemia (AML), M6 (Erythrol...</td>\n",
       "      <td>blood</td>\n",
       "      <td>AML</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id stripped_cell_line_name  \\\n",
       "0  ACH-000001               NIHOVCAR3   \n",
       "1  ACH-000002                    HL60   \n",
       "2  ACH-000003                   CACO2   \n",
       "3  ACH-000004                     HEL   \n",
       "4  ACH-000005                 HEL9217   \n",
       "\n",
       "                                    ccle_name     sex cas9_activity  \\\n",
       "0                             NIHOVCAR3_OVARY  Female           NaN   \n",
       "1     HL60_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE  Female           NaN   \n",
       "2                       CACO2_LARGE_INTESTINE    Male           NaN   \n",
       "3      HEL_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE    Male          52.4   \n",
       "4  HEL9217_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE    Male          86.6   \n",
       "\n",
       "  primary_or_metastasis          primary_disease  \\\n",
       "0            Metastasis           Ovarian Cancer   \n",
       "1               Primary                 Leukemia   \n",
       "2                   NaN  Colon/Colorectal Cancer   \n",
       "3                   NaN                 Leukemia   \n",
       "4                   NaN                 Leukemia   \n",
       "\n",
       "                                             subtype     lineage  \\\n",
       "0                  Adenocarcinoma, high grade serous       ovary   \n",
       "1  Acute Myelogenous Leukemia (AML), M3 (Promyelo...       blood   \n",
       "2                                     Adenocarcinoma  colorectal   \n",
       "3  Acute Myelogenous Leukemia (AML), M6 (Erythrol...       blood   \n",
       "4  Acute Myelogenous Leukemia (AML), M6 (Erythrol...       blood   \n",
       "\n",
       "             lineage_subtype  \n",
       "0       ovary_adenocarcinoma  \n",
       "1                        AML  \n",
       "2  colorectal_adenocarcinoma  \n",
       "3                        AML  \n",
       "4                        AML  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_info_columns = [\n",
    "    \"depmap_id\",\n",
    "    \"stripped_cell_line_name\",\n",
    "    \"ccle_name\",\n",
    "    \"sex\",\n",
    "    \"cas9_activity\",\n",
    "    \"primary_or_metastasis\",\n",
    "    \"primary_disease\",\n",
    "    \"subtype\",\n",
    "    \"lineage\",\n",
    "    \"lineage_subtype\",\n",
    "]\n",
    "\n",
    "sample_info = pd.read_csv(data_dir / \"sample_info.csv\").clean_names()[\n",
    "    sample_info_columns\n",
    "]\n",
    "sample_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:30:45.066996Z",
     "iopub.status.busy": "2021-05-31T14:30:45.066093Z",
     "iopub.status.idle": "2021-05-31T14:30:45.087872Z",
     "shell.execute_reply": "2021-05-31T14:30:45.088513Z"
    }
   },
   "outputs": [],
   "source": [
    "sample_info.to_csv(save_dir / \"sample_info.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'Achilles_guide_map.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:30:45.091853Z",
     "iopub.status.busy": "2021-05-31T14:30:45.090934Z",
     "iopub.status.idle": "2021-05-31T14:30:45.278855Z",
     "shell.execute_reply": "2021-05-31T14:30:45.279488Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgrna</th>\n",
       "      <th>genome_alignment</th>\n",
       "      <th>n_alignments</th>\n",
       "      <th>hugo_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAAATCCAGCAATGCAG</td>\n",
       "      <td>chr10_110964620_+</td>\n",
       "      <td>1</td>\n",
       "      <td>SHOC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAAACCCGTAGATAGCCT</td>\n",
       "      <td>chr12_95003615_+</td>\n",
       "      <td>1</td>\n",
       "      <td>NDUFA12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAAAAGAAGAAAAAACCAG</td>\n",
       "      <td>chr4_75970356_-</td>\n",
       "      <td>1</td>\n",
       "      <td>SDAD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAAAAGCTCAAGAAGGAGG</td>\n",
       "      <td>chr2_33588446_-</td>\n",
       "      <td>1</td>\n",
       "      <td>FAM98A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAAAAGGCTGTAAAAGCGT</td>\n",
       "      <td>chr19_19891600_+</td>\n",
       "      <td>1</td>\n",
       "      <td>ZNF253</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sgrna   genome_alignment  n_alignments hugo_symbol\n",
       "0  AAAAAAATCCAGCAATGCAG  chr10_110964620_+             1       SHOC2\n",
       "1  AAAAAACCCGTAGATAGCCT   chr12_95003615_+             1     NDUFA12\n",
       "2  AAAAAAGAAGAAAAAACCAG    chr4_75970356_-             1       SDAD1\n",
       "3  AAAAAAGCTCAAGAAGGAGG    chr2_33588446_-             1      FAM98A\n",
       "4  AAAAAAGGCTGTAAAAGCGT   chr19_19891600_+             1      ZNF253"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "achilles_guide_map = (\n",
    "    pd.read_csv(data_dir / \"Achilles_guide_map.csv\")\n",
    "    .clean_names()\n",
    "    .assign(hugo_symbol=lambda x: [a.split(\" \")[0] for a in x.gene])\n",
    "    .drop([\"gene\"], axis=1)\n",
    ")\n",
    "\n",
    "achilles_guide_map.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'Achilles_dropped_guides.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:30:45.282534Z",
     "iopub.status.busy": "2021-05-31T14:30:45.281625Z",
     "iopub.status.idle": "2021-05-31T14:30:45.303031Z",
     "shell.execute_reply": "2021-05-31T14:30:45.303674Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgrna</th>\n",
       "      <th>genomic_coordinates</th>\n",
       "      <th>gene</th>\n",
       "      <th>n_alignments</th>\n",
       "      <th>fail_reason</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAGCTTCCGCCTGATGG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_aligned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAATCCTAAAATAAAATA</td>\n",
       "      <td>chrX_145827835.0_-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>in_dropped_guides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAACAGAATATAGTCAGTG</td>\n",
       "      <td>chrX_145827787.0_-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>guide_dropped_by_ceres</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAACAGGACGATGTGCGGC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_aligned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAACATCGACCGAAAGCGT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_aligned</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sgrna genomic_coordinates gene  n_alignments  \\\n",
       "0  AAAAAGCTTCCGCCTGATGG                 NaN  NaN           0.0   \n",
       "1  AAAAATCCTAAAATAAAATA  chrX_145827835.0_-  NaN           1.0   \n",
       "2  AAAACAGAATATAGTCAGTG  chrX_145827787.0_-  NaN           1.0   \n",
       "3  AAAACAGGACGATGTGCGGC                 NaN  NaN           0.0   \n",
       "4  AAAACATCGACCGAAAGCGT                 NaN  NaN           0.0   \n",
       "\n",
       "              fail_reason  \n",
       "0             not_aligned  \n",
       "1       in_dropped_guides  \n",
       "2  guide_dropped_by_ceres  \n",
       "3             not_aligned  \n",
       "4             not_aligned  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "achilles_dropped_guides = (\n",
    "    pd.read_csv(data_dir / \"Achilles_dropped_guides.csv\")\n",
    "    .clean_names()\n",
    "    .rename({\"unnamed_0\": \"sgrna\"}, axis=1)\n",
    ")\n",
    "\n",
    "achilles_dropped_guides.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:30:45.306661Z",
     "iopub.status.busy": "2021-05-31T14:30:45.305768Z",
     "iopub.status.idle": "2021-05-31T14:30:45.312592Z",
     "shell.execute_reply": "2021-05-31T14:30:45.313210Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2554"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(np.unique(achilles_dropped_guides.sgrna))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:30:45.316046Z",
     "iopub.status.busy": "2021-05-31T14:30:45.315154Z",
     "iopub.status.idle": "2021-05-31T14:30:45.338038Z",
     "shell.execute_reply": "2021-05-31T14:30:45.338696Z"
    }
   },
   "outputs": [],
   "source": [
    "achilles_guide_map = achilles_guide_map[\n",
    "    ~achilles_guide_map.sgrna.isin(achilles_dropped_guides.sgrna)\n",
    "]\n",
    "achilles_guide_map = achilles_guide_map.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:30:45.341718Z",
     "iopub.status.busy": "2021-05-31T14:30:45.340821Z",
     "iopub.status.idle": "2021-05-31T14:30:45.494104Z",
     "shell.execute_reply": "2021-05-31T14:30:45.494759Z"
    }
   },
   "outputs": [],
   "source": [
    "achilles_guide_map.to_csv(save_dir / \"achilles_guide_map.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'Achilles_replicate_map.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:30:45.497813Z",
     "iopub.status.busy": "2021-05-31T14:30:45.496897Z",
     "iopub.status.idle": "2021-05-31T14:30:45.516993Z",
     "shell.execute_reply": "2021-05-31T14:30:45.517635Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>replicate_id</th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>pdna_batch</th>\n",
       "      <th>passes_qc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pe_ca-pj-34clone c12-311cas9_repb_p5_batch2</td>\n",
       "      <td>ACH-000606</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nci-h1944-311cas9 rep c p5_batch3</td>\n",
       "      <td>ACH-000414</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mdamb231-311cas9_repb_p5_batch2</td>\n",
       "      <td>ACH-000768</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>skrc20-311cas9_repa_p6_batch3</td>\n",
       "      <td>ACH-000385</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>skpndw-311cas9_repb_p6_batch3</td>\n",
       "      <td>ACH-001193</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  replicate_id   depmap_id  pdna_batch  \\\n",
       "0  pe_ca-pj-34clone c12-311cas9_repb_p5_batch2  ACH-000606           2   \n",
       "1            nci-h1944-311cas9 rep c p5_batch3  ACH-000414           3   \n",
       "2              mdamb231-311cas9_repb_p5_batch2  ACH-000768           2   \n",
       "3                skrc20-311cas9_repa_p6_batch3  ACH-000385           3   \n",
       "4                skpndw-311cas9_repb_p6_batch3  ACH-001193           3   \n",
       "\n",
       "   passes_qc  \n",
       "0       True  \n",
       "1       True  \n",
       "2       True  \n",
       "3       True  \n",
       "4       True  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "achilles_replicate_map = (\n",
    "    pd.read_csv(data_dir / \"Achilles_replicate_map.csv\")\n",
    "    .clean_names()\n",
    "    .assign(replicate_id=lambda x: x.replicate_id.str.lower())\n",
    ")\n",
    "achilles_replicate_map.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:30:45.520631Z",
     "iopub.status.busy": "2021-05-31T14:30:45.519738Z",
     "iopub.status.idle": "2021-05-31T14:30:45.525747Z",
     "shell.execute_reply": "2021-05-31T14:30:45.526359Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.max(achilles_replicate_map.pdna_batch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:30:45.529161Z",
     "iopub.status.busy": "2021-05-31T14:30:45.528273Z",
     "iopub.status.idle": "2021-05-31T14:30:45.534922Z",
     "shell.execute_reply": "2021-05-31T14:30:45.535548Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1759"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(np.unique(achilles_replicate_map.replicate_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:30:45.538388Z",
     "iopub.status.busy": "2021-05-31T14:30:45.537508Z",
     "iopub.status.idle": "2021-05-31T14:30:45.543449Z",
     "shell.execute_reply": "2021-05-31T14:30:45.544078Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.948"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.round(np.mean(achilles_replicate_map.passes_qc), 3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:30:45.546828Z",
     "iopub.status.busy": "2021-05-31T14:30:45.545941Z",
     "iopub.status.idle": "2021-05-31T14:30:45.561351Z",
     "shell.execute_reply": "2021-05-31T14:30:45.561994Z"
    }
   },
   "outputs": [],
   "source": [
    "achilles_replicate_map.to_csv(save_dir / \"achilles_replicate_map.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'Achilles_logfold_change.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:30:45.565085Z",
     "iopub.status.busy": "2021-05-31T14:30:45.564178Z",
     "iopub.status.idle": "2021-05-31T14:33:14.858106Z",
     "shell.execute_reply": "2021-05-31T14:33:14.858776Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgrna</th>\n",
       "      <th>replicate_id</th>\n",
       "      <th>lfc</th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>pdna_batch</th>\n",
       "      <th>passes_qc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAAATCCAGCAATGCAG</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>0.289694</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAAACCCGTAGATAGCCT</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>0.170172</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAAAAGAAGAAAAAACCAG</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>-0.695947</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAAAAGCTCAAGAAGGAGG</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>-0.324935</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAAAAGGCTGTAAAAGCGT</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>0.142874</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sgrna                 replicate_id       lfc   depmap_id  \\\n",
       "0  AAAAAAATCCAGCAATGCAG  143b-311cas9_repa_p6_batch3  0.289694  ACH-001001   \n",
       "1  AAAAAACCCGTAGATAGCCT  143b-311cas9_repa_p6_batch3  0.170172  ACH-001001   \n",
       "2  AAAAAAGAAGAAAAAACCAG  143b-311cas9_repa_p6_batch3 -0.695947  ACH-001001   \n",
       "3  AAAAAAGCTCAAGAAGGAGG  143b-311cas9_repa_p6_batch3 -0.324935  ACH-001001   \n",
       "4  AAAAAAGGCTGTAAAAGCGT  143b-311cas9_repa_p6_batch3  0.142874  ACH-001001   \n",
       "\n",
       "   pdna_batch  passes_qc  \n",
       "0           3       True  \n",
       "1           3       True  \n",
       "2           3       True  \n",
       "3           3       True  \n",
       "4           3       True  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "achilles_logfold_change = (\n",
    "    pd.read_csv(data_dir / \"Achilles_logfold_change.csv\")\n",
    "    .rename({\"Construct Barcode\": \"sgrna\"}, axis=1)\n",
    "    .set_index(\"sgrna\")\n",
    "    .melt(var_name=\"replicate_id\", value_name=\"lfc\", ignore_index=False)\n",
    "    .reset_index()\n",
    "    .assign(replicate_id=lambda x: x.replicate_id.str.lower())\n",
    "    .merge(achilles_replicate_map, on=\"replicate_id\", how=\"left\")\n",
    "    .pipe(lambda x: x[x.passes_qc])\n",
    ")\n",
    "\n",
    "achilles_logfold_change.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:33:14.861947Z",
     "iopub.status.busy": "2021-05-31T14:33:14.861047Z",
     "iopub.status.idle": "2021-05-31T14:33:18.487565Z",
     "shell.execute_reply": "2021-05-31T14:33:18.488210Z"
    }
   },
   "outputs": [],
   "source": [
    "if achilles_logfold_change.depmap_id.isnull().values.any():\n",
    "    raise Exception(\"Some data points are missing cell line assignments.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:33:18.491287Z",
     "iopub.status.busy": "2021-05-31T14:33:18.490391Z",
     "iopub.status.idle": "2021-05-31T14:33:18.507747Z",
     "shell.execute_reply": "2021-05-31T14:33:18.508391Z"
    }
   },
   "outputs": [],
   "source": [
    "if not np.all(achilles_logfold_change.passes_qc):\n",
    "    raise Exception(\"Some data does not pass QC.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:33:18.511353Z",
     "iopub.status.busy": "2021-05-31T14:33:18.510465Z",
     "iopub.status.idle": "2021-05-31T14:33:18.516406Z",
     "shell.execute_reply": "2021-05-31T14:33:18.517041Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(123756359, 6)"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "achilles_logfold_change.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:33:18.520038Z",
     "iopub.status.busy": "2021-05-31T14:33:18.519149Z",
     "iopub.status.idle": "2021-05-31T14:40:12.070136Z",
     "shell.execute_reply": "2021-05-31T14:40:12.071298Z"
    }
   },
   "outputs": [],
   "source": [
    "achilles_logfold_change.to_csv(save_dir / \"achilles_logfold_change.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'CCLE_mutations.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:40:12.076748Z",
     "iopub.status.busy": "2021-05-31T14:40:12.075224Z",
     "iopub.status.idle": "2021-05-31T14:40:26.705516Z",
     "shell.execute_reply": "2021-05-31T14:40:26.706185Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>hugo_symbol</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>tumor_seq_allele1</th>\n",
       "      <th>cdna_change</th>\n",
       "      <th>codon_change</th>\n",
       "      <th>protein_change</th>\n",
       "      <th>isdeleterious</th>\n",
       "      <th>istcgahotspot</th>\n",
       "      <th>iscosmichotspot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000986</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858743</td>\n",
       "      <td>58858743</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>c.1456G&gt;A</td>\n",
       "      <td>c.(1456-1458)Gac&gt;Aac</td>\n",
       "      <td>p.D486N</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000988</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858810</td>\n",
       "      <td>58858810</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>c.1389G&gt;A</td>\n",
       "      <td>c.(1387-1389)caG&gt;caA</td>\n",
       "      <td>p.Q463Q</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-002182</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858867</td>\n",
       "      <td>58858867</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>c.1332G&gt;C</td>\n",
       "      <td>c.(1330-1332)aaG&gt;aaC</td>\n",
       "      <td>p.K444N</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000985</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858872</td>\n",
       "      <td>58858872</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>c.1327G&gt;A</td>\n",
       "      <td>c.(1327-1329)Gtg&gt;Atg</td>\n",
       "      <td>p.V443M</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-001793</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858914</td>\n",
       "      <td>58858914</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>c.1285G&gt;A</td>\n",
       "      <td>c.(1285-1287)Gac&gt;Aac</td>\n",
       "      <td>p.D429N</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id hugo_symbol chromosome  start_position  end_position  \\\n",
       "0  ACH-000986        A1BG         19        58858743      58858743   \n",
       "1  ACH-000988        A1BG         19        58858810      58858810   \n",
       "2  ACH-002182        A1BG         19        58858867      58858867   \n",
       "3  ACH-000985        A1BG         19        58858872      58858872   \n",
       "4  ACH-001793        A1BG         19        58858914      58858914   \n",
       "\n",
       "  variant_classification variant_type reference_allele tumor_seq_allele1  \\\n",
       "0      Missense_Mutation          SNP                C                 T   \n",
       "1                 Silent          SNP                C                 T   \n",
       "2      Missense_Mutation          SNP                C                 G   \n",
       "3      Missense_Mutation          SNP                C                 T   \n",
       "4      Missense_Mutation          SNP                C                 T   \n",
       "\n",
       "  cdna_change          codon_change protein_change isdeleterious  \\\n",
       "0   c.1456G>A  c.(1456-1458)Gac>Aac        p.D486N         False   \n",
       "1   c.1389G>A  c.(1387-1389)caG>caA        p.Q463Q         False   \n",
       "2   c.1332G>C  c.(1330-1332)aaG>aaC        p.K444N         False   \n",
       "3   c.1327G>A  c.(1327-1329)Gtg>Atg        p.V443M         False   \n",
       "4   c.1285G>A  c.(1285-1287)Gac>Aac        p.D429N         False   \n",
       "\n",
       "   istcgahotspot  iscosmichotspot  \n",
       "0          False            False  \n",
       "1          False            False  \n",
       "2          False            False  \n",
       "3          False            False  \n",
       "4          False            False  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccle_mutations_columns = [\n",
    "    \"depmap_id\",\n",
    "    \"hugo_symbol\",\n",
    "    \"chromosome\",\n",
    "    \"start_position\",\n",
    "    \"end_position\",\n",
    "    \"variant_classification\",\n",
    "    \"variant_type\",\n",
    "    \"reference_allele\",\n",
    "    \"tumor_seq_allele1\",\n",
    "    \"cdna_change\",\n",
    "    \"codon_change\",\n",
    "    \"protein_change\",\n",
    "    \"isdeleterious\",\n",
    "    \"istcgahotspot\",\n",
    "    \"iscosmichotspot\",\n",
    "]\n",
    "\n",
    "ccle_mutations = pd.read_csv(\n",
    "    data_dir / \"CCLE_mutations.csv\", delimiter=\"\\t\", low_memory=False\n",
    ").clean_names()[ccle_mutations_columns]\n",
    "\n",
    "ccle_mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:40:26.709415Z",
     "iopub.status.busy": "2021-05-31T14:40:26.708520Z",
     "iopub.status.idle": "2021-05-31T14:40:32.412792Z",
     "shell.execute_reply": "2021-05-31T14:40:32.413988Z"
    }
   },
   "outputs": [],
   "source": [
    "ccle_mutations.to_csv(save_dir / \"ccle_mutations.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### *KRAS* mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:40:32.419515Z",
     "iopub.status.busy": "2021-05-31T14:40:32.417813Z",
     "iopub.status.idle": "2021-05-31T14:40:36.679464Z",
     "shell.execute_reply": "2021-05-31T14:40:36.680111Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>protein_change</th>\n",
       "      <th>isdeleterious</th>\n",
       "      <th>istcgahotspot</th>\n",
       "      <th>iscosmichotspot</th>\n",
       "      <th>codon</th>\n",
       "      <th>is_kras_hotspot</th>\n",
       "      <th>kras_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000981</td>\n",
       "      <td>25368390</td>\n",
       "      <td>25368390</td>\n",
       "      <td>frame_shift_del</td>\n",
       "      <td>del</td>\n",
       "      <td>p.K185fs</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>185</td>\n",
       "      <td>False</td>\n",
       "      <td>frame_shift_del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-001650</td>\n",
       "      <td>25368390</td>\n",
       "      <td>25368390</td>\n",
       "      <td>frame_shift_del</td>\n",
       "      <td>del</td>\n",
       "      <td>p.K185fs</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>185</td>\n",
       "      <td>False</td>\n",
       "      <td>frame_shift_del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-002238</td>\n",
       "      <td>25368390</td>\n",
       "      <td>25368390</td>\n",
       "      <td>frame_shift_del</td>\n",
       "      <td>del</td>\n",
       "      <td>p.K185fs</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>185</td>\n",
       "      <td>False</td>\n",
       "      <td>frame_shift_del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000996</td>\n",
       "      <td>25368455</td>\n",
       "      <td>25368455</td>\n",
       "      <td>nonsense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.R164*</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>164</td>\n",
       "      <td>False</td>\n",
       "      <td>nonsense_mutation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000218</td>\n",
       "      <td>25378561</td>\n",
       "      <td>25378561</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.A146V</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>146</td>\n",
       "      <td>True</td>\n",
       "      <td>A146V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id  start_position  end_position variant_classification  \\\n",
       "0  ACH-000981        25368390      25368390        frame_shift_del   \n",
       "1  ACH-001650        25368390      25368390        frame_shift_del   \n",
       "2  ACH-002238        25368390      25368390        frame_shift_del   \n",
       "3  ACH-000996        25368455      25368455      nonsense_mutation   \n",
       "4  ACH-000218        25378561      25378561      missense_mutation   \n",
       "\n",
       "  variant_type protein_change isdeleterious  istcgahotspot  iscosmichotspot  \\\n",
       "0          del       p.K185fs          True          False            False   \n",
       "1          del       p.K185fs          True          False            False   \n",
       "2          del       p.K185fs          True          False            False   \n",
       "3          snp        p.R164*          True          False            False   \n",
       "4          snp        p.A146V         False           True             True   \n",
       "\n",
       "  codon  is_kras_hotspot      kras_mutation  \n",
       "0   185            False    frame_shift_del  \n",
       "1   185            False    frame_shift_del  \n",
       "2   185            False    frame_shift_del  \n",
       "3   164            False  nonsense_mutation  \n",
       "4   146             True              A146V  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras_mutations_columns = [\n",
    "    \"depmap_id\",\n",
    "    \"start_position\",\n",
    "    \"end_position\",\n",
    "    \"variant_classification\",\n",
    "    \"variant_type\",\n",
    "    \"protein_change\",\n",
    "    \"isdeleterious\",\n",
    "    \"istcgahotspot\",\n",
    "    \"iscosmichotspot\",\n",
    "]\n",
    "\n",
    "kras_hotspot_codons = [\"12\", \"13\", \"61\", \"146\"]\n",
    "\n",
    "\n",
    "kras_mutations = (\n",
    "    ccle_mutations[ccle_mutations.hugo_symbol == \"KRAS\"][kras_mutations_columns]\n",
    "    .assign(\n",
    "        variant_classification=lambda x: x.variant_classification.str.lower(),\n",
    "        variant_type=lambda x: x.variant_type.str.lower(),\n",
    "        codon=lambda x: [re.sub(r\"\\D\", \"\", a) for a in x.protein_change],\n",
    "        is_kras_hotspot=lambda x: x.codon.isin(kras_hotspot_codons),\n",
    "    )\n",
    "    .pipe(lambda x: x[x.variant_classification != \"silent\"])\n",
    "    .pipe(\n",
    "        lambda x: x[\n",
    "            x.is_kras_hotspot | x.iscosmichotspot | x.istcgahotspot | x.isdeleterious\n",
    "        ]\n",
    "    )\n",
    "    .assign(\n",
    "        kras_mutation=lambda x: [\n",
    "            pc if vt == \"missense_mutation\" else vt\n",
    "            for vt, pc in zip(x.variant_classification, x.protein_change)\n",
    "        ]\n",
    "    )\n",
    "    .assign(kras_mutation=lambda x: [re.sub(\"p\\\\.\", \"\", a) for a in x.kras_mutation])\n",
    "    .drop_duplicates()\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "\n",
    "kras_mutations.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:40:36.683163Z",
     "iopub.status.busy": "2021-05-31T14:40:36.682270Z",
     "iopub.status.idle": "2021-05-31T14:40:36.705735Z",
     "shell.execute_reply": "2021-05-31T14:40:36.706372Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kras_mutation</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G12D</th>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12V</th>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12C</th>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12A</th>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G13D</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q61H</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12R</th>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12S</th>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A146T</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G13C</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q61K</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>frame_shift_del</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A59T</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K117N</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A146V</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V14I</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q61L</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q61R</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>in_frame_ins</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nonsense_mutation</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V14L</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A146P</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L19F</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12F</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D33E</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A59G</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>splice_site</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   depmap_id\n",
       "kras_mutation               \n",
       "G12D                      68\n",
       "G12V                      47\n",
       "G12C                      27\n",
       "G12A                      16\n",
       "G13D                      14\n",
       "Q61H                       9\n",
       "G12R                       8\n",
       "G12S                       8\n",
       "A146T                      7\n",
       "G13C                       6\n",
       "Q61K                       4\n",
       "frame_shift_del            3\n",
       "A59T                       3\n",
       "K117N                      3\n",
       "A146V                      3\n",
       "V14I                       2\n",
       "Q61L                       2\n",
       "Q61R                       2\n",
       "in_frame_ins               1\n",
       "nonsense_mutation          1\n",
       "V14L                       1\n",
       "A146P                      1\n",
       "L19F                       1\n",
       "G12F                       1\n",
       "D33E                       1\n",
       "A59G                       1\n",
       "splice_site                1"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras_mutation_counts = (\n",
    "    kras_mutations[[\"kras_mutation\", \"depmap_id\"]]\n",
    "    .groupby(\"kras_mutation\")\n",
    "    .count()\n",
    "    .sort_values(\"depmap_id\", ascending=False)\n",
    ")\n",
    "\n",
    "# List of KRAS mutations to group as \"other\".\n",
    "kras_others = (\n",
    "    kras_mutation_counts.reset_index().pipe(lambda x: x[x.depmap_id < 2]).kras_mutation\n",
    ")\n",
    "\n",
    "kras_mutation_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:40:36.709281Z",
     "iopub.status.busy": "2021-05-31T14:40:36.708384Z",
     "iopub.status.idle": "2021-05-31T14:40:36.759011Z",
     "shell.execute_reply": "2021-05-31T14:40:36.759690Z"
    }
   },
   "outputs": [],
   "source": [
    "# Assign rare KRAS mutations to \"other\" group.\n",
    "kras_mutations = kras_mutations.assign(\n",
    "    kras_mutation=lambda x: [\n",
    "        \"other\" if kras_others.isin([a]).any() else a for a in x.kras_mutation\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:40:36.762750Z",
     "iopub.status.busy": "2021-05-31T14:40:36.761841Z",
     "iopub.status.idle": "2021-05-31T14:40:36.774641Z",
     "shell.execute_reply": "2021-05-31T14:40:36.775266Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>kras_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000718</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000249</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000264</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000314</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000344</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACH-001094</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ACH-001378</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ACH-001650</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ACH-001857</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id  kras_mutation\n",
       "0  ACH-000718              3\n",
       "1  ACH-000249              2\n",
       "2  ACH-000264              2\n",
       "3  ACH-000314              2\n",
       "4  ACH-000344              2\n",
       "5  ACH-001001              2\n",
       "6  ACH-001094              2\n",
       "7  ACH-001378              2\n",
       "8  ACH-001650              2\n",
       "9  ACH-001857              2"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mult_kras_mutations = (\n",
    "    kras_mutations[[\"depmap_id\", \"kras_mutation\"]]\n",
    "    .groupby(\"depmap_id\")\n",
    "    .count()\n",
    "    .pipe(lambda x: x[x.kras_mutation > 1])\n",
    "    .sort_values(\"kras_mutation\", ascending=False)\n",
    "    .reset_index(drop=False)\n",
    ")\n",
    "\n",
    "mult_kras_mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:40:36.778139Z",
     "iopub.status.busy": "2021-05-31T14:40:36.777246Z",
     "iopub.status.idle": "2021-05-31T14:40:36.796795Z",
     "shell.execute_reply": "2021-05-31T14:40:36.797429Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>protein_change</th>\n",
       "      <th>isdeleterious</th>\n",
       "      <th>istcgahotspot</th>\n",
       "      <th>iscosmichotspot</th>\n",
       "      <th>codon</th>\n",
       "      <th>is_kras_hotspot</th>\n",
       "      <th>kras_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000249</td>\n",
       "      <td>25380275</td>\n",
       "      <td>25380275</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61H</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000264</td>\n",
       "      <td>25380277</td>\n",
       "      <td>25380277</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000264</td>\n",
       "      <td>25380277</td>\n",
       "      <td>25380278</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>dnp</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000314</td>\n",
       "      <td>25380275</td>\n",
       "      <td>25380275</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61H</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000314</td>\n",
       "      <td>25398282</td>\n",
       "      <td>25398282</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G13C</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>13</td>\n",
       "      <td>True</td>\n",
       "      <td>G13C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACH-000344</td>\n",
       "      <td>25380277</td>\n",
       "      <td>25380277</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACH-000344</td>\n",
       "      <td>25380277</td>\n",
       "      <td>25380278</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>dnp</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ACH-000718</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ACH-000718</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12C</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ACH-000718</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>dnp</td>\n",
       "      <td>p.G12F</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12S</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ACH-001094</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12S</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ACH-001378</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ACH-001378</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12S</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ACH-001650</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12D</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ACH-001857</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12C</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ACH-001857</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12V</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     depmap_id  start_position  end_position variant_classification  \\\n",
       "0   ACH-000249        25380275      25380275      missense_mutation   \n",
       "1   ACH-000264        25380277      25380277      missense_mutation   \n",
       "2   ACH-000264        25380277      25380278      missense_mutation   \n",
       "3   ACH-000314        25380275      25380275      missense_mutation   \n",
       "4   ACH-000314        25398282      25398282      missense_mutation   \n",
       "5   ACH-000344        25380277      25380277      missense_mutation   \n",
       "6   ACH-000344        25380277      25380278      missense_mutation   \n",
       "7   ACH-000718        25398284      25398284      missense_mutation   \n",
       "8   ACH-000718        25398285      25398285      missense_mutation   \n",
       "9   ACH-000718        25398284      25398285      missense_mutation   \n",
       "10  ACH-001001        25398285      25398285      missense_mutation   \n",
       "11  ACH-001094        25398285      25398285      missense_mutation   \n",
       "12  ACH-001378        25398284      25398284      missense_mutation   \n",
       "13  ACH-001378        25398285      25398285      missense_mutation   \n",
       "14  ACH-001650        25398284      25398284      missense_mutation   \n",
       "15  ACH-001857        25398285      25398285      missense_mutation   \n",
       "16  ACH-001857        25398284      25398284      missense_mutation   \n",
       "\n",
       "   variant_type protein_change isdeleterious  istcgahotspot  iscosmichotspot  \\\n",
       "0           snp         p.Q61H         False           True             True   \n",
       "1           snp         p.Q61K         False           True             True   \n",
       "2           dnp         p.Q61K         False           True             True   \n",
       "3           snp         p.Q61H         False           True             True   \n",
       "4           snp         p.G13C         False           True             True   \n",
       "5           snp         p.Q61K         False           True             True   \n",
       "6           dnp         p.Q61K         False           True             True   \n",
       "7           snp         p.G12V         False           True             True   \n",
       "8           snp         p.G12C         False           True             True   \n",
       "9           dnp         p.G12F         False           True             True   \n",
       "10          snp         p.G12S         False           True             True   \n",
       "11          snp         p.G12S         False           True             True   \n",
       "12          snp         p.G12V         False           True             True   \n",
       "13          snp         p.G12S         False           True             True   \n",
       "14          snp         p.G12D         False           True             True   \n",
       "15          snp         p.G12C         False           True             True   \n",
       "16          snp         p.G12V         False           True             True   \n",
       "\n",
       "   codon  is_kras_hotspot kras_mutation  \n",
       "0     61             True          Q61H  \n",
       "1     61             True          Q61K  \n",
       "2     61             True          Q61K  \n",
       "3     61             True          Q61H  \n",
       "4     13             True          G13C  \n",
       "5     61             True          Q61K  \n",
       "6     61             True          Q61K  \n",
       "7     12             True          G12V  \n",
       "8     12             True          G12C  \n",
       "9     12             True         other  \n",
       "10    12             True          G12S  \n",
       "11    12             True          G12S  \n",
       "12    12             True          G12V  \n",
       "13    12             True          G12S  \n",
       "14    12             True          G12D  \n",
       "15    12             True          G12C  \n",
       "16    12             True          G12V  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras_mult_mutations_fix = (\n",
    "    kras_mutations[kras_mutations.depmap_id.isin(mult_kras_mutations.depmap_id)]\n",
    "    .reset_index(drop=True)\n",
    "    .pipe(lambda x: x[x.is_kras_hotspot])\n",
    "    .sort_values(\"depmap_id\")\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "kras_mult_mutations_fix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:40:36.800414Z",
     "iopub.status.busy": "2021-05-31T14:40:36.799504Z",
     "iopub.status.idle": "2021-05-31T14:40:36.811314Z",
     "shell.execute_reply": "2021-05-31T14:40:36.811952Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>kras_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000718</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000264</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000314</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000344</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-001378</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACH-001857</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACH-000249</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ACH-001094</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ACH-001650</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id  kras_mutation\n",
       "0  ACH-000718              3\n",
       "1  ACH-000264              2\n",
       "2  ACH-000314              2\n",
       "3  ACH-000344              2\n",
       "4  ACH-001378              2\n",
       "5  ACH-001857              2\n",
       "6  ACH-000249              1\n",
       "7  ACH-001001              1\n",
       "8  ACH-001094              1\n",
       "9  ACH-001650              1"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Cell lines to ignore because they have multiple KRAS mutations that are\n",
    "# not easily resolved.\n",
    "true_kras_double_muts = [\"ACH-000718\", \"ACH-000314\", \"ACH-001378\", \"ACH-001857\"]\n",
    "\n",
    "(\n",
    "    kras_mult_mutations_fix[[\"depmap_id\", \"kras_mutation\"]]\n",
    "    .groupby(\"depmap_id\")\n",
    "    .count()\n",
    "    .sort_values(\"kras_mutation\", ascending=False)\n",
    "    .reset_index(drop=False)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:40:36.814842Z",
     "iopub.status.busy": "2021-05-31T14:40:36.813949Z",
     "iopub.status.idle": "2021-05-31T14:40:36.835059Z",
     "shell.execute_reply": "2021-05-31T14:40:36.835715Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>variant_classification</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>protein_change</th>\n",
       "      <th>isdeleterious</th>\n",
       "      <th>istcgahotspot</th>\n",
       "      <th>iscosmichotspot</th>\n",
       "      <th>codon</th>\n",
       "      <th>is_kras_hotspot</th>\n",
       "      <th>kras_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000249</td>\n",
       "      <td>25380275</td>\n",
       "      <td>25380275</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61H</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000264</td>\n",
       "      <td>25380277</td>\n",
       "      <td>25380277</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000314</td>\n",
       "      <td>25380275</td>\n",
       "      <td>25380275</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61H</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>double_mut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000344</td>\n",
       "      <td>25380277</td>\n",
       "      <td>25380277</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.Q61K</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>True</td>\n",
       "      <td>Q61K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000718</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>double_mut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12S</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ACH-001094</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12S</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ACH-001378</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>double_mut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ACH-001650</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12D</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>G12D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ACH-001857</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>missense_mutation</td>\n",
       "      <td>snp</td>\n",
       "      <td>p.G12C</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>double_mut</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id  start_position  end_position variant_classification  \\\n",
       "0  ACH-000249        25380275      25380275      missense_mutation   \n",
       "1  ACH-000264        25380277      25380277      missense_mutation   \n",
       "2  ACH-000314        25380275      25380275      missense_mutation   \n",
       "3  ACH-000344        25380277      25380277      missense_mutation   \n",
       "4  ACH-000718        25398284      25398284      missense_mutation   \n",
       "5  ACH-001001        25398285      25398285      missense_mutation   \n",
       "6  ACH-001094        25398285      25398285      missense_mutation   \n",
       "7  ACH-001378        25398284      25398284      missense_mutation   \n",
       "8  ACH-001650        25398284      25398284      missense_mutation   \n",
       "9  ACH-001857        25398285      25398285      missense_mutation   \n",
       "\n",
       "  variant_type protein_change  isdeleterious  istcgahotspot  iscosmichotspot  \\\n",
       "0          snp         p.Q61H          False           True             True   \n",
       "1          snp         p.Q61K          False           True             True   \n",
       "2          snp         p.Q61H          False           True             True   \n",
       "3          snp         p.Q61K          False           True             True   \n",
       "4          snp         p.G12V          False           True             True   \n",
       "5          snp         p.G12S          False           True             True   \n",
       "6          snp         p.G12S          False           True             True   \n",
       "7          snp         p.G12V          False           True             True   \n",
       "8          snp         p.G12D          False           True             True   \n",
       "9          snp         p.G12C          False           True             True   \n",
       "\n",
       "  codon  is_kras_hotspot kras_mutation  \n",
       "0    61             True          Q61H  \n",
       "1    61             True          Q61K  \n",
       "2    61             True    double_mut  \n",
       "3    61             True          Q61K  \n",
       "4    12             True    double_mut  \n",
       "5    12             True          G12S  \n",
       "6    12             True          G12S  \n",
       "7    12             True    double_mut  \n",
       "8    12             True          G12D  \n",
       "9    12             True    double_mut  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras_mult_mutations_fix = (\n",
    "    kras_mult_mutations_fix.assign(\n",
    "        kras_mutation=lambda x: [\n",
    "            \"double_mut\" if a in true_kras_double_muts else b\n",
    "            for a, b in zip(x.depmap_id, x.kras_mutation)\n",
    "        ]\n",
    "    )\n",
    "    .groupby(\"depmap_id\")\n",
    "    .first()\n",
    "    .reset_index(drop=False)\n",
    ")\n",
    "kras_mult_mutations_fix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:40:36.838779Z",
     "iopub.status.busy": "2021-05-31T14:40:36.837876Z",
     "iopub.status.idle": "2021-05-31T14:40:36.847035Z",
     "shell.execute_reply": "2021-05-31T14:40:36.847696Z"
    }
   },
   "outputs": [],
   "source": [
    "kras_mutations = kras_mutations.pipe(\n",
    "    lambda x: x[~x.depmap_id.isin(kras_mult_mutations_fix.depmap_id)]\n",
    ")\n",
    "\n",
    "kras_mutations = pd.concat([kras_mutations, kras_mult_mutations_fix])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:40:36.850732Z",
     "iopub.status.busy": "2021-05-31T14:40:36.849802Z",
     "iopub.status.idle": "2021-05-31T14:40:36.860599Z",
     "shell.execute_reply": "2021-05-31T14:40:36.861236Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of cell lines with multiple KRAS mutations: 0\n"
     ]
    }
   ],
   "source": [
    "mult_kras_mutations = (\n",
    "    kras_mutations[[\"depmap_id\", \"kras_mutation\"]]\n",
    "    .groupby(\"depmap_id\")\n",
    "    .count()\n",
    "    .pipe(lambda x: x[x.kras_mutation > 1])\n",
    "    .sort_values(\"kras_mutation\", ascending=False)\n",
    "    .reset_index(drop=False)\n",
    ")\n",
    "\n",
    "print(f\"Number of cell lines with multiple KRAS mutations: {len(mult_kras_mutations)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:40:36.864294Z",
     "iopub.status.busy": "2021-05-31T14:40:36.863388Z",
     "iopub.status.idle": "2021-05-31T14:40:36.872597Z",
     "shell.execute_reply": "2021-05-31T14:40:36.873243Z"
    }
   },
   "outputs": [],
   "source": [
    "kras_mutations.to_csv(save_dir / \"kras_mutations.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'CCLE_segment_cn.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:40:36.876322Z",
     "iopub.status.busy": "2021-05-31T14:40:36.875413Z",
     "iopub.status.idle": "2021-05-31T14:40:37.282680Z",
     "shell.execute_reply": "2021-05-31T14:40:37.283314Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>num_probes</th>\n",
       "      <th>segment_mean</th>\n",
       "      <th>source</th>\n",
       "      <th>copy_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1969745</td>\n",
       "      <td>286</td>\n",
       "      <td>2.546065</td>\n",
       "      <td>Sanger WES</td>\n",
       "      <td>5.840389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>1</td>\n",
       "      <td>1969746</td>\n",
       "      <td>6354345</td>\n",
       "      <td>365</td>\n",
       "      <td>2.175759</td>\n",
       "      <td>Sanger WES</td>\n",
       "      <td>4.518235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>1</td>\n",
       "      <td>6354346</td>\n",
       "      <td>6958256</td>\n",
       "      <td>100</td>\n",
       "      <td>3.109430</td>\n",
       "      <td>Sanger WES</td>\n",
       "      <td>8.630414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>1</td>\n",
       "      <td>6958257</td>\n",
       "      <td>15977206</td>\n",
       "      <td>884</td>\n",
       "      <td>2.134831</td>\n",
       "      <td>Sanger WES</td>\n",
       "      <td>4.391857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>1</td>\n",
       "      <td>15977207</td>\n",
       "      <td>16174774</td>\n",
       "      <td>57</td>\n",
       "      <td>2.952592</td>\n",
       "      <td>Sanger WES</td>\n",
       "      <td>7.741386</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id chromosome     start       end  num_probes  segment_mean  \\\n",
       "0  ACH-000001          1         1   1969745         286      2.546065   \n",
       "1  ACH-000001          1   1969746   6354345         365      2.175759   \n",
       "2  ACH-000001          1   6354346   6958256         100      3.109430   \n",
       "3  ACH-000001          1   6958257  15977206         884      2.134831   \n",
       "4  ACH-000001          1  15977207  16174774          57      2.952592   \n",
       "\n",
       "       source  copy_number  \n",
       "0  Sanger WES     5.840389  \n",
       "1  Sanger WES     4.518235  \n",
       "2  Sanger WES     8.630414  \n",
       "3  Sanger WES     4.391857  \n",
       "4  Sanger WES     7.741386  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccle_segment_cn = (\n",
    "    pd.read_csv(data_dir / \"CCLE_segment_cn.csv\")\n",
    "    .clean_names()\n",
    "    .assign(copy_number=lambda x: [2 ** a for a in x.segment_mean])\n",
    ")\n",
    "ccle_segment_cn.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:40:37.286270Z",
     "iopub.status.busy": "2021-05-31T14:40:37.285387Z",
     "iopub.status.idle": "2021-05-31T14:40:38.738818Z",
     "shell.execute_reply": "2021-05-31T14:40:38.739990Z"
    }
   },
   "outputs": [],
   "source": [
    "ccle_segment_cn.to_csv(save_dir / \"ccle_semgent_cn.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'CCLE_gene_cn.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:40:38.745316Z",
     "iopub.status.busy": "2021-05-31T14:40:38.743738Z",
     "iopub.status.idle": "2021-05-31T14:41:37.584863Z",
     "shell.execute_reply": "2021-05-31T14:41:37.585542Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>hugo_symbol</th>\n",
       "      <th>log2_cn_p1</th>\n",
       "      <th>copy_number</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1.179621</td>\n",
       "      <td>1.265172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000002</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1.009801</td>\n",
       "      <td>1.013633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000003</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1.022828</td>\n",
       "      <td>1.031898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000004</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1.232225</td>\n",
       "      <td>1.349290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000005</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>1.151880</td>\n",
       "      <td>1.222033</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id hugo_symbol  log2_cn_p1  copy_number\n",
       "0  ACH-000001        A1BG    1.179621     1.265172\n",
       "1  ACH-000002        A1BG    1.009801     1.013633\n",
       "2  ACH-000003        A1BG    1.022828     1.031898\n",
       "3  ACH-000004        A1BG    1.232225     1.349290\n",
       "4  ACH-000005        A1BG    1.151880     1.222033"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccle_gene_cn = (\n",
    "    pd.read_csv(data_dir / \"CCLE_gene_cn.csv\")\n",
    "    .rename({\"Unnamed: 0\": \"depmap_id\"}, axis=1)\n",
    "    .melt(id_vars=\"depmap_id\", var_name=\"hugo_symbol\", value_name=\"log2_cn_p1\")\n",
    "    .assign(\n",
    "        hugo_symbol=lambda x: [a.split(\" \")[0] for a in x.hugo_symbol],\n",
    "        copy_number=lambda x: (2 ** x.log2_cn_p1) - 1,\n",
    "    )\n",
    ")\n",
    "\n",
    "ccle_gene_cn.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:41:37.588816Z",
     "iopub.status.busy": "2021-05-31T14:41:37.587910Z",
     "iopub.status.idle": "2021-05-31T14:41:41.871727Z",
     "shell.execute_reply": "2021-05-31T14:41:41.872358Z"
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAtQAAAHSCAYAAADMnFxwAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAb6ElEQVR4nO3df6zd913f8de7doJZm4DTOF3cOHPDXGuAWBo5mUcG8ig/AkOUIloasdKNjoBGEaQTg4I02KZKaKMEtC2NDI1KtzZtIbQU6ICslFaRemntcikpbZM2SWMTKzattxSBmzr57A8fm2vn2Dn2557zveeex0O6uvd8zjn3vKOTmz717ed8v9VaCwAAcGGeNfQAAAAwzwQ1AAB0ENQAANBBUAMAQAdBDQAAHQQ1AAB02Dj0AD0uv/zytn379qHHAABgndu/f/9ftda2jLtvroN6+/bt2bdv39BjAACwzlXVZ892ny0fAADQQVADAEAHQQ0AAB0ENQAAdBDUAADQQVADAEAHQQ0AAB0ENQAAdBDUAADQQVADAEAHQQ0AAB0ENQAAdBDUAADQQVADAEAHQQ0AAB0ENQAAdBDUAADQQVADAECHjUMPALNy7NixLC0tPW199+7d2bRp0wATAQDrgaBmYSwtLeXW29+dzdt2nFo7euCB3JZkz549g80FAMw3Qc1C2bxtR67Yed3QYwAA64g91AAA0EFQAwBAB0ENAAAdBDUAAHQQ1AAA0EFQAwBAB0ENAAAdBDUAAHQQ1AAA0EFQAwBAB0ENAAAdBDUAAHQQ1AAA0EFQAwBAB0ENAAAdBDUAAHQQ1AAA0EFQAwBAB0ENAAAdBDUAAHQQ1AAA0EFQAwBAB0ENAAAdphbUVXVnVR2uqvtWrL2jqpZHXw9X1fJofXtV/e2K++6Y1lwAALCaNk7xd785yX9P8paTC6217zv5c1W9Icn/W/H4z7TWrp3iPAAAsOqmFtSttQ9W1fZx91VVJXl5km+a1usDAMAsDLWH+huSPNZae2DF2guq6k+r6gNV9Q1ne2JV3VJV+6pq35EjR6Y/KQAAnMNQQX1zkrtW3D6U5OrW2ouSvDbJ26rq0nFPbK3tba3taq3t2rJlywxGBQCAs5t5UFfVxiTfk+QdJ9daa19srX1u9PP+JJ9J8sJZzwYAAOdriCPU35zkk621gycXqmpLVW0Y/XxNkh1JHhxgNgAAOC/TPG3eXUk+lGRnVR2sqleP7npFTt/ukSTfmORjVfVnSX4zyY+01j4/rdkAAGC1TPMsHzefZf1fjVm7O8nd05oFAACmxZUSAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOUwvqqrqzqg5X1X0r1n6+qv6yqpZHX9+x4r7XVdWnq+pTVfVt05oLAABW0zSPUL85yU1j1m9rrV07+npvklTVVyd5RZKvGT3n9qraMMXZAABgVUwtqFtrH0zy+Qkf/pIkb2+tfbG19lCSTye5YVqzAQDAahliD/Vrqupjoy0hm0drz09yYMVjDo7WnqaqbqmqfVW178iRI9OeFQAAzmnWQf3GJF+V5Nokh5K8YbReYx7bxv2C1tre1tqu1tquLVu2TGVIAACY1EyDurX2WGvtydbaU0l+NX+3reNgkm0rHnpVkkdnORsAAFyImQZ1VV254uZLk5w8A8h7kryiqr6sql6QZEeSD89yNgAAuBAbp/WLq+quJHuSXF5VB5P8XJI9VXVtTmzneDjJDydJa+3jVfXOJH+R5HiSH22tPTmt2QAAYLVMLahbazePWX7TOR7/+iSvn9Y8AAAwDa6UCAAAHQQ1AAB0ENQAANBBUAMAQAdBDQAAHQQ1AAB0ENQAANBBUAMAQAdBDQAAHQQ1AAB0ENQAANBBUAMAQAdBDQAAHQQ1AAB0ENQAANBBUAMAQAdBDQAAHQQ1AAB0ENQAANBBUAMAQAdBDQAAHQQ1AAB0ENQAANBBUAMAQAdBDQAAHQQ1AAB02Dj0ADANx44dy9LS0mlry8vLeaptGGgiAGC9EtSsS0tLS7n19ndn87Ydp9Ye2X9vLtt5/YBTAQDrkaBm3dq8bUeu2HndqdtHD9w/4DQAwHplDzUAAHQQ1AAA0EFQAwBAB0ENAAAdBDUAAHQQ1AAA0EFQAwBAB0ENAAAdBDUAAHQQ1AAA0EFQAwBAB0ENAAAdBDUAAHQQ1AAA0EFQAwBAB0ENAAAdBDUAAHSYWlBX1Z1Vdbiq7lux9l+r6pNV9bGqeldVfeVofXtV/W1VLY++7pjWXAAAsJqmeYT6zUluOmPtniRf21r7uiT3J3ndivs+01q7dvT1I1OcCwAAVs3Ugrq19sEknz9j7Q9ba8dHN5eSXDWt1wcAgFkYcg/1Dyb53ytuv6Cq/rSqPlBV33C2J1XVLVW1r6r2HTlyZPpTAgDAOQwS1FX1s0mOJ3nraOlQkqtbay9K8tokb6uqS8c9t7W2t7W2q7W2a8uWLbMZGAAAzmLmQV1Vr0rynUm+v7XWkqS19sXW2udGP+9P8pkkL5z1bAAAcL5mGtRVdVOSn0ryXa21v1mxvqWqNox+vibJjiQPznI2AAC4EBun9Yur6q4ke5JcXlUHk/xcTpzV48uS3FNVSbI0OqPHNyb5T1V1PMmTSX6ktfb5sb8YAADWkKkFdWvt5jHLbzrLY+9Ocve0ZgEAgGlxpUQAAOggqAEAoIOgBgCADoIaAAA6CGoAAOggqAEAoIOgBgCADoIaAAA6CGoAAOggqAEAoIOgBgCADoIaAAA6CGoAAOggqAEAoIOgBgCADoIaAAA6CGoAAOggqAEAoIOgBgCADoIaAAA6CGoAAOggqAEAoIOgBgCADoIaAAA6CGoAAOggqAEAoIOgBgCADoIaAAA6CGoAAOggqAEAoIOgBgCADoIaAAA6CGoAAOggqAEAoIOgBgCADoIaAAA6CGoAAOggqAEAoIOgBgCADhMFdVXdOMkaAAAsmkmPUP+3CdcAAGChbDzXnVX1T5N8fZItVfXaFXddmmTDNAcDAIB5cM6gTnJxkueMHnfJivXHk3zvtIYCAIB5cc6gbq19IMkHqurNrbXPzmgmAACYG890hPqkL6uqvUm2r3xOa+2bpjEUAADMi0mD+jeS3JHk15I8Ob1xAABgvkx6lo/jrbU3ttY+3Frbf/LrXE+oqjur6nBV3bdi7bKquqeqHhh937zivtdV1aer6lNV9W0X+M8DAAAzNWlQ/05V/duqunIUxZdV1WXP8Jw3J7npjLWfTvK+1tqOJO8b3U5VfXWSVyT5mtFzbq8qZxEBAGDNm3TLx6tG339yxVpLcs3ZntBa+2BVbT9j+SVJ9ox+/vUkf5zkp0brb2+tfTHJQ1X16SQ3JPnQhPMBAMAgJgrq1toLVun1ntdaOzT6nYeq6orR+vOTLK143MHR2tNU1S1JbkmSq6++epXGAgCACzNRUFfVD4xbb629ZZXmqHG//iyvuTfJ3iTZtWvX2McAAMCsTLrl4/oVP29K8uIkH01yvkH9WFVdOTo6fWWSw6P1g0m2rXjcVUkePc/fDQAAMzfplo8fW3m7qr4iyf+8gNd7T07sx/6F0fffXrH+tqr6pSRbk+xI8uEL+P0AADBTkx6hPtPf5ET0nlVV3ZUTH0C8vKoOJvm5nAjpd1bVq5M8kuRlSdJa+3hVvTPJXyQ5nuRHW2vOdw0AwJo36R7q38nf7WnekOQfJXnnuZ7TWrv5LHe9+CyPf32S108yDwAArBWTHqH+xRU/H0/y2dbawSnMAwAAc2WiC7u01j6Q5JNJLkmyOckT0xwKAADmxURBXVUvz4kPCb4sycuT/ElVfe80BwMAgHkw6ZaPn01yfWvtcJJU1ZYk/yfJb05rMAAAmAcTHaFO8qyTMT3yufN4LgAArFuTHqH+/ar6gyR3jW5/X5L3TmckAACYH+cM6qr6h0me11r7yar6niT/LCcuE/6hJG+dwXwAALCmPdO2jV9O8oUkaa39Vmvtta21W3Pi6PQvT3c0AABY+54pqLe31j525mJrbV+S7VOZCAAA5sgzBfWmc9z35as5CAAAzKNnCuqPVNUPnblYVa9Osn86IwEAwPx4prN8/ESSd1XV9+fvAnpXkouTvHSKcwEAwFw4Z1C31h5L8vVV9c+TfO1o+fdaa3809ckAAGAOTHQe6tba+5O8f8qzAADA3HG1QwAA6CCoAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOghoAADpsnPULVtXOJO9YsXRNkv+Q5CuT/FCSI6P1n2mtvXe20wEAwPmZeVC31j6V5NokqaoNSf4yybuS/Oskt7XWfnHWMwEAwIUaesvHi5N8prX22YHnAACACzJ0UL8iyV0rbr+mqj5WVXdW1eahhgIAgEkNFtRVdXGS70ryG6OlNyb5qpzYDnIoyRvO8rxbqmpfVe07cuTIuIcAAMDMDHmE+tuTfLS19liStNYea6092Vp7KsmvJrlh3JNaa3tba7taa7u2bNkyw3EBAODphgzqm7Niu0dVXbnivpcmuW/mEwEAwHma+Vk+kqSq/l6Sb0nywyuW/0tVXZukJXn4jPtgKp48/qUsLy8/bX337t3ZtGnT7AcCAObOIEHdWvubJM89Y+2VQ8zCYnv80EO548Fj2XrgolNrRw88kNuS7NmzZ7C5AID5MUhQw1py6dZrcsXO64YeAwCYU0OfNg8AAOaaoAYAgA6CGgAAOghqAADoIKgBAKCDoAYAgA6CGgAAOghqAADoIKgBAKCDoAYAgA6CGgAAOghqAADoIKgBAKCDoAYAgA6CGgAAOghqAADoIKgBAKCDoAYAgA6CGgAAOghqAADoIKgBAKCDoAYAgA6CGgAAOghqAADoIKgBAKCDoAYAgA6CGgAAOghqAADoIKgBAKCDoAYAgA6CGgAAOghqAADoIKgBAKCDoAYAgA6CGgAAOghqAADoIKgBAKCDoAYAgA6CGgAAOmwcegDodezYsSwtLZ22try8nKfahoEmAgAWiaBm7i0tLeXW29+dzdt2nFp7ZP+9uWzn9QNOBQAsCkHNurB5245csfO6U7ePHrh/wGkAgEViDzUAAHQQ1AAA0EFQAwBAh0H2UFfVw0m+kOTJJMdba7uq6rIk70iyPcnDSV7eWjs6xHwAADCpIY9Q//PW2rWttV2j2z+d5H2ttR1J3je6DQAAa9pa2vLxkiS/Pvr515N893CjAADAZIYK6pbkD6tqf1XdMlp7XmvtUJKMvl8x0GwAADCxoc5DfWNr7dGquiLJPVX1yUmfOArwW5Lk6quvntZ8AAAwkUGOULfWHh19P5zkXUluSPJYVV2ZJKPvh8/y3L2ttV2ttV1btmyZ1cgAADDWzIO6qp5dVZec/DnJtya5L8l7krxq9LBXJfntWc8GAADna4gtH89L8q6qOvn6b2ut/X5VfSTJO6vq1UkeSfKyAWYDAIDzMvOgbq09mOQfj1n/XJIXz3oeAADosZZOmwcAAHNHUAMAQIehTpsHa9aTx7+U5eXlp63v3r07mzZtmv1AAMCaJqjhDI8feih3PHgsWw9cdGrt6IEHcluSPXv2DDYXALA2CWoY49Kt1+SKndcNPQYAMAfsoQYAgA6CGgAAOghqAADoIKgBAKCDoAYAgA6CGgAAOjhtHnPl2LFjWVpaOm1teXk5T7UNA00EACw6Qc1cWVpayq23vzubt+04tfbI/ntz2c7rB5wKAFhkgpq5s3nbjtMuunL0wP0DTgMALDp7qAEAoIOgBgCADoIaAAA6CGoAAOjgQ4msCeNOh5cku3fvzqZNmwaY6HRPHv9SlpeXT1tbK7MBAMMS1KwJ406Hd/TAA7ktyZ49ewab66THDz2UOx48lq0HLkqytmYDAIYlqFkzzjwd3lpz6dZr1vR8AMAw7KEGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOghoAADoIagAA6CCoAQCgg6AGAIAOG4ceAM7myeNfyvLy8mlry8vLeaptGGYgAIAxBDVr1uOHHsodDx7L1gMXnVp7ZP+9uWzn9QNOBQBwOkHNmnbp1mtyxc7rTt0+euD+AacBAHg6e6gBAKCDI9RM1bFjx7K0tPS09d27d2fTpk0DTAQAsLoENVO1tLSUW29/dzZv23Fq7eiBB3Jbkj179gw2FwDAahHUTN3mbTtO2we9How7A0niyDsALCJBDRdg3BlIHHkHgMU086Cuqm1J3pLk7yd5Ksne1tqvVNXPJ/mhJEdGD/2Z1tp7Zz0fTOrMM5AAAItpiCPUx5P8u9baR6vqkiT7q+qe0X23tdZ+cYCZAADggsw8qFtrh5IcGv38har6RJLnz3oOAABYDYPuoa6q7UlelORPktyY5DVV9QNJ9uXEUeyjA44H58UHFQFgMQ0W1FX1nCR3J/mJ1trjVfXGJP85SRt9f0OSHxzzvFuS3JIkV1999ewGhmfgg4oAsJgGCeqquignYvqtrbXfSpLW2mMr7v/VJL877rmttb1J9ibJrl272vSnZbWNO5K7vLycp9qGYQZaRT6oCACLZ4izfFSSNyX5RGvtl1asXznaX50kL01y36xnYzbGHcl9ZP+9uWzn9QNOBQBwYYY4Qn1jklcm+fOqWh6t/UySm6vq2pzY8vFwkh8eYDZm5MwjuUcP3D/gNAAAF26Is3zcm6TG3OWc0wAAzB1XSoQ14NixY1laWnraujOEAMDaJ6hhDVhaWsqtt787m7ftOLXmDCEAMB8ENawRm7ftcIYQAJhDghpmbNz2jvVy2kAAWESCGmZs3PYOpw0EgPklqBeYD8IN58ztHU4bCADzS1AvMB+EAwDoJ6gX3IV+EG7c0e0nnngiSXLxxRefWrM3GABY7wQ1F2T8PuD3Z+Mlz83WF37dijV7gwGA9U1Qc8HG7QO+aPNWe4NXePL4l7K8vHzamqP2ALC+CGqYoscPPZQ7HjyWrQcuOrXmqD0ArC+CGqbs0q3XOGoPAOvYs4YeAAAA5pmgBgCADoIaAAA6CGoAAOggqAEAoIOgBgCADk6bB3Nu3GXgk2T37t3ZtGnTABMBwGIR1DDnxl0G/uiBB3Jbkj179gw2FwAsCkHNMxp3BNTls6dv3GXLk/FHns+8DDwAMDuCmmc07gioy2dP37jLljvyDABrj6BmImceAXX57Nk487LlAMDa4ywfAADQQVADAEAHWz5gjoz7oKIPiALAsAQ1zJFxH1T0AVEAGJagXhCTnvrOEdC178wPKvqAKAAMS1AviElPfecIKADA+RHUC2TSU985AgoAMDln+QAAgA6OUMM6dD6XLQcA+ghqWIdcthwAZkdQwzrlsuUAMBv2UAMAQAdBDQAAHQQ1AAB0ENQAANDBhxLn3LhLiidOj8Zk/PsDAP0E9Zwbd0lxp0djUv79AYB+gnodOPOS4jDOuIu9LC8v5yuu+qrT/v0Z97gnnngiSXLxxRefcy1xdBuAxSOoYUGMu9jLI/vvzWU7r5/gce/Pxkuem60v/Lpzrjm6DcAiEtRzZNx+1+Xl5TzVNpy2drYjkWc+jsVz5sVejh64f+LHXbR56zOuAcAiEtRzZNx+18mPMD79cQAA9BPUc+bM/dLnc4QRpm3S/deJvdYArB+CGlg1k+6/ttcagPVkzQV1Vd2U5FeSbEjya621Xxh4JOA8TLL/GgDWkzUV1FW1Icn/SPItSQ4m+UhVvae19hfDTna61b4YxrjfN+7/JvfBQtaLnlPz9WwhmfRvbdLXmPS/BZO+7qT/HOO4SM/a4v2APvP2N7SmgjrJDUk+3Vp7MEmq6u1JXpJkTQX1al8MY/yHDcedpswHC1kfek7N17OFZPK/tcleY9L/Fkz6uqv93xFba4bj/YA+8/Y3tNaC+vlJDqy4fTDJPxlolvN25hG3aTzv8UcfzOHnPOfU7S8c/sts/Ntj1ma8tlbmmOu1S56b1TTJ39GF/o2e7fln+32TPm6S5/Y+r/efmQvj/YA+8/a3Uq21oWc4papeluTbWmv/ZnT7lUluaK392IrH3JLkltHNnUk+1fmylyf5q87fwfzwfi8W7/di8X4vDu/1Ylkr7/c/aK1tGXfHWjtCfTDJthW3r0ry6MoHtNb2Jtm7Wi9YVftaa7tW6/extnm/F4v3e7F4vxeH93qxzMP7/ayhBzjDR5LsqKoXVNXFSV6R5D0DzwQAAGe1po5Qt9aOV9VrkvxBTpw2787W2scHHgsAAM5qTQV1krTW3pvkvTN8yVXbPsJc8H4vFu/3YvF+Lw7v9WJZ8+/3mvpQIgAAzJu1tocaAADmykIHdVXdVFWfqqpPV9VPDz0P01NVd1bV4aq6b+hZmK6q2lZV76+qT1TVx6vqx4eeiempqk1V9eGq+rPR+/0fh56J6auqDVX1p1X1u0PPwnRV1cNV9edVtVxV+4ae52wWdsvH6DLn92fFZc6T3LzWLnPO6qiqb0zy10ne0lr72qHnYXqq6sokV7bWPlpVlyTZn+S7/W2vT1VVSZ7dWvvrqrooyb1Jfry19vRrFrNuVNVrk+xKcmlr7TuHnofpqaqHk+xqra2F81Cf1SIfoT51mfPW2hNJTl7mnHWotfbBJJ8feg6mr7V2qLX20dHPX0jyiZy4CivrUDvhr0c3Lxp9LeaRogVRVVcl+RdJfm3oWeCkRQ7qcZc59z+6sI5U1fYkL0ryJwOPwhSN/u//5SSHk9zTWvN+r2+/nOTfJ3lq4DmYjZbkD6tq/+hq2WvSIgd1jVlzVAPWiap6TpK7k/xEa+3xoedhelprT7bWrs2Jq+veUFW2da1TVfWdSQ631vYPPQszc2Nr7bok357kR0dbONecRQ7qZ7zMOTCfRntp707y1tbabw09D7PRWvu/Sf44yU3DTsIU3Zjku0b7at+e5Juq6n8NOxLT1Fp7dPT9cJJ35cSW3TVnkYPaZc5hHRp9SO1NST7RWvuloedhuqpqS1V95ejnL0/yzUk+OehQTE1r7XWttataa9tz4n+3/6i19i8HHospqapnjz5cnqp6dpJvTbImz9a1sEHdWjue5ORlzj+R5J0uc75+VdVdST6UZGdVHayqVw89E1NzY5JX5sSRq+XR13cMPRRTc2WS91fVx3LiQMk9rTWnUoP14XlJ7q2qP0vy4SS/11r7/YFnGmthT5sHAACrYWGPUAMAwGoQ1AAA0EFQAwBAB0ENAAAdBDUAAHQQ1AAA0EFQAwBAB0ENAAAd/j9WfWCf15Zc0QAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 864x576 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig = plt.figure(figsize=(12, 8))\n",
    "x = [np.min([a, 6]) for a in ccle_gene_cn.copy_number.sample(n=1000)]\n",
    "sns.histplot(x, kde=False)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:41:41.875287Z",
     "iopub.status.busy": "2021-05-31T14:41:41.874400Z",
     "iopub.status.idle": "2021-05-31T14:44:26.029081Z",
     "shell.execute_reply": "2021-05-31T14:44:26.030155Z"
    }
   },
   "outputs": [],
   "source": [
    "ccle_gene_cn.to_csv(save_dir / \"ccle_gene_cn.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'CCLE_fusions.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:44:26.035650Z",
     "iopub.status.busy": "2021-05-31T14:44:26.034174Z",
     "iopub.status.idle": "2021-05-31T14:44:26.251178Z",
     "shell.execute_reply": "2021-05-31T14:44:26.251817Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>fusion_name</th>\n",
       "      <th>junction_read_count</th>\n",
       "      <th>spanning_frag_count</th>\n",
       "      <th>splice_type</th>\n",
       "      <th>left_gene</th>\n",
       "      <th>left_break_point</th>\n",
       "      <th>right_gene</th>\n",
       "      <th>right_break_point</th>\n",
       "      <th>large_anchor_support</th>\n",
       "      <th>ffpm</th>\n",
       "      <th>left_break_dinuc</th>\n",
       "      <th>left_break_entropy</th>\n",
       "      <th>right_break_dinuc</th>\n",
       "      <th>right_break_entropy</th>\n",
       "      <th>annots</th>\n",
       "      <th>ccle_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-001113</td>\n",
       "      <td>CACNA2D1--SEMA3E</td>\n",
       "      <td>48</td>\n",
       "      <td>53</td>\n",
       "      <td>ONLY_REF_SPLICE</td>\n",
       "      <td>CACNA2D1 (ENSG00000153956)</td>\n",
       "      <td>chr7:82335135:-</td>\n",
       "      <td>SEMA3E (ENSG00000170381)</td>\n",
       "      <td>chr7:83469302:-</td>\n",
       "      <td>YES_LDAS</td>\n",
       "      <td>1.5703</td>\n",
       "      <td>GT</td>\n",
       "      <td>1.9899</td>\n",
       "      <td>AG</td>\n",
       "      <td>1.9899</td>\n",
       "      <td>[\"CCLE_StarF2019\",\"INTRACHROMOSOMAL[chr7:0.92M...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-001113</td>\n",
       "      <td>ADNP2--PLIN3</td>\n",
       "      <td>50</td>\n",
       "      <td>40</td>\n",
       "      <td>INCL_NON_REF_SPLICE</td>\n",
       "      <td>ADNP2 (ENSG00000101544)</td>\n",
       "      <td>chr18:80117650:+</td>\n",
       "      <td>PLIN3 (ENSG00000105355)</td>\n",
       "      <td>chr19:4867657:-</td>\n",
       "      <td>YES_LDAS</td>\n",
       "      <td>1.3993</td>\n",
       "      <td>GT</td>\n",
       "      <td>1.7819</td>\n",
       "      <td>AG</td>\n",
       "      <td>1.8323</td>\n",
       "      <td>[\"CCLE_StarF2019\",\"INTERCHROMOSOMAL[chr18--chr...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-001113</td>\n",
       "      <td>XRCC4--SUGCT</td>\n",
       "      <td>23</td>\n",
       "      <td>104</td>\n",
       "      <td>ONLY_REF_SPLICE</td>\n",
       "      <td>XRCC4 (ENSG00000152422)</td>\n",
       "      <td>chr5:83111203:+</td>\n",
       "      <td>SUGCT (ENSG00000175600)</td>\n",
       "      <td>chr7:40749434:+</td>\n",
       "      <td>YES_LDAS</td>\n",
       "      <td>1.9745</td>\n",
       "      <td>GT</td>\n",
       "      <td>1.9656</td>\n",
       "      <td>AG</td>\n",
       "      <td>1.9656</td>\n",
       "      <td>[\"CCLE_StarF2019\",\"INTERCHROMOSOMAL[chr5--chr7]\"]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-001113</td>\n",
       "      <td>PCYT1A--RBFOX1</td>\n",
       "      <td>26</td>\n",
       "      <td>48</td>\n",
       "      <td>ONLY_REF_SPLICE</td>\n",
       "      <td>PCYT1A (ENSG00000161217)</td>\n",
       "      <td>chr3:196247367:-</td>\n",
       "      <td>RBFOX1 (ENSG00000078328)</td>\n",
       "      <td>chr16:6654603:+</td>\n",
       "      <td>YES_LDAS</td>\n",
       "      <td>1.1505</td>\n",
       "      <td>GT</td>\n",
       "      <td>1.7819</td>\n",
       "      <td>AG</td>\n",
       "      <td>1.8295</td>\n",
       "      <td>[\"CCLE_StarF2019\",\"INTERCHROMOSOMAL[chr3--chr1...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-001113</td>\n",
       "      <td>AF117829.1--PIP4P2</td>\n",
       "      <td>25</td>\n",
       "      <td>31</td>\n",
       "      <td>ONLY_REF_SPLICE</td>\n",
       "      <td>AF117829.1 (ENSG00000251136)</td>\n",
       "      <td>chr8:89757045:-</td>\n",
       "      <td>PIP4P2 (ENSG00000155099)</td>\n",
       "      <td>chr8:90996744:-</td>\n",
       "      <td>YES_LDAS</td>\n",
       "      <td>0.8707</td>\n",
       "      <td>GT</td>\n",
       "      <td>1.8062</td>\n",
       "      <td>AG</td>\n",
       "      <td>1.9329</td>\n",
       "      <td>[\"INTRACHROMOSOMAL[chr8:1.24Mb]\"]</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id         fusion_name  junction_read_count  spanning_frag_count  \\\n",
       "0  ACH-001113    CACNA2D1--SEMA3E                   48                   53   \n",
       "1  ACH-001113        ADNP2--PLIN3                   50                   40   \n",
       "2  ACH-001113        XRCC4--SUGCT                   23                  104   \n",
       "3  ACH-001113      PCYT1A--RBFOX1                   26                   48   \n",
       "4  ACH-001113  AF117829.1--PIP4P2                   25                   31   \n",
       "\n",
       "           splice_type                     left_gene  left_break_point  \\\n",
       "0      ONLY_REF_SPLICE    CACNA2D1 (ENSG00000153956)   chr7:82335135:-   \n",
       "1  INCL_NON_REF_SPLICE       ADNP2 (ENSG00000101544)  chr18:80117650:+   \n",
       "2      ONLY_REF_SPLICE       XRCC4 (ENSG00000152422)   chr5:83111203:+   \n",
       "3      ONLY_REF_SPLICE      PCYT1A (ENSG00000161217)  chr3:196247367:-   \n",
       "4      ONLY_REF_SPLICE  AF117829.1 (ENSG00000251136)   chr8:89757045:-   \n",
       "\n",
       "                 right_gene right_break_point large_anchor_support    ffpm  \\\n",
       "0  SEMA3E (ENSG00000170381)   chr7:83469302:-             YES_LDAS  1.5703   \n",
       "1   PLIN3 (ENSG00000105355)   chr19:4867657:-             YES_LDAS  1.3993   \n",
       "2   SUGCT (ENSG00000175600)   chr7:40749434:+             YES_LDAS  1.9745   \n",
       "3  RBFOX1 (ENSG00000078328)   chr16:6654603:+             YES_LDAS  1.1505   \n",
       "4  PIP4P2 (ENSG00000155099)   chr8:90996744:-             YES_LDAS  0.8707   \n",
       "\n",
       "  left_break_dinuc  left_break_entropy right_break_dinuc  right_break_entropy  \\\n",
       "0               GT              1.9899                AG               1.9899   \n",
       "1               GT              1.7819                AG               1.8323   \n",
       "2               GT              1.9656                AG               1.9656   \n",
       "3               GT              1.7819                AG               1.8295   \n",
       "4               GT              1.8062                AG               1.9329   \n",
       "\n",
       "                                              annots  ccle_count  \n",
       "0  [\"CCLE_StarF2019\",\"INTRACHROMOSOMAL[chr7:0.92M...           2  \n",
       "1  [\"CCLE_StarF2019\",\"INTERCHROMOSOMAL[chr18--chr...           2  \n",
       "2  [\"CCLE_StarF2019\",\"INTERCHROMOSOMAL[chr5--chr7]\"]           2  \n",
       "3  [\"CCLE_StarF2019\",\"INTERCHROMOSOMAL[chr3--chr1...           4  \n",
       "4                  [\"INTRACHROMOSOMAL[chr8:1.24Mb]\"]           2  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccle_fusions = (\n",
    "    pd.read_csv(data_dir / \"CCLE_fusions.csv\", delimiter=\"\\t\")\n",
    "    .clean_names()\n",
    "    .rename(\n",
    "        {\n",
    "            \"#fusionname\": \"fusion_name\",\n",
    "            \"junctionreadcount\": \"junction_read_count\",\n",
    "            \"spanningfragcount\": \"spanning_frag_count\",\n",
    "            \"splicetype\": \"splice_type\",\n",
    "            \"leftgene\": \"left_gene\",\n",
    "            \"leftbreakpoint\": \"left_break_point\",\n",
    "            \"rightgene\": \"right_gene\",\n",
    "            \"rightbreakpoint\": \"right_break_point\",\n",
    "            \"largeanchorsupport\": \"large_anchor_support\",\n",
    "            \"leftbreakdinuc\": \"left_break_dinuc\",\n",
    "            \"leftbreakentropy\": \"left_break_entropy\",\n",
    "            \"rightbreakdinuc\": \"right_break_dinuc\",\n",
    "            \"rightbreakentropy\": \"right_break_entropy\",\n",
    "        },\n",
    "        axis=1,\n",
    "    )\n",
    ")\n",
    "ccle_fusions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:44:26.254737Z",
     "iopub.status.busy": "2021-05-31T14:44:26.253846Z",
     "iopub.status.idle": "2021-05-31T14:44:26.523287Z",
     "shell.execute_reply": "2021-05-31T14:44:26.523943Z"
    }
   },
   "outputs": [],
   "source": [
    "ccle_fusions.to_csv(save_dir / \"ccle_fusions.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'CCLE_expression_full_v2.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:44:26.527075Z",
     "iopub.status.busy": "2021-05-31T14:44:26.526173Z",
     "iopub.status.idle": "2021-05-31T14:46:28.193920Z",
     "shell.execute_reply": "2021-05-31T14:46:28.194610Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>hugo_symbol</th>\n",
       "      <th>rna_expr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-001097</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-001636</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-001804</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>4.934988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000534</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>0.839960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-001498</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>1.744161</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id hugo_symbol  rna_expr\n",
       "0  ACH-001097      TSPAN6  0.000000\n",
       "1  ACH-001636      TSPAN6  0.000000\n",
       "2  ACH-001804      TSPAN6  4.934988\n",
       "3  ACH-000534      TSPAN6  0.839960\n",
       "4  ACH-001498      TSPAN6  1.744161"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ccle_expression = (\n",
    "    pd.read_csv(data_dir / \"CCLE_expression_full.csv\")\n",
    "    .rename({\"Unnamed: 0\": \"depmap_id\"}, axis=1)\n",
    "    .melt(id_vars=\"depmap_id\", var_name=\"hugo_symbol\", value_name=\"rna_expr\")\n",
    "    .assign(hugo_symbol=lambda x: [a.split(\" \")[0] for a in x.hugo_symbol])\n",
    ")\n",
    "\n",
    "ccle_expression.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:46:28.197929Z",
     "iopub.status.busy": "2021-05-31T14:46:28.197021Z",
     "iopub.status.idle": "2021-05-31T14:48:45.309908Z",
     "shell.execute_reply": "2021-05-31T14:48:45.310984Z"
    }
   },
   "outputs": [],
   "source": [
    "ccle_expression.to_csv(save_dir / \"ccle_expression.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 'Achilles_gene_effect.csv' & 'Achilles_gene_effect_unscaled.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:48:45.316117Z",
     "iopub.status.busy": "2021-05-31T14:48:45.314738Z",
     "iopub.status.idle": "2021-05-31T14:49:32.524864Z",
     "shell.execute_reply": "2021-05-31T14:49:32.525517Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>hugo_symbol</th>\n",
       "      <th>gene_effect</th>\n",
       "      <th>gene_effect_unscaled</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000004</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.181037</td>\n",
       "      <td>0.542787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000005</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>-0.090250</td>\n",
       "      <td>0.440018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000007</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.071568</td>\n",
       "      <td>0.104889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000009</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.110311</td>\n",
       "      <td>0.478258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000011</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>0.275160</td>\n",
       "      <td>0.589944</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id hugo_symbol  gene_effect  gene_effect_unscaled\n",
       "0  ACH-000004        A1BG     0.181037              0.542787\n",
       "1  ACH-000005        A1BG    -0.090250              0.440018\n",
       "2  ACH-000007        A1BG     0.071568              0.104889\n",
       "3  ACH-000009        A1BG     0.110311              0.478258\n",
       "4  ACH-000011        A1BG     0.275160              0.589944"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def prep_gene_effect_data(fpath, col_name):\n",
    "    d = (\n",
    "        pd.read_csv(fpath)\n",
    "        .rename({\"DepMap_ID\": \"depmap_id\"}, axis=1)\n",
    "        .melt(id_vars=\"depmap_id\", var_name=\"hugo_symbol\", value_name=col_name)\n",
    "        .assign(hugo_symbol=lambda x: [a.split(\" \")[0] for a in x.hugo_symbol])\n",
    "    )\n",
    "    return d\n",
    "\n",
    "\n",
    "gene_effect_scaled = prep_gene_effect_data(\n",
    "    data_dir / \"Achilles_gene_effect.csv\", \"gene_effect\"\n",
    ")\n",
    "gene_effect_unscaled = prep_gene_effect_data(\n",
    "    data_dir / \"Achilles_gene_effect_unscaled.csv\", \"gene_effect_unscaled\"\n",
    ")\n",
    "\n",
    "# Check that the two data frames are the same shape.\n",
    "if not gene_effect_scaled.shape == gene_effect_unscaled.shape:\n",
    "    raise Exception(\"Gene effect data frames are not the same shape as expected.\")\n",
    "\n",
    "gene_effect = pd.merge(\n",
    "    gene_effect_scaled, gene_effect_unscaled, on=[\"depmap_id\", \"hugo_symbol\"]\n",
    ")\n",
    "\n",
    "gene_effect.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-05-31T14:49:32.528674Z",
     "iopub.status.busy": "2021-05-31T14:49:32.527774Z",
     "iopub.status.idle": "2021-05-31T14:50:22.349861Z",
     "shell.execute_reply": "2021-05-31T14:50:22.351087Z"
    }
   },
   "outputs": [],
   "source": [
    "gene_effect.to_csv(save_dir / \"achilles_gene_effect.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
