{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare data for modeling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "exceptions must derive from BaseException",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-03b94e7188ef>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mraise\u001b[0m \u001b[0;34m\"TODO: make faster!\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m: exceptions must derive from BaseException"
     ]
    }
   ],
   "source": [
    "raise \"TODO: make faster!\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "import janitor\n",
    "from pathlib import Path\n",
    "import re\n",
    "import plotnine as gg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "gg.theme_set(gg.theme_minimal())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = Path(\"../data\")\n",
    "save_dir = Path(\"../modeling_data\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setup dask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import dask\n",
    "import dask.dataframe as dd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table style=\"border: 2px solid white;\">\n",
       "<tr>\n",
       "<td style=\"vertical-align: top; border: 0px solid white\">\n",
       "<h3 style=\"text-align: left;\">Client</h3>\n",
       "<ul style=\"text-align: left; list-style: none; margin: 0; padding: 0;\">\n",
       "  <li><b>Scheduler: </b>tcp://127.0.0.1:41948</li>\n",
       "  <li><b>Dashboard: </b><a href='http://127.0.0.1:8787/status' target='_blank'>http://127.0.0.1:8787/status</a></li>\n",
       "</ul>\n",
       "</td>\n",
       "<td style=\"vertical-align: top; border: 0px solid white\">\n",
       "<h3 style=\"text-align: left;\">Cluster</h3>\n",
       "<ul style=\"text-align: left; list-style:none; margin: 0; padding: 0;\">\n",
       "  <li><b>Workers: </b>2</li>\n",
       "  <li><b>Cores: </b>4</li>\n",
       "  <li><b>Memory: </b>40.00 GB</li>\n",
       "</ul>\n",
       "</td>\n",
       "</tr>\n",
       "</table>"
      ],
      "text/plain": [
       "<Client: 'tcp://127.0.0.1:41948' processes=2 threads=4, memory=40.00 GB>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from dask.distributed import Client, progress\n",
    "\n",
    "client = Client(n_workers=2, threads_per_worker=2, memory_limit=\"20GB\")\n",
    "client"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cell line information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def show_counts(df, col):\n",
    "    if type(col) != list:\n",
    "        col = [col]\n",
    "\n",
    "    return (\n",
    "        df[col + [\"depmap_id\"]]\n",
    "        .drop_duplicates()\n",
    "        .groupby(col)\n",
    "        .count()\n",
    "        .sort_values(\"depmap_id\", ascending=False)\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lineage</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>lung</th>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>blood</th>\n",
       "      <td>132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>skin</th>\n",
       "      <td>113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lymphocyte</th>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>central_nervous_system</th>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>colorectal</th>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>breast</th>\n",
       "      <td>82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>upper_aerodigestive</th>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bone</th>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ovary</th>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>soft_tissue</th>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pancreas</th>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kidney</th>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gastric</th>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>peripheral_nervous_system</th>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unknown</th>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bile_duct</th>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fibroblast</th>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>uterus</th>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>urinary_tract</th>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>esophagus</th>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>plasma_cell</th>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>liver</th>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cervix</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>thyroid</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prostate</th>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>eye</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>engineered</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>engineered_bone</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>embryo</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>engineered_kidney</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>engineered_breast</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>engineered_prostate</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>engineered_ovary</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>engineered_lung</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>engineered_central_nervous_system</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>engineered_blood</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>adrenal_cortex</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   depmap_id\n",
       "lineage                                     \n",
       "lung                                     273\n",
       "blood                                    132\n",
       "skin                                     113\n",
       "lymphocyte                               109\n",
       "central_nervous_system                   107\n",
       "colorectal                                83\n",
       "breast                                    82\n",
       "upper_aerodigestive                       76\n",
       "bone                                      75\n",
       "ovary                                     74\n",
       "soft_tissue                               71\n",
       "pancreas                                  59\n",
       "kidney                                    57\n",
       "gastric                                   49\n",
       "peripheral_nervous_system                 47\n",
       "unknown                                   45\n",
       "bile_duct                                 43\n",
       "fibroblast                                43\n",
       "uterus                                    39\n",
       "urinary_tract                             39\n",
       "esophagus                                 38\n",
       "plasma_cell                               34\n",
       "liver                                     27\n",
       "cervix                                    22\n",
       "thyroid                                   21\n",
       "prostate                                  13\n",
       "eye                                        9\n",
       "engineered                                 5\n",
       "engineered_bone                            4\n",
       "embryo                                     4\n",
       "engineered_kidney                          3\n",
       "engineered_breast                          2\n",
       "engineered_prostate                        1\n",
       "engineered_ovary                           1\n",
       "engineered_lung                            1\n",
       "engineered_central_nervous_system          1\n",
       "engineered_blood                           1\n",
       "adrenal_cortex                             1"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_info = pd.read_csv(save_dir / \"sample_info.csv\")\n",
    "show_counts(sample_info, \"lineage\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "noncancerous_lineages = [\"unknown\", \"embryo\"]\n",
    "engineered_lineages = sample_info[\n",
    "    sample_info.lineage.str.contains(\"engineer\")\n",
    "].lineage.to_list()\n",
    "\n",
    "ignore_lineages = engineered_lineages + noncancerous_lineages\n",
    "sample_info = sample_info[~sample_info.lineage.isin(ignore_lineages)]\n",
    "\n",
    "sample_info_columns = [\n",
    "    \"depmap_id\",\n",
    "    \"primary_or_metastasis\",\n",
    "    \"lineage\",\n",
    "    \"lineage_subtype\",\n",
    "]\n",
    "sample_info = sample_info[sample_info_columns].drop_duplicates()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## *KRAS* mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove all cell lines with no mutation data.\n",
    "all_samples_with_mutation_data = pd.read_csv(\n",
    "    save_dir / \"ccle_mutations.csv\", low_memory=False\n",
    ").depmap_id.unique()\n",
    "\n",
    "sample_info = sample_info.pipe(\n",
    "    lambda x: x[x.depmap_id.isin(all_samples_with_mutation_data)]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "kras_mutations = pd.read_csv(save_dir / \"kras_mutations.csv\")\n",
    "kras_mutations = kras_mutations[[\"depmap_id\", \"kras_mutation\"]]\n",
    "\n",
    "sample_info = sample_info.merge(kras_mutations, on=\"depmap_id\", how=\"left\").assign(\n",
    "    kras_mutation=lambda x: x.kras_mutation.fillna(\"WT\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>primary_or_metastasis</th>\n",
       "      <th>lineage</th>\n",
       "      <th>lineage_subtype</th>\n",
       "      <th>kras_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACH-000001</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>ovary</td>\n",
       "      <td>ovary_adenocarcinoma</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACH-000002</td>\n",
       "      <td>Primary</td>\n",
       "      <td>blood</td>\n",
       "      <td>AML</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACH-000003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>colorectal</td>\n",
       "      <td>colorectal_adenocarcinoma</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACH-000004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>blood</td>\n",
       "      <td>AML</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACH-000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>blood</td>\n",
       "      <td>AML</td>\n",
       "      <td>WT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    depmap_id primary_or_metastasis     lineage            lineage_subtype  \\\n",
       "0  ACH-000001            Metastasis       ovary       ovary_adenocarcinoma   \n",
       "1  ACH-000002               Primary       blood                        AML   \n",
       "2  ACH-000003                   NaN  colorectal  colorectal_adenocarcinoma   \n",
       "3  ACH-000004                   NaN       blood                        AML   \n",
       "4  ACH-000005                   NaN       blood                        AML   \n",
       "\n",
       "  kras_mutation  \n",
       "0            WT  \n",
       "1            WT  \n",
       "2            WT  \n",
       "3            WT  \n",
       "4            WT  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_info.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Screen data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "achilles_lfc = dd.read_csv(save_dir / \"achilles_logfold_change.csv\").compute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "achilles_guide_map = dd.read_csv(save_dir / \"achilles_guide_map.csv\").compute()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sgrna</th>\n",
       "      <th>replicate_id</th>\n",
       "      <th>lfc</th>\n",
       "      <th>depmap_id</th>\n",
       "      <th>pdna_batch</th>\n",
       "      <th>passes_qc</th>\n",
       "      <th>primary_or_metastasis</th>\n",
       "      <th>lineage</th>\n",
       "      <th>lineage_subtype</th>\n",
       "      <th>kras_mutation</th>\n",
       "      <th>genome_alignment</th>\n",
       "      <th>n_alignments</th>\n",
       "      <th>hugo_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAAATCCAGCAATGCAG</td>\n",
       "      <td>143b-311cas9_repa_p6_batch3</td>\n",
       "      <td>0.289694</td>\n",
       "      <td>ACH-001001</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>Primary</td>\n",
       "      <td>bone</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>G12S</td>\n",
       "      <td>chr10_110964620_+</td>\n",
       "      <td>1</td>\n",
       "      <td>SHOC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAAAATCCAGCAATGCAG</td>\n",
       "      <td>2313287-311cas9_repa_p5_batch3</td>\n",
       "      <td>0.171917</td>\n",
       "      <td>ACH-000948</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>Primary</td>\n",
       "      <td>gastric</td>\n",
       "      <td>gastric_adenocarcinoma</td>\n",
       "      <td>WT</td>\n",
       "      <td>chr10_110964620_+</td>\n",
       "      <td>1</td>\n",
       "      <td>SHOC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAAAAATCCAGCAATGCAG</td>\n",
       "      <td>2313287-311cas9_repb_p5_batch3</td>\n",
       "      <td>-0.522717</td>\n",
       "      <td>ACH-000948</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>Primary</td>\n",
       "      <td>gastric</td>\n",
       "      <td>gastric_adenocarcinoma</td>\n",
       "      <td>WT</td>\n",
       "      <td>chr10_110964620_+</td>\n",
       "      <td>1</td>\n",
       "      <td>SHOC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAAAAATCCAGCAATGCAG</td>\n",
       "      <td>253j-311cas9_repa_p5_batch3</td>\n",
       "      <td>-0.211690</td>\n",
       "      <td>ACH-000011</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>urinary_tract</td>\n",
       "      <td>bladder_carcinoma</td>\n",
       "      <td>WT</td>\n",
       "      <td>chr10_110964620_+</td>\n",
       "      <td>1</td>\n",
       "      <td>SHOC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAAAAATCCAGCAATGCAG</td>\n",
       "      <td>42-mg-ba-311cas9_repa_p6_batch3</td>\n",
       "      <td>-1.067942</td>\n",
       "      <td>ACH-000323</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>Primary</td>\n",
       "      <td>central_nervous_system</td>\n",
       "      <td>glioma</td>\n",
       "      <td>WT</td>\n",
       "      <td>chr10_110964620_+</td>\n",
       "      <td>1</td>\n",
       "      <td>SHOC2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sgrna                     replicate_id       lfc  \\\n",
       "0  AAAAAAATCCAGCAATGCAG      143b-311cas9_repa_p6_batch3  0.289694   \n",
       "1  AAAAAAATCCAGCAATGCAG   2313287-311cas9_repa_p5_batch3  0.171917   \n",
       "2  AAAAAAATCCAGCAATGCAG   2313287-311cas9_repb_p5_batch3 -0.522717   \n",
       "3  AAAAAAATCCAGCAATGCAG      253j-311cas9_repa_p5_batch3 -0.211690   \n",
       "4  AAAAAAATCCAGCAATGCAG  42-mg-ba-311cas9_repa_p6_batch3 -1.067942   \n",
       "\n",
       "    depmap_id  pdna_batch  passes_qc primary_or_metastasis  \\\n",
       "0  ACH-001001           3       True               Primary   \n",
       "1  ACH-000948           3       True               Primary   \n",
       "2  ACH-000948           3       True               Primary   \n",
       "3  ACH-000011           3       True            Metastasis   \n",
       "4  ACH-000323           3       True               Primary   \n",
       "\n",
       "                  lineage         lineage_subtype kras_mutation  \\\n",
       "0                    bone            osteosarcoma          G12S   \n",
       "1                 gastric  gastric_adenocarcinoma            WT   \n",
       "2                 gastric  gastric_adenocarcinoma            WT   \n",
       "3           urinary_tract       bladder_carcinoma            WT   \n",
       "4  central_nervous_system                  glioma            WT   \n",
       "\n",
       "    genome_alignment  n_alignments hugo_symbol  \n",
       "0  chr10_110964620_+             1       SHOC2  \n",
       "1  chr10_110964620_+             1       SHOC2  \n",
       "2  chr10_110964620_+             1       SHOC2  \n",
       "3  chr10_110964620_+             1       SHOC2  \n",
       "4  chr10_110964620_+             1       SHOC2  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "modeling_data = pd.merge(\n",
    "    left=achilles_lfc, right=sample_info, how=\"inner\", on=[\"depmap_id\"]\n",
    ")\n",
    "\n",
    "modeling_data = pd.merge(\n",
    "    left=modeling_data, right=achilles_guide_map, how=\"inner\", on=[\"sgrna\"]\n",
    ")\n",
    "\n",
    "modeling_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Copy number at each guide target"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_segment_mean(cn_data, depmap_id, chromosome, pos):\n",
    "    d = cn_data[(cn_data.depmap_id == depmap_id) & (cn_data.chromosome == chromosome)]\n",
    "    d = d[(d.start <= pos) & (pos <= d.end)]\n",
    "    if len(d) == 0:\n",
    "        return None\n",
    "    elif len(d) > 1:\n",
    "        raise Exception(f\"Data contains more than 1 row of data: {len(d)}\")\n",
    "    else:\n",
    "        return d.segment_mean.to_list()[0]\n",
    "\n",
    "\n",
    "def segmentmean_to_copynumber(seg_mean):\n",
    "    if seg_mean == None:\n",
    "        return None\n",
    "    return 2.0 ** seg_mean\n",
    "\n",
    "\n",
    "def parse_genome_location(gloc):\n",
    "    d = gloc.split(\"_\")\n",
    "    d[0] = d[0].replace(\"chr\", \"\")\n",
    "    return (d[0], int(d[1]))\n",
    "\n",
    "\n",
    "cn_segments = dd.read_csv(save_dir / \"ccle_semgent_cn.csv\").compute()\n",
    "\n",
    "for i in range(0, len(modeling_data)):\n",
    "    genome_loc = modeling_data.loc[i, \"genome_alignment\"]\n",
    "    depmap_id = modeling_data.loc[i, \"depmap_id\"]\n",
    "    g_chr, g_pos = parse_genome_location(genome_loc)\n",
    "\n",
    "    seg_mean = get_segment_mean(\n",
    "        cn_data=cn_segments, depmap_id=depmap_id, chromosome=g_chr, pos=g_pos\n",
    "    )\n",
    "\n",
    "    modeling_data.at[i, \"chromosome\"] = g_chr\n",
    "    modeling_data.at[i, \"chr_position\"] = g_pos\n",
    "    modeling_data.at[i, \"copy_number\"] = segmentmean_to_copynumber(seg_mean)\n",
    "\n",
    "modeling_data.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eg_cellline = \"ACH-001001\"\n",
    "\n",
    "chromosome_order = [str(a) for a in range(1, 23)] + [\"X\"]\n",
    "\n",
    "d = modeling_data[modeling_data.depmap_id == eg_cellline]\n",
    "d[\"chromosome\"] = pd.Categorical(d.chromosome, categories=chromosome_order)\n",
    "\n",
    "gg.options.set_option(\"figure_size\", (12, 8))\n",
    "gg.options.set_option(\"dpi\", 400)\n",
    "\n",
    "(\n",
    "    gg.ggplot(d, gg.aes(\"chr_position\", \"copy_number\", color=\"chromosome\"))\n",
    "    + gg.facet_wrap(\"chromosome\", ncol=4, scales=\"free\")\n",
    "    + gg.geom_line(alpha=0.5, size=0.5)\n",
    "    + gg.geom_point(alpha=0.5, size=0.5)\n",
    "    + gg.theme_minimal()\n",
    "    + gg.theme(\n",
    "        axis_text_x=gg.element_blank(),\n",
    "        legend_position=\"none\",\n",
    "        subplots_adjust={\"hspace\": 0.25, \"wspace\": 0.25},\n",
    "        axis_text_y=gg.element_text(hjust=2),\n",
    "    )\n",
    "    + gg.labs(\n",
    "        x=\"chromosome position\",\n",
    "        y=\"copy number\",\n",
    "        title=f\"Example of copy number variation for a single cell line: {eg_cellline}\",\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mutation data of each gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_mutations_df = pd.read_csv(\n",
    "    save_dir / \"ccle_mutations.csv\", dtype={\"chromosome\": str}, low_memory=False\n",
    ")\n",
    "\n",
    "mutations_data_columns = [\n",
    "    \"depmap_id\",\n",
    "    \"hugo_symbol\",\n",
    "    \"variant_classification\",\n",
    "    \"variant_type\",\n",
    "    \"isdeleterious\",\n",
    "    \"istcgahotspot\",\n",
    "    \"iscosmichotspot\",\n",
    "]\n",
    "\n",
    "\n",
    "def mod_variant_classifications(d):\n",
    "    d.variant_classification = d.variant_classification.fillna(\"unknown\")\n",
    "    d.variant_classification = d.variant_classification.astype(\"str\")\n",
    "    d.variant_classification = [a.lower() for a in d.variant_classification]\n",
    "    return d\n",
    "\n",
    "\n",
    "mutations_df = (\n",
    "    full_mutations_df[mutations_data_columns]\n",
    "    .rename(\n",
    "        columns={\n",
    "            \"isdeleterious\": \"is_deleterious\",\n",
    "            \"istcgahotspot\": \"is_tcga_hotspot\",\n",
    "            \"iscosmichotspot\": \"is_cosmic_hotspot\",\n",
    "        }\n",
    "    )\n",
    "    .pipe(mod_variant_classifications)\n",
    ")\n",
    "\n",
    "\n",
    "def any_mutations_true(mut_data, col_name):\n",
    "    return [any([a[col_name] for a in m]) for m in mut_data]\n",
    "\n",
    "\n",
    "mutations_df = (\n",
    "    pd.DataFrame(\n",
    "        mutations_df.set_index([\"depmap_id\", \"hugo_symbol\"])\n",
    "        .apply(lambda d: d.to_dict(), axis=1)\n",
    "        .groupby([\"depmap_id\", \"hugo_symbol\"])\n",
    "        .agg(mutation_data=lambda x: list(x))\n",
    "    )\n",
    "    .assign(n_muts=lambda d: [len(a) for a in d.mutation_data])\n",
    "    .reset_index()\n",
    "    .assign(\n",
    "        any_deleterious=lambda df: any_mutations_true(\n",
    "            df.mutation_data, \"is_deleterious\"\n",
    "        ),\n",
    "        any_tcga_hotspot=lambda df: any_mutations_true(\n",
    "            df.mutation_data, \"is_tcga_hotspot\"\n",
    "        ),\n",
    "        any_cosmic_hotspot=lambda df: any_mutations_true(\n",
    "            df.mutation_data, \"is_cosmic_hotspot\"\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "\n",
    "mutations_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check that there is one row per `depmap_id` x `hugo_symbol` pair.\n",
    "An error will be raised if this is not true."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = (\n",
    "    mutations_df.groupby([\"depmap_id\", \"hugo_symbol\"])\n",
    "    .count()\n",
    "    .query(\"mutation_data > 1\")\n",
    ")\n",
    "if x.shape[0] > 0:\n",
    "    raise \"More than one row per cell line x gene pair\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "modeling_data = pd.merge(\n",
    "    left=modeling_data, right=mutations_df, how=\"left\", on=[\"depmap_id\", \"hugo_symbol\"]\n",
    ").fillna(\n",
    "    value={\n",
    "        \"n_muts\": 0,\n",
    "        \"any_deleterious\": False,\n",
    "        \"any_tcga_hotspot\": False,\n",
    "        \"any_cosmic_hotspot\": False,\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "modeling_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## A column to indicate if there is a mutation at the guide target location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "modeling_data[\"mutated_at_target\"] = False\n",
    "\n",
    "for i in range(modeling_data.shape[0]):\n",
    "    g_chr = modeling_data.chromosome[i]\n",
    "    g_pos = modeling_data.chr_position[i]\n",
    "    g_depmapid = modeling_data.depmap_id[i]\n",
    "\n",
    "    mut_d = (\n",
    "        full_mutations_df.query(f\"depmap_id == '{g_depmapid}'\")\n",
    "        .query(f\"chromosome == '{g_chr}'\")\n",
    "        .query(f\"start_position <= {g_pos} <= end_position\")\n",
    "    )\n",
    "\n",
    "    if mut_d.shape[0] >= 1:\n",
    "        modeling_data.loc[i, \"mutated_at_target\"] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "modeling_data.groupby(\"mutated_at_target\").count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RNA expression of the target gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_df = (\n",
    "    dd.read_csv(save_dir / \"ccle_expression.csv\")\n",
    "    .rename(columns={\"dep_map_id\": \"depmap_id\"})\n",
    "    .compute()\n",
    ")\n",
    "rna_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "modeling_data = pd.merge(\n",
    "    left=modeling_data, right=rna_df, how=\"left\", on=[\"depmap_id\", \"hugo_symbol\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Final data frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "modeling_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "modeling_data.to_csv(save_dir / \"depmap_modeling_dataframe.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
